---
title: "Class 5: Working with gene lists"
author: "Rui Fu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Class-5}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE, message = FALSE, warning = FALSE, echo = TRUE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center"
)
```

```{r load, message = FALSE, warning = FALSE}
# run before class:
# devtools::install_github('idpt7810/practical-data-analysis')
# BiocManager::install("biomaRt", version = "3.8")

library(pbda)
library(tidyverse)
```

## Did experiment, ran through analysis pipeline, what now?
Handle gene lists

Try to connect prior knowledge to the results (GO terms, or something like `yeast_prot_prop`)

Recall some basic R and dataframe manipulation commands

## Retrieve gene lists from RNA-seq or scRNA-seq
Recall how data is saved in objects

Or whatever file you already saved containing lists of gene names

```{r getlist, message = FALSE, warning = FALSE}
# from DEseq output, get a tibble back
loc <- system.file("extdata", "res_class4.rds", package = 'pbda')
res <- readRDS(loc) # loading a DESeq2 object automatically loads DESeq2

class(res)

head(res)

res_tibble <- as.data.frame(res) %>%
  tibble::rownames_to_column(var = "gene") %>%
  as.tibble()

res_tibble

# res_tibble contains calculations for all genes, pick out "interesting" genes
gene_tibble <- res_tibble %>% 
  filter(!is.na(padj) & padj <= 0.01) %>% # p <= 0.05 is common, but can be lowered, sometimes even 0.1 is acceptable
  filter(log2FoldChange >= 1 | log2FoldChange <= -1) %>%
  arrange(desc(log2FoldChange))

gene_tibble

# get a vector of gene names
gene_vector <- gene_tibble$gene
gene_vector <- gene_tibble %>% pull(gene) # at least 5 more ways to do this
```

```{r getlist2, message = FALSE, warning = FALSE}
# or load already saved file on disk
loc <- system.file("extdata", "gene_tibble.csv", package = 'pbda')
gene_tibble2 <- read_csv(loc) # reads out as tibble
write_csv(gene_tibble, "gene_tibble.csv")

# or write as txt file, without col name
write_csv(gene_tibble %>% select(gene), "gene_tibble.txt", col_names = FALSE) # write from tibble
write_lines(gene_vector, "gene_tibble_v.txt") # write for vector
```

## Feed gene lists to other software outside of R, get enrichment score back
DAVID https://david.ncifcrf.gov/

```{r david, message = FALSE, warning = FALSE, fig.height=5, fig.width=6}
# save output and load in R
loc <- system.file("extdata", "david.txt", package = 'pbda')
david <- read_tsv(loc)

david

# do whatever you want, for instance plot fold change of all genes in row 1
g1 <- david$Genes[1]
g1_split <- str_split(g1, ", ")[[1]]
library(cowplot)
ggplot(gene_tibble %>% filter(gene %in% g1_split), 
       aes(x = gene, y = log2FoldChange, fill = -log10(padj))) +
  geom_bar(stat = "identity") +
  labs(x = david$Term[1]) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_gradient2(low = "white", high = "black")
```

other options:

Panther http://www.geneontology.org/page/go-enrichment-analysis

GSEA (requires ranked list, for instance by fold-change, and requires full list of genes instead of just "significant") http://software.broadinstitute.org/gsea/index.jsp

or, GO and GSEA right in R (will require additional packages):

topGO: https://bioconductor.org/packages/release/bioc/html/topGO.html

fgsea: http://bioconductor.org/packages/release/bioc/html/fgsea.html

## Sometimes transformation of names is needed - `stringr` package, part of tidyverse
Depending on the transcriptome and analysis options used in RNA-seq/scRNA-seq, the list of names can be messy.

```{r stringr, message = FALSE, warning = FALSE}
loc <- system.file("extdata", "ensg_tibble.csv", package = 'pbda')
ensg_tibble <- read_csv(loc)

ensg_tibble
```

`str_replace`: find and replace
(run `?str_replace` for documentation)

```{r str_replace, message = FALSE, warning = FALSE}
ensg_tibble_rm <- ensg_tibble %>% mutate(gene = str_replace(gene, pattern = "ENSG", replacement = ""))

ensg_tibble_rm

# trimming of gene/transcript version
ensg_tibble2 <- ensg_tibble %>% mutate(gene = str_replace(gene, pattern = "\\..*", replacement = "")) # regular expressions are useful and worth the time to learn, see ?base::regex

ensg_tibble2
```

`str_to_lower` and `str_to_upper`: for changing cases

```{r str_to, message = FALSE, warning = FALSE}
gene_tibble_l <- gene_tibble %>% mutate(gene = str_to_lower(gene))

gene_tibble_l

gene_tibble_u <- gene_tibble_l %>% mutate(gene = str_to_upper(gene))

gene_tibble_u # note that this is not ideal for most biology situations, see below for better way
```

`str_length`: get length of strings

```{r str_length, message = FALSE, warning = FALSE}
gene_tibble3 <- gene_tibble %>% mutate(len = str_length(gene))

gene_tibble3
```

`str_detect`: spits out `TRUE` or `FALSE`

```{r str_detect, message = FALSE, warning = FALSE}
# "-AS" denotes antisense genes, we can find all those
gene_tibble_as <- gene_tibble %>% filter(str_detect(gene, "-AS"))

gene_tibble_as
```

`str_c`: concatenate strings

```{r str_c, message = FALSE, warning = FALSE}
# also add annotations of our own
gene_tibble_as2 <- gene_tibble_as %>% mutate(gene = str_c("HUMAN-", gene))

gene_tibble_as2
```

`str_sub`: extract substrings

```{r str_sub, message = FALSE, warning = FALSE}
# to remove annotations
gene_tibble_as3 <- gene_tibble_as2 %>% mutate(gene = str_sub(gene, 2, 4)) # from gene name, "HUMAN-ELOVL2-AS1", the string 2-4 is "UMA"

gene_tibble_as3

gene_tibble_as4 <- gene_tibble_as2 %>% mutate(gene = str_sub(gene, 7, -5)) # extract just the gene name from "HUMAN-ELOVL2-AS1"

gene_tibble_as4
```

## Various conversions in R - BioMart
BioMart can be used outside of R too: http://www.ensembl.org/biomart/martview

We will be using `biomaRt`

```{r biomart, message = FALSE, warning = FALSE}
library(biomaRt)

# biomaRt links to many useful datasets
listDatasets(useMart("ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org")) %>% as_tibble() # hundreds of organisms

listAttributes(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% as_tibble() # thousands of rows/attributes

listFilters(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% as_tibble() # ways to filter the data
```

common usage 1: id to gene symbol conversion

```{r biomartid, message = FALSE, warning = FALSE}
# gene id to gene symbol conversion
human_mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

ensg_sym2 <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"), # columns you want back
                 filters = "ensembl_gene_id",                       # type of info you are providing
                 mart = human_mart,                                 # dataset of certain species
                 values = ensg_tibble2$gene) %>% as_tibble()        # values of the filter

ensg_sym2

ensg_sym <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 mart = human_mart, 
                 values = ensg_tibble$gene) %>% as_tibble() 

ensg_sym # with . version #, can't find matches
```

common usage 2: functional GO term annotation

```{r lookgo, message = FALSE, warning = FALSE}
listAttributes(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% 
  as_tibble() %>% 
  filter(str_detect(description, "GO")) # look at attributes related to GO 
```

```{r getgo, message = FALSE, warning = FALSE}
drerio_mart <- useMart("ensembl", dataset = "drerio_gene_ensembl")

loc <- system.file("extdata", "genes_drerio.csv", package = 'pbda')
genes_drerio <- read_csv(loc)

genes_drerio

genesGO <- getBM(attributes = c("zfin_id_symbol", "go_id", "name_1006", "namespace_1003"), 
                 filters = "zfin_id_symbol", 
                 mart = drerio_mart, 
                 values = genes_drerio$gene, TRUE) %>% as_tibble()

genesGO
genesGO <- genesGO %>% filter(go_id != "") # note that there are lines with empty go terms, remember to remove those

# join GO dataframe with original RNA-seq/scRNA-seq data if needed
genes_drerio_joinedGO <- left_join(genes_drerio, genesGO, 
                                   by = c("gene" = "zfin_id_symbol"))

genes_drerio_joinedGO

# or, get all genes with GO term of interest, by providing multiple filters
genesGO_filtered <- getBM(attributes = c("zfin_id_symbol"), 
                  filters = c("zfin_id_symbol", "go"), 
                  mart = drerio_mart, 
                  values = list(genes_drerio$gene, c("GO:0003677")))

genesGO_filtered
```

common usage 3: ortholog conversion between species

```{r biomartpara, message = FALSE, warning = FALSE}
# example of zebrafish to human gene symbol conversion
genes_human <- getLDS(attributes = c("hgnc_symbol"),
                      filters = "hgnc_symbol", 
                      mart = human_mart, 
                      attributesL = c("zfin_id_symbol"), 
                      martL = drerio_mart, 
                      uniqueRows=T, 
                      values = genes_drerio$gene) %>% as_tibble()

genes_human
```

## Exercise
1. Delete one argument from code above used to generate `genesGO_filtered` to obtain a list of all genes with GO id of "GO:0003677".

2. Take a look at `genesGO`. Describe in pseudo-code, how you could add GO-CC back into the brauer_gene_exp.

## Additional resources/reading
More sophisticated analysis than simply gene lists:
Find enriched sequence elements in promoters, UTRs, etc. See https://bedtools.readthedocs.io/en/latest/, http://meme-suite.org/, etc

WGCNA: weighted correlation network analysis

SCENIC: infer gene regulatory networks and cell types from single-cell RNA-seq data


