---
title: "Class 5: Working with gene lists"
author: "Rui Fu"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include = FALSE, message = FALSE, warning = FALSE, echo = TRUE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "img/class-5-",
  fig.align = "center"
)
```

```{r load, message = FALSE, warning = FALSE}
devtools::install_github('idpt7810/practical-data-analysis')
library(pbda)
library(tidyverse)
```

# Did experiment, ran through analysis pipeline, what now?
Handle gene lists

Try to connect prior knowledge to the results

Recall some basic R and dataframe manipulation commands

### Retrieve gene lists from RNA-seq or scRNA-seq
Recall how data is saved in objects

Or whatever file you already saved containing lists of gene names

```{r getlist, message = FALSE, warning = FALSE}
# from DEseq output, get a tibble back
loc <- system.file("extdata", "res_class4.rds", package = 'pbda')
res <- readRDS(loc)

class(res)
head(res)

res_tibble <- as.data.frame(res) %>%
  tibble::rownames_to_column(var = "gene") %>%
  as.tibble()
res_tibble

# res_tibble contains calculations for all genes, pick out "interesting" genes
gene_tibble <- res_tibble %>% 
  filter(!is.na(padj) & padj <= 0.01) %>% # p <= 0.05 is common, but can be lowered
  filter(log2FoldChange >= 1 | log2FoldChange <= -1) %>%
  arrange(desc(log2FoldChange))
gene_tibble

# get a vector of gene names
gene_vector <- gene_tibble$gene
gene_vector <- gene_tibble %>% pull(gene)
```

```{r getlist2, message = FALSE, warning = FALSE}
# or load already saved file on disk
loc <- system.file("extdata", "gene_tibble.csv", package = 'pbda')
gene_tibble2 <- read_csv(loc) # reads out as tibble
write_csv(gene_tibble, "gene_tibble.csv")

# or write as txt file, without col name
write_delim(gene_tibble %>% select(gene), "gene_tibble.txt", col_names = FALSE) # write from tibble
write.table(gene_vector, "gene_tibble_v.txt", row.names = FALSE, col.names = FALSE, quote = FALSE) # write for vector
```

### Feed gene lists to other software outside of R, get enrichment score back
DAVID https://david.ncifcrf.gov/

```{r david, message = FALSE, warning = FALSE}
# save output and load in R
# for macs: curl -o david.txt -k https://david.ncifcrf.gov/data/download/chart_F3B3023831201544032516222.txt
loc <- system.file("extdata", "david.txt", package = 'pbda')
david <- read_tsv(loc)
View(david)

# do whatever you want, for instance plot fold change of all genes in row 1
g1 <- david$Genes[1]
g1_split <- strsplit(g1, ", ")[[1]]
ggplot(gene_tibble %>% filter(gene %in% g1s), 
       aes(x = gene, y = log2FoldChange)) +
  geom_bar(stat = "identity")
```

Panther http://www.geneontology.org/page/go-enrichment-analysis

GSEA (requires ranked list, for instance by fold-change, and requires full list of genes instead of just "significant") http://software.broadinstitute.org/gsea/index.jsp

### GO and GSEA right in R (will require additional packages):
topGO: https://bioconductor.org/packages/release/bioc/html/topGO.html

fgsea: http://bioconductor.org/packages/release/bioc/html/fgsea.html

### Sometimes transformation of names is needed - stringr package, part of tidyverse
Depending on the transcriptome and analysis options used in RNA-seq/scRNA-seq, the list of names can be messy.

For instance, trimming of gene/transcript version.

```{r stringr, message = FALSE, warning = FALSE}
#exp2 <- exp %>% mutate(id = str_replace(id, pattern = "/.*", replacement = ""))

# some other simple and useful functions
#str_to_upper(x) # note that this is not ideal for most biology situations, see below for better way
#str_to_lower(x)
#str_length()

#str_detect() # can be very useful for filtering
```

### Ortholog conversion in R - BioMart
Sometimes conversion between organisms is needed.

```{r biomart, message = FALSE, warning = FALSE}
# install.packages("biomaRt") if needed
library(biomaRt)

# biomaRt links to many useful datasets
listDatasets(useMart("ENSEMBL_MART_ENSEMBL", host = "www.ensembl.org")) %>% as_tibble() # hundreds of organisms

listAttributes(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% as_tibble() # thousands of rows/attributes

listFilters(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% as_tibble() # ways to filter the data
```

```{r biomartpara, message = FALSE, warning = FALSE}
# example of zebrafish to human gene symbol conversion
human_mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
drerio_mart <- useMart("ensembl", dataset = "drerio_gene_ensembl")

genes_drerio <- read_csv("/Users/rf/genes_drerio.csv")
genes_drerio

genes_human <- getLDS(attributes = c("hgnc_symbol"), 
                      filters = "hgnc_symbol", 
                      mart = human_mart, 
                      attributesL = c("zfin_id_symbol"), 
                      martL = drerio_mart, 
                      uniqueRows=T, 
                      values = genes_drerio$gene) %>% as_tibble()
genes_human
```

```{r biomarttranscript, message = FALSE, warning = FALSE}
# similarly, you can convert transcript ids to gene symbols, etc
```

### Functional annotation in R - BioMart
GO terms are included in the BioMart databases as well.

```{r lookgo, message = FALSE, warning = FALSE}
attributesGO <- listAttributes(useMart("ensembl", dataset = "hsapiens_gene_ensembl")) %>% as_tibble() %>% 
  filter(str_detect(description, "GO")) # look at attributes related to GO 
attributesGO
```

```{r getgo, message = FALSE, warning = FALSE}
# convert a dataframe
genesGO <- getBM(attributes = c("zfin_id_symbol", "go_id", "name_1006", "namespace_1003"), 
                 filters = c("zfin_id_symbol", "with_go"), 
                 mart = drerio_mart, 
                 values = list(genes_drerio$gene, TRUE)) %>% as_tibble()
genesGO

# for a slightly cleaner look, notice that 1 gene symbol was not found or has no GO assigned
genesGO_nested <- genesGO %>%
  group_by(zfin_id_symbol) %>%
  nest(.key = "GO")
genesGO_nested

genesGO_nested$GO[1]

# join GO dataframe with original RNA-seq/scRNA-seq data if needed
genes_drerio_joinedGO <- left_join(genes_drerio, genesGO, 
                                   by = c("gene" = "zfin_id_symbol"))
genes_drerio_joinedGO

# or, get all genes with GO term of interest
genesGO_filtered <- getBM(attributes = c("zfin_id_symbol"), 
                  filters = c("zfin_id_symbol", "go"), 
                  mart = drerio_mart, 
                  values = list(genes_drerio$gene, c("GO:0003677")))
genesGO_filtered
```

### BioMart can be used outside of R too
http://www.ensembl.org/biomart/martview

## Exercise 2
1. Delete one argument from code above to obtain a list of all human genes with GO id of "GO:0048468".

2. Find the shared genes between marker genes of cluster 1 and known list from tcell.csv.

3. List the human ensemble ids of transcription factors from gene list 1.

4. Feed list to GSEA, and giving brief analysis of highly enriched results.

5. ggplot of some sort?

6. add CC back into the bauer set?

### Additional resources/reading
More sophisticated analysis than simply gene lists:

WGCNA: weighted correlation network analysis

SCENIC: infer gene regulatory networks and cell types from single-cell RNA-seq data





