---
title: "Class 7: Heatmaps and Clustering"
author: "Kent Riemondy"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Class-7}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center"
)
library(tidyverse)
library(ComplexHeatmap)
library(pbda)
```

```{r reinstall, eval = F}
devtools::install_github("IDPT7810/practical-data-analysis")
library(tidyverse)
library(ComplexHeatmap)
library(pbda)
```

Goals:
 
  1) Learn additional operations on matrices. 
  
  2) Demonstrate principles to effectively visualize large datasets with heatmaps. 
  
  3) Use clustering algorithms to identify patterns in the data.
  

## Matrix operations

For some of the exercises today we will be using a single cell RNA-seq dataset that
is part of the `pbda` package. For simplicity we will just use the raw matrix rather
than working through a Seurat object. The `esc_mat` matrix contains raw read counts 
for each gene in each cell. 

This dataset contains `r ncol(esc_mat)` cells from mouse embryos 
at various stages of development ([Single-Cell RNA-Seq Reveals Dynamic, Random Monoallelic Gene Expression in Mammalian Cells](http://science.sciencemag.org/content/343/6167/193)). This dataset, as well as many
others are available in a preprocessed form from the Hemberg lab (
https://hemberg-lab.github.io/scRNA.seq.datasets/). Note that the number of cells has been reduced for complexity reasons. 



```{r}
esc_mat[1:3, 1:3]
```

Many mathmatical operations that work on vectors also work on matrices. 

```{r}
sum(esc_mat)
min(esc_mat)
max(esc_mat)
mean(esc_mat)
median(esc_mat)
```

Logical comparisons also work. Note that the `.` is a placeholder that indicates the data being passed from the pipe. 

```{r}
is.na(esc_mat) %>% .[1:3, 1:3]

# equivalently
na_values <- is.na(esc_mat) 
na_values[1:3, 1:3]


(esc_mat > 0) %>% .[1:3, 1:3]
```

How many elements contain NA or zero values in the matrix? Note that `TRUE == 1` and `FALSE == 0`

```{r}
TRUE == 1 
FALSE == 1

sum(c(TRUE, FALSE, TRUE))

sum(is.na(esc_mat))

sum(esc_mat > 0)

sum(esc_mat > 0) / length(esc_mat)
```

### Apply functions per row or per column

A common task with matrices is to compute a row-wise or column-wise summary. We have already used a few functions that perform these operations.

```{r}
colSums(esc_mat) %>% head()
rowSums(esc_mat) %>% head()
```

How would we apply an arbitrary function to operate row-wise or column-wise?

In base R there is a function called `apply`. There is no tidyverse equivalent for operating on matrices at this point. 

```{r}
?apply
```

We can use apply to calculate rowSums and colSums.

```{r}
apply(esc_mat, 1, sum) %>% head()
apply(esc_mat, 2, sum) %>% head()
```

### Exercise

Remove all genes with low expression from the `esc_mat` matrix. Save this new matrix as `norm_mat`. Keep genes that have greater than 10 samples with non-zero counts. How many genes remain? 

```{r}

genes_to_keep <- apply(esc_mat, 1, 
                       function(x) sum(x > 0) > 10)

# same result as above
genes_to_keep <- rowSums(esc_mat > 0) > 10

filtered_mat <- esc_mat[genes_to_keep, ]

nrow(filtered_mat)
```


If the function has multiple arguments, there are two options to supply the arguments to apply.

```{r}
#make a smaller matrix
mat <- esc_mat[1:10, 1:5]

apply(mat, 1, sum, na.rm = TRUE)

apply(mat, 1, function(row) sum(row, na.rm = TRUE))

# same as above
apply(mat, 1, function(x) sum(x, na.rm = TRUE))


# using a custom function
max_minus_min <- function(x){
  max(x) - min(x)
}

apply(mat, 2, max_minus_min)
```



### Exercise

Normalize each the expression values for each cell. Divide each column by the total number of reads in each column, multiply by 10000, and log transform.  


```{r}
# normalize a matrix
norm_mat <- apply(esc_mat, 2, function(x) x / sum(x)) 
norm_mat <- 10000 * norm_mat
norm_mat <- log(norm_mat + 1) # or log1p(norm_mat)


# another approach 
lognorm <- function(vec){
  norm_values <- 10000 * (vec / sum(vec))
  log(norm_values + 1)
}

norm_mat <- apply(esc_mat, 2, lognorm)
```

### Converting matrices to data.frames and tibbles

```{r}
# to data.frame
df <- as.data.frame(mat)

df[1:3, 1:5]

# to tibble
as_tibble(df, rownames = "gene")
# or equivalently
rownames_to_column(df, "gene") %>% as_tibble()

# as one step 

tbl_data <- mat %>% as.data.frame() %>% as_tibble(rownames = "gene")
tbl_data
```

Converting data.frames to matrices is also easily accomplished using `as.matrix()`

```{r}

tbl_mat <- tbl_data %>% 
  as.data.frame()

rownames(tbl_mat) <- tbl_mat$gene
tbl_mat <- tbl_mat[, -1]

tbl_mat <- as.matrix(tbl_mat)

tbl_mat[1:3, 1:3]
class(tbl_mat)

```

Converting vectors to tibbles is also very useful.

```{r}
# make a vector from colnames
col_ids <- colnames(tbl_mat)

#build tibble directly
tibble(sample_names = col_ids)

#additional columns can be added as long as all of the vectors are the same lenght

tibble(sample_names = col_ids,
       id = 1:5)
```

### Exercise

Tidy the `clust_mat` matrix into long format and plot the values across time points for each gene. What do you notice about the gene expression. Which are the most similar? similar expression-levels? similar pattern?


```{r, fig.height=5, fig.width=7}
tidy_mat <- clust_mat %>% 
  as.data.frame() %>% 
  as_tibble(rownames = "gene") %>% 
  gather(timepoint, value, -gene) %>% 
  mutate(timepoint = str_remove(timepoint, "timepoint_"),
         timepoint = as.numeric(timepoint))

ggplot(tidy_mat, aes(x = timepoint, y = value,
                     color = gene)) +
  geom_point() +
  geom_line()

```

```{r, echo = FALSE}
clustering_example_plot <- ggplot(tidy_mat, aes(x = timepoint, y = value,
                     color = gene)) +
  geom_point() +
  geom_line()

```

## Generating Heatmaps using ComplexHeatmap

How do we visualize this dataset to examine relationships between samples? Visualizing the entire scRNA-seq matrix will be computationally expensive and
likely completely unintepretable. Ideally we'd like to reduce the dataset to less than a few thousand genes for visualization.
  
How do we select which genes to plot in a heatmap?
  1) For bulk RNA-Seq, generally first identify differentially expressed genes and make a heatmap of these genes.  
  2) For scRNA-seq you can use the variable genes in the dataset, 
  or genes that are markers of defined clusters.   
  3) Pick genes of interest related to the hypothesis.   
  
```{r heatmaps, eval = F}
?Heatmap
??Heatmap
```

```{r}
gene_variance <- apply(norm_mat, 1, var) %>%
  sort(decreasing = T)

vargenes <- names(gene_variance)[1:30]

head(vargenes)
```

Next we'll pick a color palette and generate a vector of colors to be used in the heatmap.

```{r}
vcols <- viridis::viridis(256)
head(vcols)
```


```{r, fig.height=5, fig.width=7}
var_mat <- norm_mat[vargenes, ]

Heatmap(var_mat, col = vcols, name = "log scale",
        cluster_rows = TRUE,
        cluster_columns = FALSE)
```

Often visualizing gene expression abundance can obscure some patterns, as lower expression values are not as strongly represented. A common data transformation is mean-centering and scaling each gene. This is also known as a z-score. 

`(x - mean(x)) / sd(x)` 

By generating z-scores, the gene expression data now represents standard deviations away from the mean. A built in function in R called `scale` can generate z-scores. This is also easy to do using an apply function. 


```{r zscores}
?scale
```

The scale function performs scaling per column in a matrix. However, it is common in genomics for matricies to be represented with samples (i.e. cells) as columns and genes as rows. This is done as commonly the number of genes >> than the number of samples, and so it is conviently (and computaionally easier) to represent data in this format.  

However, is this a tidy format? 

Are the variables stored as rows or columns?

How do we transpose rows to columns and vice-versa?

```{r, eval = F}
?t
```

```{r}
var_mat[1:3, 1:3]

t(var_mat) %>% .[1:3, 1:3]


zmat <- t(scale(t(var_mat), 
                center = TRUE, 
                scale = TRUE)) 

# confirm that columns are the same after scaling
all(colnames(zmat) == colnames(var_mat))
```


```{r, fig.height=5, fig.width=7}
Heatmap(zmat, col = vcols,
        name = "z-score",
        cluster_rows = TRUE,
        cluster_columns = FALSE)

```

Multiple heatmaps can be generated and plotted together if they 
share genes. For example, if we wanted to compare the zscore values 
to the log expression values. 

```{r, fig.height=5, fig.width=14}
# save each heatmap as an object
h1 <- Heatmap(zmat, col = vcols,
        name = "z-score",
        cluster_rows = TRUE,
        cluster_columns = FALSE)

h2 <- Heatmap(var_mat, col = vcols,
        name = "log",
        cluster_rows = TRUE,
        cluster_columns = FALSE)

# use the + operator to combine
h1 + h2
```

Lastly, heatmaps can be saved in a variety of formats using built in r functions, `pdf`, `png`, `svg`, and `jpeg`. `ggplots` plots can be saved with `ggsave` or the cowplot wrapper `save_plot`.

```{r, eval = F}
pdf("figure1.pdf") # name of plot
  h1 + h2 # code that generates the plot
dev.off() # saves the plot

png("figure1.png") 
  h1 + h2 
dev.off() 

svg("figure1.svg") 
  h1 + h2 
dev.off() 

tiff("figure1.tiff") 
  h1 + h2 
dev.off() 
```

### Exercises

Create a heatmap using the top 200 variables genes, using either zscores or log normalized data. Consult the help menu to generate a heatmap and save it to a file. 

Your heatmap should be made in the following manner:

  1) cluster by row
  2) hide the rownames
  3) add a title to the legend
  4) add a title to the columns

What appears to be happening between the `early2cell` and `mid2cell` stages?

```{r, fig.height=5, fig.width=7}
gene_variance <- apply(norm_mat, 1, var) %>%
  sort(decreasing = T)

vargenes <- names(gene_variance)[1:200]

var_mat <- norm_mat[vargenes, ]

Heatmap(var_mat, col = vcols, name = "log scale",
        cluster_rows = TRUE,
        show_row_names = FALSE,
        cluster_columns = FALSE,
        column_title = "Samples", 
        row_title = "Genes")
```

## Clustering

Reviews: 

  - [Avoiding common pitfalls when clustering biological data](http://stke.sciencemag.org/content/9/432/re6)  
  
  - [How does gene expression clustering work?](https://www.nature.com/articles/nbt1205-1499)  
  
  - [Heatmaps](https://www.nature.com/articles/nmeth.1902)  
  

Consider the following example matrix `clust_mat` which is part of the package. Data and concept from a [blog post](http://www.opiniomics.org/you-probably-dont-understand-heatmaps/) by Mick Watson.

By default many Heatmap packages perform clustering. It's important to know what these defaults do, and when it is appropriate to use them. Consider the following toy dataset, which is clustered by row.

```{r}
clustering_example_plot
```

Here is the heatmap for this data.

```{r, fig.height=5, fig.width=7}
Heatmap(clust_mat, 
        col = vcols,
        cluster_rows = TRUE, 
        cluster_columns = FALSE)

```

Does this clustering make sense? Should I1 and I2 be grouped together? 

By default hierarchical clustering is performed using euclidean distance as a distance metric. There are many distance metrics, and some are more appropriate for different data. For this data, using euclidean distance will group the highly expressed genes (`h1` and `h2`) separate from the lowly expressed (`l1`, `l2`).  However we can see that the pattern of change over time is more similar between `h1` and `l1`

We can see the euclidean distances between genes using `dist`. 

```{r}
dist(clust_mat)
```

The hierarchical clustering is done by default using `dist` as input. 

```{r, fig.height=5, fig.width=7}
hc <- hclust(dist(clust_mat))
plot(hc)
```

We can change the distance metric used for the Heatmap by changing the `clustering_distance_rows` argument. A correlation measure, is not sensitive to abundance, and can recover the shared patterns in the data.

```{r}
cor(t(clust_mat))
```

```{r, fig.height=5, fig.width=7}

Heatmap(clust_mat, col = vcols,
        clustering_distance_rows = "pearson",
        cluster_rows = TRUE, 
        cluster_columns = FALSE)

```

Alternatively we could also generate zscores and use these to plot. 

```{r, fig.height=5, fig.width=7}
zmat <- t(scale(t(clust_mat)))

Heatmap(zmat, col = vcols,
        clustering_distance_rows = "euclidean",
        cluster_rows = TRUE, 
        cluster_columns = FALSE)

```


### Extracting out clusters

First let's return to our heatmap of variable genes from the single cell data. 

```{r, fig.height=5, fig.width=7}
h1 <- Heatmap(var_mat, col = vcols, name = "log scale",
        cluster_rows = TRUE,
        show_row_names = FALSE,
        cluster_columns = FALSE,
        column_title = "Correlation", 
        row_title = "Genes")
h1
```

How do we select out individual clusters? How do we define clusters?

First let's generate the dendogram:

```{r, fig.height=5, fig.width=7}
hc <- hclust(dist(var_mat))
plot(hc)
```

Next we can use `cutree()` to cut the tree into groups

```{r, eval = F}
?cutree
```

```{r, fig.height=5, fig.width=7}
# split into 5 groups
kclusters <- cutree(hc, k = 5)

plot(hc, labels = kclusters)

# split based on height of tree
hclusters <- cutree(hc, h = 20)

plot(hc, labels = hclusters)
```


Now that we have assigned clusters we can replot our heatmaps split out by each cluster.


```{r, fig.height=5, fig.width=7}
h1 <- Heatmap(var_mat, col = vcols, name = "log scale",
        cluster_rows = TRUE,
        show_row_names = FALSE,
        cluster_columns = FALSE,
        split = kclusters, 
        row_title = "Genes")
h1
```


### K-means

K-means clustering is another simple, yet powerful clustering techinuqe. It also scales to much larger datasets than hierarchical clustering. 

description:
  - [interactive description](http://onmyphd.com/?p=k-means.clustering)   
  
  - [detailed description](https://en.wikipedia.org/wiki/K-means_clustering)   
  
  - [determining number of clusters](https://stackoverflow.com/questions/15376075/cluster-analysis-in-r-determine-the-optimal-number-of-clusters)   
  

Kmeans is easily implemented using `kmeans()`

```{r}
clusters <- kmeans(var_mat, centers = 5)
names(clusters)
```

Cluster assignments can be extracted using `$cluster`

```{r}
clusters$cluster %>% head()
```

To recover a reproducible kmeans classification set a seed for the random number generator

```{r}
set.seed(42)
clusters <- kmeans(var_mat, centers = 5)
clusters$cluster %>% head
```

### Exercise:

1) Make a Heatmap split by each kmeans cluster

2) use your tidying skills to plot the average expression per cluster.

