---
title: "Exercises 3"
author: "Your Name Here"
date: "12/7/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

Here is a [link](https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-3.Rmd) to the text of these exercises.

# Question 1

The Encode project has performed `eCLIP` experiments to map RNA binding protein  (RBP) binding sites for > 100 RBPs. Go to the [Encode website](https://www.encodeproject.org/matrix/?type=Experiment), search for `eCLIP` experiments, select an RBP eClip experiment, and load a `BED` file (**bed narrowPeak** file format) with the binding sites into R using just the url. 

Use the [`read_narrowPeak()`](https://rnabioco.github.io/valr/reference/read_bed.html) function from `valr` to read in your bed file using just the URL. Do not download the file locally.

Also, make sure that the file was generated using the `hg19` genome build, and that you do **not** select the `eCLIP mock input` samples, which are negative controls. Display the bed tibble that you pulled from the Encode database and give a one to two sentence summary about the known functions of the RBP. 

Here's an example. An eClip experiment for `PTPB1` was performed ([info here](https://www.encodeproject.org/experiments/ENCSR384KAN/). I selected `hg19` and in the `file details` menu selected a bed narrowPeak file [here](https://www.encodeproject.org/files/ENCFF471GJE/). I copied the URL by right-clicking on `Download ENCFF471GJe` and selecting copy link address.  Note that the extension of the file in the URL is `.bed.gz`, do not use `bigBed`, `fastq`, `bam`, or `bigWig` files. 

```{r q1_example}
library(valr)
url <- "https://www.encodeproject.org/files/ENCFF471GJE/@@download/ENCFF471GJE.bed.gz"

read_narrowpeak(url)
```

## Strategy

```{r q1}



```

## Interpretation

# Question 2

Use `bed_intersect()` and some dplyr functions to determine how many `eCLIP` peaks are found in genes and how many are intergenic. (see the documentation on the `invert` argument for `bed_intersect()`). Use the `hg19_genes.bed.gz` bed file as your gene bed file (location shown below). Explain in the interpretation whether or not the RBP primarily binds to known (geneic) or unannotated (intergeneic) regions. 

```{r, eval = F}
genes <- system.file("extdata", "hg19_genes.bed.gz", package = "eda")
read_bed(genes, n_fields = 6)
```

## Strategy

```{r q2}

```

## Interpretation

# Question 3 

Determine which gene has the largest number of binding sites for your `RBP`. 

## Strategy

```{r q3}

```

## Interpretation

# Question 4

Use another function from `valr` (overview and documentation [here](https://rnabioco.github.io/valr/articles/valr.html) to study a relationship between your eClip binding sites and genes in the `gene` bed file. Explain your question and answer. Some `valr` functions require a `genome` file, which indicates the lengths of the chromosomes.You can access this file as shown below.

```{r, eval = F}
genome_file <- system.file("extdata", "hg19_genome.txt.gz", package = "eda")
genome <- read_genome(genome_file)

genome
```

## Strategy

```{r q4}

```

## Interpretation

