<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Weekend Assignment 2 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Genomic Interval Analysis part 2</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: RNA-Seq Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-7.html">Class 7: RNA-Seq Analysis part 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li>
      <a href="../articles/exercises-4.html">Exercises 4</a>
    </li>
    <li>
      <a href="../articles/exercises-5.html">Exercises 5</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Weekend Assignment 2</h1>
                        <h4 class="author">YOUR NAME HERE</h4>
            
            <h4 class="date">12/05/2017</h4>
          </div>

    
    
<div class="contents">
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Here is a <a href="https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/assignment-2.Rmd">link</a> to the Rmarkdown for this assignment. Download that and update your name at the top.</p>
<p>For this assignment you will be generating a “metagene” plot which summarizes genomic signals over a specific genomic feature in the human genome. This type of analysis is commonly performed with Chip-Seq and Clip-Seq datasets to investigate patterns of DNA or RNA binding protein occupancy around a genomic feature (splice site, transcription start site, start/stop codons, etc.). For this assignment we will be generating a metagene profile surrounding the transcription start site that summarize the coverage of an RNA PolII Chip-Seq experiment genome-wide.</p>
<p>The main idea of this analysis is depicted in Figure 1. We first calculate the coverage across at each individual TSS for every gene, then we average these profiles across every gene to produce a single metagene.</p>
<div class="figure">
<img src="https://github.com/rnabioco/eda/raw/master/vignettes/img/assignment-2-metagene.png" alt="Figure 1. Conceptual metagene of Chip-Seq coverage over the TSS. (A) Shown are examples Chip-Seq read coverage (reads per nucleotide) across intervals +/- 1000 nucleotides surrounding the TSS for 5 genes. (B) Metagene that shows the average coverage binned in 10 nucleotide windows across a TSS metagene +/- 1000 nucleotides surrounding the TSS"><p class="caption">Figure 1. Conceptual metagene of Chip-Seq coverage over the TSS. (A) Shown are examples Chip-Seq read coverage (reads per nucleotide) across intervals +/- 1000 nucleotides surrounding the TSS for 5 genes. (B) Metagene that shows the average coverage binned in 10 nucleotide windows across a TSS metagene +/- 1000 nucleotides surrounding the TSS</p>
</div>
</div>
<div id="files-used-for-this-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#files-used-for-this-analysis" class="anchor"></a>Files used for this analysis:</h2>
<p>You will use the following files in this analysis:</p>
<table style="width:83%;" class="table">
<colgroup>
<col width="47%">
<col width="36%">
</colgroup>
<thead><tr class="header">
<th>Filename</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>hg19_genes.bed.gz</td>
<td>bed file with 6 fields with the start and end coordinates for every human gene (<code>hg19</code>)</td>
</tr>
<tr class="even">
<td>hg19_genome.txt.gz</td>
<td>genome file containing the lengths of every human chromosome</td>
</tr>
<tr class="odd">
<td>polra2_hela_1e6.bg.gz</td>
<td>bedgraph file of RNA Pol II Chip-Seq coverage data (subsampled to 1 million rows), and hosted on a server accessible by URL</td>
</tr>
</tbody>
</table>
<p>These files can be loaded into R by following these steps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(eda)
<span class="kw">library</span>(tidyverse)

genes_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genes.bed.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(genes_file, <span class="dt">n_fields =</span> <span class="dv">6</span>)
genes</code></pre></div>
<pre><code>## # A tibble: 55,006 x 6
##    chrom start   end         name score strand
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;
##  1  chr1 11868 14362 LOC102725121     3      +
##  2  chr1 11873 14409      DDX11L1     3      +
##  3  chr1 14361 29370       WASH7P    11      -
##  4  chr1 17368 17436    MIR6859-1     1      -
##  5  chr1 17368 17436    MIR6859-2     1      -
##  6  chr1 17368 17436    MIR6859-3     1      -
##  7  chr1 17368 17436    MIR6859-4     1      -
##  8  chr1 30365 30503    MIR1302-2     1      +
##  9  chr1 30365 30503    MIR1302-9     1      +
## 10  chr1 30365 30503   MIR1302-10     1      +
## # ... with 54,996 more rows</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genome.txt.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_genome">read_genome</a></span>(genome_file)
genome</code></pre></div>
<pre><code>## # A tibble: 93 x 2
##    chrom      size
##    &lt;chr&gt;     &lt;int&gt;
##  1  chr1 249250621
##  2  chr2 243199373
##  3  chr3 198022430
##  4  chr4 191154276
##  5  chr5 180915260
##  6  chr6 171115067
##  7  chr7 159138663
##  8  chrX 155270560
##  9  chr8 146364022
## 10  chr9 141213431
## # ... with 83 more rows</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(<span class="st">"https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz"</span>, 
                <span class="dt">n_fields =</span> <span class="dv">4</span>)
<span class="co"># rename name to value</span>
bg &lt;-<span class="st"> </span><span class="kw">select</span>(bed, <span class="kw">everything</span>(), <span class="dt">value =</span> name)

<span class="co"># change value from a character to a numeric value</span>
bg &lt;-<span class="st"> </span><span class="kw">mutate</span>(bg, 
              <span class="dt">value =</span> <span class="kw">as.numeric</span>(value))
bg</code></pre></div>
<pre><code>## # A tibble: 1,000,000 x 4
##    chrom  start    end value
##    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt; &lt;dbl&gt;
##  1  chr1  69504  69517   1.5
##  2  chr1 125185 125219   0.0
##  3  chr1 566764 566765  13.5
##  4  chr1 566810 566811  14.6
##  5  chr1 566880 566881  18.7
##  6  chr1 566919 566922  12.1
##  7  chr1 568723 568750   0.0
##  8  chr1 569892 569897   3.1
##  9  chr1 664732 664751   1.6
## 10  chr1 713315 713316  16.5
## # ... with 999,990 more rows</code></pre>
</div>
<div id="assignment" class="section level2">
<h2 class="hasAnchor">
<a href="#assignment" class="anchor"></a>Assignment</h2>
<p>I’ve provides some pseudo-code to get you started with this task. The <code>...</code> is pseudo-code that you will need to fill in with your own code.</p>
<div id="exercise-1" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-1" class="anchor"></a>Exercise 1</h3>
<p>Generate a tibble that contains the positions of transcription start sites for all genes on the plus strand.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss &lt;-<span class="st"> </span>genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(strand <span class="op">==</span><span class="st"> "+"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>...

tss</code></pre></div>
<p>Your <code>tss</code> bed tibble that you generate should look like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 28,035 x 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  11868  11869 LOC102725121     3      +</span>
<span class="co">#&gt;  2  chr1  11873  11874      DDX11L1     3      +</span>
<span class="co">#&gt;  3  chr1  30365  30366    MIR1302-2     1      +</span>
<span class="co">#&gt;  4  chr1  30365  30366    MIR1302-9     1      +</span>
<span class="co">#&gt;  5  chr1  30365  30366   MIR1302-10     1      +</span>
<span class="co">#&gt;  6  chr1  30365  30366   MIR1302-11     1      +</span>
<span class="co">#&gt;  7  chr1  69090  69091        OR4F5     1      +</span>
<span class="co">#&gt;  8  chr1 323891 323892 LOC100132287     3      +</span>
<span class="co">#&gt;  9  chr1 323891 323892 LOC100132062     3      +</span>
<span class="co">#&gt; 10  chr1 323891 323892 LOC100133331     4      +</span>
<span class="co">#&gt; # ... with 28,025 more rows</span></code></pre></div>
</div>
<div id="exercise-2" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-2" class="anchor"></a>Exercise 2</h3>
<p>Generate a tibble that contains the intervals +/- 1000 nucleotides surrounding the TSS position. Conceptually, these are the intervals that are shown in Figure 1A. <code>valr</code> provides the <a href="https://rnabioco.github.io/valr/reference/bed_slop.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop()</a></code></a> function, which adds nucleotides to intervals.</p>
<p><a href="https://rnabioco.github.io/valr/reference/bed_slop.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop()</a></code></a> requires the <code>genome</code> tibble which contains the length of each chromosome. In your Rmarkdown document, speculate as to why this function might require the lengths of each chromosome to produce the correct output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">slop_bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop</a></span>(tss, genome, ...) 

slop_bed</code></pre></div>
<p>Your <code>slop_bed</code> bed tibble that you generate should look like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 28,034 x 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  10868  12869 LOC102725121     3      +</span>
<span class="co">#&gt;  2  chr1  10873  12874      DDX11L1     3      +</span>
<span class="co">#&gt;  3  chr1  29365  31366    MIR1302-2     1      +</span>
<span class="co">#&gt;  4  chr1  29365  31366    MIR1302-9     1      +</span>
<span class="co">#&gt;  5  chr1  29365  31366   MIR1302-10     1      +</span>
<span class="co">#&gt;  6  chr1  29365  31366   MIR1302-11     1      +</span>
<span class="co">#&gt;  7  chr1  68090  70091        OR4F5     1      +</span>
<span class="co">#&gt;  8  chr1 322891 324892 LOC100132287     3      +</span>
<span class="co">#&gt;  9  chr1 322891 324892 LOC100132062     3      +</span>
<span class="co">#&gt; 10  chr1 322891 324892 LOC100133331     4      +</span>
<span class="co">#&gt; # ... with 28,024 more rows</span></code></pre></div>
</div>
<div id="exercise-3" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-3" class="anchor"></a>Exercise 3</h3>
<p>Now, we have obtained intervals surrounding each gene’s TSS, similar to the intervals shown in Figure 1A. For our next task we are going to break up each gene’s TSS interval into smaller bins of 10 nucleotides to give us a high resolution metagene. In following steps, coverage values will be averaged across all genes at the same bin. For example bin 1 represents the first 10 nucleotides of our TSS intervals (i.e. positions -1000 to -990 for every gene).</p>
<p>Smaller bin values will increase the resolution of the final plot, but dramatically increase the size of your bed interval tibble (i.e. if you break every 2001 nt region into 1 nt, then for each gene you generate 2001 new intervals). Similarly, larger bin values will decrease resolution of the metagene, but compute faster.</p>
<p><code>valr</code> provides the <a href="https://rnabioco.github.io/valr/reference/bed_makewindows.htmll"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows()</a></code></a> function, which will split up each interval into sub-intervals (i.e windows). This function has many options to control the size of the windows generated. Read the documentation for <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows()</a></code> to pick an option that will generate non-overlapping 10 nucleotide windows across each TSS interval.</p>
<p>Note that <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows()</a></code> will dramatically increase the size of your output tibble. <strong>Do not select a window size less than 10, as this will generate a tibble with potentially 10s of millions of rows(!).</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss_windows &lt;-<span class="st"> </span>slop_bed <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">bed_makewindow</span>(genome, ...)</code></pre></div>
<p>Your output <code>tss_windows</code> tibble should look like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 5,634,834 x 7</span>
<span class="co">#&gt;    chrom start   end         name score strand .win_id</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;</span>
<span class="co">#&gt;  1  chr1 10868 10878 LOC102725121     3      +       1</span>
<span class="co">#&gt;  2  chr1 10878 10888 LOC102725121     3      +       2</span>
<span class="co">#&gt;  3  chr1 10888 10898 LOC102725121     3      +       3</span>
<span class="co">#&gt;  4  chr1 10898 10908 LOC102725121     3      +       4</span>
<span class="co">#&gt;  5  chr1 10908 10918 LOC102725121     3      +       5</span>
<span class="co">#&gt;  6  chr1 10918 10928 LOC102725121     3      +       6</span>
<span class="co">#&gt;  7  chr1 10928 10938 LOC102725121     3      +       7</span>
<span class="co">#&gt;  8  chr1 10938 10948 LOC102725121     3      +       8</span>
<span class="co">#&gt;  9  chr1 10948 10958 LOC102725121     3      +       9</span>
<span class="co">#&gt; 10  chr1 10958 10968 LOC102725121     3      +      10</span>
<span class="co">#&gt; # ... with 5,634,824 more rows</span></code></pre></div>
</div>
<div id="exercise-4" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-4" class="anchor"></a>Exercise 4</h3>
<p>Examine the <code>tss_windows</code> tibble that you generated in the previous exercise. What information does the <code>.win_id</code> column contain ?</p>
</div>
<div id="exercise-5" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-5" class="anchor"></a>Exercise 5</h3>
<p>We have now obtained a proper set set of intervals to calculate Chip-Seq coverage values. The Chip-seq data that we will use is derived from an Encode experiment in which RNA polII was Chip’ed in Hela-S3 cells. The Chip experiment is described <a href="https://www.encodeproject.org/experiments/ENCSR000EZL/">here</a>. The coverage data is supplied as a <a href="http://genome.ucsc.edu/goldenPath/help/bigWig.html"><code>bigWig</code></a> file, which is an indexed binary format, not suitable for R. I’ve converted the <code>bigWig</code> to a <code>bedGraph</code> file using the <code>bigWigToBedgraph</code> C program available from <code>USCS</code> (only available for <a href="http://hgdownload.soe.ucsc.edu/admin/exe/macOSX.x86_64/">macOS</a> and <a href="http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/">linux</a>). <code>bedGraph</code> files can be read into R using the <code><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed()</a></code> function.</p>
<p>The <code>bedGraph</code> file contains four columns, <code>chrom</code>, <code>start</code>, <code>end</code>, and <code>score</code>. The <code>score</code> column contains the number of reads overlapping the interval. The score is normalized by the total number of reads in the experiment, and is represented as <strong>R</strong>eads <strong>P</strong>er <strong>M</strong>illon:</p>
<p><span class="math display">\[\frac{reads\ overlapping\ the\ interval* 1e6}{total\ number\ of\ reads}\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(<span class="st">"https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz"</span>, 
                <span class="dt">n_fields =</span> <span class="dv">4</span>)
<span class="co"># rename name to value</span>
bg &lt;-<span class="st"> </span><span class="kw">select</span>(bed, <span class="kw">everything</span>(), <span class="dt">value =</span> name)

<span class="co"># change value from a character to a numeric value</span>
bg &lt;-<span class="st"> </span><span class="kw">mutate</span>(bg, 
              <span class="dt">value =</span> <span class="kw">as.numeric</span>(value))
bg</code></pre></div>
<pre><code>## # A tibble: 1,000,000 x 4
##    chrom  start    end value
##    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt; &lt;dbl&gt;
##  1  chr1  69504  69517   1.5
##  2  chr1 125185 125219   0.0
##  3  chr1 566764 566765  13.5
##  4  chr1 566810 566811  14.6
##  5  chr1 566880 566881  18.7
##  6  chr1 566919 566922  12.1
##  7  chr1 568723 568750   0.0
##  8  chr1 569892 569897   3.1
##  9  chr1 664732 664751   1.6
## 10  chr1 713315 713316  16.5
## # ... with 999,990 more rows</code></pre>
<p>Now we would like to perform two steps:<br>
1) For each TSS interval, find all intervals in the <code>bedGraph</code> that overlap<br>
2) For each TSS interval, sum all of the <code>values</code> for each <code>bedGraph</code> interval that overlaps the TSS interval</p>
<p>The <a href="https://rnabioco.github.io/valr/reference/bed_map.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map()</a></code></a> function accomplishes both of these steps. Shown below is a schematic which illustrates the behavior of <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map()</a></code>.</p>
<div class="figure">
<img src="assignment-2_files/figure-html/map_glyph-1.png" alt="Imagine that each X interval is a TSS interval and the Y intervals are the Pol2 ChipSeq bedgraph intervals. The numbers 12, 20, 30, etc. indicate the  Chip-Seq coverage across each y interval. What bed_map() will do is find all intervals in y that overlap each x interval, then allow you to compute a summary statistic, in this case a sum. Conceptually this is similar to the group_by() and summarize() functions in dplyr, but instead of grouping by columns we are grouping y intervals that overlap the x intervals. If there are no Y intervals overlaping an X interval then an NA value is reported." width="672"><p class="caption">
Imagine that each X interval is a TSS interval and the Y intervals are the Pol2 ChipSeq bedgraph intervals. The numbers 12, 20, 30, etc. indicate the Chip-Seq coverage across each y interval. What bed_map() will do is find all intervals in y that overlap each x interval, then allow you to compute a summary statistic, in this case a sum. Conceptually this is similar to the group_by() and summarize() functions in dplyr, but instead of grouping by columns we are grouping y intervals that overlap the x intervals. If there are no Y intervals overlaping an X interval then an NA value is reported.
</p>
</div>
<p>You can pass functions to <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map()</a></code> similar to <code>summarize()</code> such as <code>mean</code>, <code>sum</code>, <code>var</code>, and name their outputs. The function outputs then become new columns in the output tibble.</p>
<p>For example, imagine we have an x and y bed tibble.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x</code></pre></div>
<pre><code>## # A tibble: 3 x 3
##   chrom start   end
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1  chr1     1   100
## 2  chr1   150   175
## 3  chr1   200   300</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y</code></pre></div>
<pre><code>## # A tibble: 5 x 4
##   chrom start   end value
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1  chr1     1    20    10
## 2  chr1    30    50    20
## 3  chr1    80    95    30
## 4  chr1   210   230     5
## 5  chr1   265   290    10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(x, y, 
        <span class="dt">read_total =</span> <span class="kw">sum</span>(value))</code></pre></div>
<pre><code>## # A tibble: 3 x 4
##   chrom start   end read_total
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
## 1  chr1     1   100         60
## 2  chr1   150   175         NA
## 3  chr1   200   300         15</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(x, y, 
        <span class="dt">read_mean =</span> <span class="kw">mean</span>(value))</code></pre></div>
<pre><code>## # A tibble: 3 x 4
##   chrom start   end read_mean
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;
## 1  chr1     1   100      20.0
## 2  chr1   150   175        NA
## 3  chr1   200   300       7.5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(x, y, 
        <span class="dt">read_mean =</span> <span class="kw">mean</span>(value),
        <span class="dt">read_total =</span> <span class="kw">sum</span>(value))</code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   chrom start   end read_mean read_total
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
## 1  chr1     1   100      20.0         60
## 2  chr1   150   175        NA         NA
## 3  chr1   200   300       7.5         15</code></pre>
<p>Using <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map()</a></code> calculate the total coverage (i.e. <code>sum()</code>) of the Chip-Seq coverage values (the <code>bedGraph</code> named <code>bg</code>) across the tss window intervals (<code>tss_windows</code>). This function might take 30-60 seconds to complete.</p>
<p>If you are having trouble getting the proper output, it may be useful to generate a smaller tss_windows bed file to test your function before running on the whole dataset (use <code>sample_n()</code>), so that you don’t have to wait 30-60 second per try.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chip_coverage &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(tss_windows, 
                         bg, 
                         ...)</code></pre></div>
<p>The expected output is shown below. Note that the <code>total_reads</code> column name may another name depending on what you assign your summary statistic to (i.e. <code>total_reads = sum(value)</code>, or <code>total_coverage = sum(value)</code>). Also note that there are many NA values in the total_reads column. This is expected as not every gene is being transcribed in the genome, so some regions have no coverage in the ChIP-Seq experiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 5,626,622 x 8</span>
<span class="co">#&gt;    chrom start   end         name score strand .win_id total_reads</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;       &lt;dbl&gt;</span>
<span class="co">#&gt;  1  chr1 10868 10878 LOC102725121     3      +       1          NA</span>
<span class="co">#&gt;  2  chr1 10873 10883      DDX11L1     3      +       1          NA</span>
<span class="co">#&gt;  3  chr1 10878 10888 LOC102725121     3      +       2          NA</span>
<span class="co">#&gt;  4  chr1 10883 10893      DDX11L1     3      +       2          NA</span>
<span class="co">#&gt;  5  chr1 10888 10898 LOC102725121     3      +       3          NA</span>
<span class="co">#&gt;  6  chr1 10893 10903      DDX11L1     3      +       3          NA</span>
<span class="co">#&gt;  7  chr1 10898 10908 LOC102725121     3      +       4          NA</span>
<span class="co">#&gt;  8  chr1 10903 10913      DDX11L1     3      +       4          NA</span>
<span class="co">#&gt;  9  chr1 10908 10918 LOC102725121     3      +       5          NA</span>
<span class="co">#&gt; 10  chr1 10913 10923      DDX11L1     3      +       5          NA</span>
<span class="co">#&gt; # ... with 5,626,612 more rows</span></code></pre></div>
</div>
<div id="exercise-6" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-6" class="anchor"></a>Exercise 6</h3>
<p>Now we have arrived at the exciting part of the analysis. To review, we first defined a set of intervals surrounding each TSS. Then we split these intervals into bins of sub-intervals. Finally, we summed up the Chip-Seq coverage across each sub-interval. Now our bed file <code>chip_coverage</code> contains the RNA polII coverage surrounding the TSS for every gene across the genome. Now in this last step we will summarize the coverage in each bin and plot as a metagene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">group_by</span>(chip_coverage, ...) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(...) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>()</code></pre></div>
<p>The x-axis values displayed will not be ideal. To center the x-axis values around zero use the following code.</p>
<p>Add informative <code>x-axis</code>, <code>y-axis</code>, and <code>main</code> plot titles and explain what the plot demonstrates. Additionally add a figure caption using the <code>fig.cap="your caption"</code> chunk option.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">region_size &lt;-<span class="st"> </span><span class="dv">1000</span>
win_size &lt;-<span class="st"> </span><span class="dv">10</span>

x_labels &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span>region_size, region_size, <span class="dt">by =</span> win_size <span class="op">*</span><span class="st"> </span><span class="dv">25</span>)
x_breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">201</span>, <span class="dt">by =</span> <span class="dv">25</span>)

<span class="co"># add this to your ggplot call:</span>
<span class="kw">scale_x_continuous</span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks) </code></pre></div>
<div class="figure">
<img src="https://github.com/rnabioco/eda/raw/master/vignettes/img/assignment-2-metagene_plot.png" alt="Figure 2. Example metagene plot"><p class="caption">Figure 2. Example metagene plot</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#files-used-for-this-analysis">Files used for this analysis:</a></li>
      <li><a href="#assignment">Assignment</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
