<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 7: RNA-seq Analysis part 2 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Genomic Interval Analysis part 2</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: RNA-Seq Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-7.html">Class 7: RNA-Seq Analysis part 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li>
      <a href="../articles/exercises-4.html">Exercises 4</a>
    </li>
    <li>
      <a href="../articles/exercises-5.html">Exercises 5</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 7: RNA-seq Analysis part 2</h1>
                        <h4 class="author">Austin Gillen</h4>
            
            <h4 class="date">2017-12-14</h4>
          </div>

    
    
<div class="contents">
<div id="single-cell-rna-seq-analysis-in-r-with-seurat" class="section level1">
<h1 class="hasAnchor">
<a href="#single-cell-rna-seq-analysis-in-r-with-seurat" class="anchor"></a>Single cell RNA-seq analysis in R with Seurat</h1>
<p>(Adapted from the <a href="http://satijalab.org/seurat/pbmc3k_tutorial.html">Seurat PBMC 3k Tutorial</a>)</p>
<ul>
<li>Introduce sparse matrices</li>
<li>QC considerations for scRNA-seq data sets</li>
<li>Build a seurat object, cluster cells and and calculate differential gene expression</li>
</ul>
<div id="after-updating-eda-a-new-package-named-seurat-should-be-installed" class="section level2">
<h2 class="hasAnchor">
<a href="#after-updating-eda-a-new-package-named-seurat-should-be-installed" class="anchor"></a>After updating eda, a new package named <code>Seurat</code> should be installed</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(eda)

<span class="kw"><a href="../reference/update_eda.html">update_eda</a></span>()

<span class="kw">library</span>(Seurat)
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(Matrix)</code></pre></div>
</div>
<div id="single-cell-rna-seq-analysis-in-r-starts-with-the-same-data-as-bulk-rna-seq-a-count-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-rna-seq-analysis-in-r-starts-with-the-same-data-as-bulk-rna-seq-a-count-matrix" class="anchor"></a>Single cell RNA-seq analysis in R starts with the same data as bulk RNA-seq: a count matrix</h2>
<p>However, for efficiency, scRNA-seq data is stored in sparse matrix objects. An example data set consisting of 2,700 peripheral blood mononuclear cells (PMBCs) is included in the eda package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the path to the example data</span>
pbmc_10x &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pbmc_gene_bc_matrices"</span>, <span class="st">"hg19"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)

<span class="co"># read the example data</span>
pbmc.data &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/Read10X">Read10X</a></span>(pbmc_10x)

<span class="co"># print the size of the sparse matrix</span>
<span class="kw">object.size</span>(<span class="dt">x =</span> pbmc.data) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">format</span>(<span class="dt">units =</span> <span class="st">"auto"</span>)
<span class="co">#&gt; [1] "36.9 Mb"</span>

<span class="co"># print the size of the full matrix</span>
<span class="kw">object.size</span>(<span class="kw">as.matrix</span>(<span class="dt">x =</span> pbmc.data)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">format</span>(<span class="dt">units =</span> <span class="st">"auto"</span>)
<span class="co">#&gt; [1] "676.4 Mb"</span></code></pre></div>
</div>
<div id="seurat-uses-a-custom-object-to-store-counts-and-data-similar-to-the-summarizedexperiment-deseqdataset" class="section level2">
<h2 class="hasAnchor">
<a href="#seurat-uses-a-custom-object-to-store-counts-and-data-similar-to-the-summarizedexperiment-deseqdataset" class="anchor"></a>Seurat uses a custom object to store counts and data (similar to the SummarizedExperiment &amp; DESeqDataSet)</h2>
<p>First, we’ll generate a Seurat object with the raw count data, keeping all genes that are expressed in at least 3 cells and all cells with at least 200 detectable genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/CreateSeuratObject">CreateSeuratObject</a></span>(<span class="dt">raw.data =</span> pbmc.data,
                           <span class="dt">min.cells =</span> <span class="dv">3</span>,
                           <span class="dt">min.genes =</span> <span class="dv">200</span>,
                           <span class="dt">project =</span> <span class="st">"10x_PBMC"</span>)</code></pre></div>
</div>
<div id="filtering-and-qc-for-scrna-seq" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-and-qc-for-scrna-seq" class="anchor"></a>Filtering and QC for scRNA-seq</h2>
<p>scRNA-seq data sets are further filtered to remove cells with too few detectable genes, cells with clear outlier numbers of detectable genes (likely doublets) and cells with high proportions of mitochondrial RNAs (likely dead cells).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Calculate percent of reads that are mitochondrial</span>
mito.genes &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="dt">pattern =</span> <span class="st">"^MT-"</span>,
                   <span class="dt">x =</span> <span class="kw">rownames</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>data),
                   <span class="dt">value =</span> <span class="ot">TRUE</span>)

percent.mito &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/Matrix/topics/colSums">colSums</a></span>(pbmc<span class="op">@</span>raw.data[mito.genes, ])<span class="op">/</span>Matrix<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/Matrix/topics/colSums">colSums</a></span>(pbmc<span class="op">@</span>raw.data)

<span class="co"># Add percent.mite as metadata to Seurat object</span>
pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(<span class="dt">object =</span> pbmc,
                    <span class="dt">metadata =</span> percent.mito,
                    <span class="dt">col.name =</span> <span class="st">"percent.mito"</span>)

<span class="co"># Plot number of genes, number of UMIs and percent.mito</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/VlnPlot">VlnPlot</a></span>(<span class="dt">object =</span> pbmc,
        <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">"nGene"</span>, <span class="st">"nUMI"</span>, <span class="st">"percent.mito"</span>),
        <span class="dt">nCol =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="img/class-7-filtering-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Outliers can be identified visually using the <code>GenePlot</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/GenePlot">GenePlot</a></span>(<span class="dt">object =</span> pbmc, <span class="dt">gene1 =</span> <span class="st">"nUMI"</span>, <span class="dt">gene2 =</span> <span class="st">"percent.mito"</span>)</code></pre></div>
<p><img src="img/class-7-outlier_plot-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/GenePlot">GenePlot</a></span>(<span class="dt">object =</span> pbmc, <span class="dt">gene1 =</span> <span class="st">"nUMI"</span>, <span class="dt">gene2 =</span> <span class="st">"nGene"</span>)</code></pre></div>
<p><img src="img/class-7-outlier_plot-2.png" width="576" style="display: block; margin: auto;"></p>
<p>Finally, we can filter the pbmc object with <code>FilterCells</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/FilterCells">FilterCells</a></span>(<span class="dt">object =</span> pbmc,
                    <span class="dt">subset.names =</span> <span class="kw">c</span>(<span class="st">"nGene"</span>, <span class="st">"percent.mito"</span>), 
                    <span class="dt">low.thresholds =</span> <span class="kw">c</span>(<span class="dv">200</span>, <span class="op">-</span><span class="ot">Inf</span>),
                    <span class="dt">high.thresholds =</span> <span class="kw">c</span>(<span class="dv">2500</span>, <span class="fl">0.05</span>))</code></pre></div>
</div>
<div id="normalizing-the-data-detecting-variable-genes-and-removing-uninteresting-variation" class="section level2">
<h2 class="hasAnchor">
<a href="#normalizing-the-data-detecting-variable-genes-and-removing-uninteresting-variation" class="anchor"></a>Normalizing the data, detecting variable genes, and removing uninteresting variation</h2>
<p>After filtering, we need to normalize gene expression across cells with the <code>NormalizeData</code> function. This function normalizes gene expression measurements by per-cell total expression, multiplies the result by an arbitrary value (10,000 here) and log transforms the result.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/NormalizeData">NormalizeData</a></span>(<span class="dt">object =</span> pbmc,
                      <span class="dt">normalization.method =</span> <span class="st">"LogNormalize"</span>, 
                      <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</code></pre></div>
<p>Next, we identify variable genes using a mean-dispersion plot to mark visual outliers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/FindVariableGenes">FindVariableGenes</a></span>(<span class="dt">object =</span> pbmc,
                          <span class="dt">mean.function =</span> ExpMean,
                          <span class="dt">dispersion.function =</span> LogVMR, 
                          <span class="dt">x.low.cutoff =</span> <span class="fl">0.0125</span>,
                          <span class="dt">x.high.cutoff =</span> <span class="dv">3</span>,
                          <span class="dt">y.cutoff =</span> <span class="fl">0.5</span>)</code></pre></div>
<p><img src="img/class-7-pkg_data-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw">length</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>var.genes)
<span class="co">#&gt; [1] 1838</span></code></pre></div>
<p>Single cell data sets contain various types of ‘uninteresting’ variation. These can include technical noise, batch effects and confounding biological variation (such as cell cycle stage). We can regress these signals out of the analysis using <code>ScaleData</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/ScaleData">ScaleData</a></span>(<span class="dt">object =</span> pbmc,
                  <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">"nUMI"</span>, <span class="st">"percent.mito"</span>))</code></pre></div>
</div>
<div id="performing-dimensionality-reduction-and-clustering-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-dimensionality-reduction-and-clustering-pca" class="anchor"></a>Performing dimensionality reduction and clustering: PCA</h2>
<p>As in our bulk RNA-seq example, we need a way to project highly multidimensional data onto a 2D plane. PCA alone is insufficient to capture the variation in these data due to the complexity, so we’ll introduce a new method: t-distributed stochastic neighbor embedding (tSNE). A useful interactive tool to help you understand tSNE can be found <a href="https://distill.pub/2016/misread-tsne/">here</a>. First, we’ll calculate the principle components (used by the clustering and tSNE algorithms):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/RunPCA">RunPCA</a></span>(<span class="dt">object =</span> pbmc,
               <span class="dt">pc.genes =</span> pbmc<span class="op">@</span>var.genes,
               <span class="dt">do.print =</span> <span class="ot">TRUE</span>,
               <span class="dt">pcs.print =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,
               <span class="dt">genes.print =</span> <span class="dv">5</span>)
<span class="co">#&gt; [1] "PC1"</span>
<span class="co">#&gt; [1] "CST3"   "TYROBP" "FCN1"   "LST1"   "AIF1"  </span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "PTPRCAP" "IL32"    "LTB"     "CD2"     "CTSW"   </span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "PC2"</span>
<span class="co">#&gt; [1] "NKG7" "GZMB" "PRF1" "CST7" "GZMA"</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "CD79A"    "MS4A1"    "HLA-DQA1" "TCL1A"    "HLA-DQB1"</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "PC3"</span>
<span class="co">#&gt; [1] "PF4"   "PPBP"  "SDPR"  "SPARC" "GNG11"</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "CYBA"     "HLA-DPA1" "HLA-DPB1" "HLA-DRB1" "CD37"    </span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "PC4"</span>
<span class="co">#&gt; [1] "IL32"   "GIMAP7" "AQP3"   "FYB"    "MAL"   </span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "CD79A"    "HLA-DQA1" "CD79B"    "MS4A1"    "HLA-DQB1"</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "PC5"</span>
<span class="co">#&gt; [1] "FCGR3A"        "CTD-2006K23.1" "IFITM3"        "ABI3"         </span>
<span class="co">#&gt; [5] "CEBPB"        </span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] "FCER1A"  "LGALS2"  "MS4A6A"  "S100A8"  "CLEC10A"</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; [1] ""</span></code></pre></div>
<p>Next, we’ll generate a PCA plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/PCAPlot">PCAPlot</a></span>(<span class="dt">object =</span> pbmc,
        <span class="dt">dim.1 =</span> <span class="dv">1</span>, 
        <span class="dt">dim.2 =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="img/class-7-pca_plot-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Multiple subpopulations are clearly visible in this plot, but as we’ll see later, there is even more interesting variation present in these data. However, in order to use the clustering and tSNE algorithms, we’ll need to determine how many PCs comprise the ‘true dimensionality’ of the data set. One approach is to view a heatmap of the 500 most extreme cells for each principle component.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/PCHeatmap">PCHeatmap</a></span>(<span class="dt">object =</span> pbmc,
          <span class="dt">pc.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>,
          <span class="dt">cells.use =</span> <span class="dv">500</span>,
          <span class="dt">do.balanced =</span> <span class="ot">TRUE</span>, 
          <span class="dt">label.columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">use.full =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="img/class-7-pca_heatmap-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Another approach is the jackstraw plot, which compares the distribution of p-values for each principle component with a uniform distribution. ‘Significant’ PCs show an enrichment of genes with low p-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/JackStraw">JackStraw</a></span>(<span class="dt">object =</span> pbmc,
                  <span class="dt">num.replicate =</span> <span class="dv">100</span>,
                  <span class="dt">do.print =</span> <span class="ot">FALSE</span>)

<span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/JackStrawPlot">JackStrawPlot</a></span>(<span class="dt">object =</span> pbmc, <span class="dt">PCs =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>)</code></pre></div>
<div class="figure">
<img src="img/jackstraw.png">
</div>
<p>Finally, we can generate an ‘elbow plot’ of the standard deviation of the principle components. The ‘elbow’ indicated the last significant principle component.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/PCElbowPlot">PCElbowPlot</a></span>(<span class="dt">object =</span> pbmc)</code></pre></div>
<p><img src="img/class-7-elbow_plot-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Based on these plots, we conclude that the first 7-10 PCs contain sufficient variation for use downstream.</p>
</div>
<div id="performing-dimensionality-reduction-and-clustering-clustering-and-tsne" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-dimensionality-reduction-and-clustering-clustering-and-tsne" class="anchor"></a>Performing dimensionality reduction and clustering: Clustering and tSNE</h2>
<p>Now that we’ve identified the ‘true dimensionality’ of the data, we can cluster the cells and project them on a tSNE plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/FindClusters">FindClusters</a></span>(<span class="dt">object =</span> pbmc,
                     <span class="dt">reduction.type =</span> <span class="st">"pca"</span>,
                     <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, 
                     <span class="dt">resolution =</span> <span class="fl">0.6</span>,
                     <span class="dt">print.output =</span> <span class="dv">0</span>,
                     <span class="dt">save.SNN =</span> <span class="ot">TRUE</span>)

pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/RunTSNE">RunTSNE</a></span>(<span class="dt">object =</span> pbmc,
                <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,
                <span class="dt">do.fast =</span> <span class="ot">TRUE</span>)

<span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(<span class="dt">object =</span> pbmc)</code></pre></div>
<p><img src="img/class-7-find_clusters-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Clustering reveals seven distinct populations that are accurately separated in the tSNE projection - substantially more complexity than was captured in the PCA plot!</p>
</div>
<div id="identifying-marker-genes-differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#identifying-marker-genes-differential-expression" class="anchor"></a>Identifying marker genes (differential expression)</h2>
<p>Seurat provides the <code>FindAllMarkers</code> and <code>FindMarkers</code> functions for identifying differentially expressed genes in all clusters vs. all other clusters and in specific pairs of clusters, respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc.markers &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/FindAllMarkers">FindAllMarkers</a></span>(<span class="dt">object =</span> pbmc,
                               <span class="dt">only.pos =</span> <span class="ot">TRUE</span>,
                               <span class="dt">min.pct =</span> <span class="fl">0.25</span>, 
                               <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)

pbmc.markers <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">top_n</span>(<span class="dv">2</span>, avg_diff)
<span class="co">#&gt; # A tibble: 16 x 6</span>
<span class="co">#&gt; # Groups:   cluster [8]</span>
<span class="co">#&gt;            p_val avg_diff pct.1 pct.2 cluster     gene</span>
<span class="co">#&gt;            &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;fctr&gt;    &lt;chr&gt;</span>
<span class="co">#&gt;  1 9.480623e-268 1.149058 0.924 0.483       0     LDHB</span>
<span class="co">#&gt;  2 3.519020e-133 1.068122 0.662 0.202       0     IL7R</span>
<span class="co">#&gt;  3  0.000000e+00 3.827593 0.996 0.216       1   S100A9</span>
<span class="co">#&gt;  4  0.000000e+00 3.786535 0.973 0.123       1   S100A8</span>
<span class="co">#&gt;  5  0.000000e+00 2.977399 0.936 0.042       2    CD79A</span>
<span class="co">#&gt;  6 1.030004e-191 2.492236 0.624 0.022       2    TCL1A</span>
<span class="co">#&gt;  7 3.158793e-231 2.158812 0.974 0.230       3     CCL5</span>
<span class="co">#&gt;  8 4.103013e-125 2.113428 0.588 0.050       3     GZMK</span>
<span class="co">#&gt;  9 1.304430e-165 2.151881 1.000 0.316       4     LST1</span>
<span class="co">#&gt; 10 1.420369e-138 2.275509 0.962 0.137       4   FCGR3A</span>
<span class="co">#&gt; 11 1.127318e-215 3.763928 0.961 0.131       5     GNLY</span>
<span class="co">#&gt; 12 8.182907e-187 3.334634 0.955 0.068       5     GZMB</span>
<span class="co">#&gt; 13  3.261141e-48 2.729243 0.844 0.011       6   FCER1A</span>
<span class="co">#&gt; 14  7.655042e-30 1.965168 1.000 0.513       6 HLA-DPB1</span>
<span class="co">#&gt; 15  1.165274e-56 5.889503 1.000 0.023       7     PPBP</span>
<span class="co">#&gt; 16  3.238887e-44 4.952160 0.933 0.010       7      PF4</span></code></pre></div>
<p>Differentially expressed genes can be visualized as violin plots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/VlnPlot">VlnPlot</a></span>(<span class="dt">object =</span> pbmc,
        <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">"LDHB"</span>, <span class="st">"CD79A"</span>))</code></pre></div>
<p><img src="img/class-7-plot_genes_violin-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Or as ‘feature plots’ on the tSNE projection:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/FeaturePlot">FeaturePlot</a></span>(<span class="dt">object =</span> pbmc,
            <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">"MS4A1"</span>,
                              <span class="st">"GNLY"</span>,
                              <span class="st">"CD3E"</span>,
                              <span class="st">"CD14"</span>, 
                              <span class="st">"FCER1A"</span>,
                              <span class="st">"FCGR3A"</span>,
                              <span class="st">"LYZ"</span>,
                              <span class="st">"PPBP"</span>,
                              <span class="st">"CD8A"</span>),
            <span class="dt">cols.use =</span> <span class="kw">c</span>(<span class="st">"grey"</span>, <span class="st">"blue"</span>), 
            <span class="dt">reduction.use =</span> <span class="st">"tsne"</span>)</code></pre></div>
<p><img src="img/class-7-plot_genes_feature-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="manipulating-seurat-objects-renaming-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#manipulating-seurat-objects-renaming-clusters" class="anchor"></a>Manipulating Seurat objects: renaming clusters</h2>
<p>The marker genes identified above can be used to reclassify clusters as known cell types. This provides a useful example of how metadata is handled in Seurat.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a vector of current cluster IDs</span>
current.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>)

<span class="co"># Create a vector of new cluster IDs (cell types)</span>
new.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"CD4 T cells"</span>,
                     <span class="st">"CD14+ Monocytes"</span>,
                     <span class="st">"B cells"</span>,
                     <span class="st">"CD8 T cells"</span>, 
                     <span class="st">"FCGR3A+ Monocytes"</span>,
                     <span class="st">"NK cells"</span>,
                     <span class="st">"Dendritic cells"</span>,
                     <span class="st">"Megakaryocytes"</span>)

<span class="co"># Save the cluster numbers as "cluster.id"</span>
pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/StashIdent">StashIdent</a></span>(<span class="dt">object =</span> pbmc, <span class="dt">save.name =</span> <span class="st">"cluster.id"</span>)

<span class="co"># Use plyr to remap the idents from the current IDs to the new IDs</span>
pbmc<span class="op">@</span>ident &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/plyr/topics/mapvalues">mapvalues</a></span>(<span class="dt">x =</span> pbmc<span class="op">@</span>ident,
                              <span class="dt">from =</span> current.cluster.ids,
                              <span class="dt">to =</span> new.cluster.ids)

<span class="co"># Plot the tSNE with new cluster IDs</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(<span class="dt">object =</span> pbmc,
         <span class="dt">do.label =</span> <span class="ot">TRUE</span>,
         <span class="dt">pt.size =</span> <span class="fl">0.5</span>)</code></pre></div>
<p><img src="img/class-7-rename_clusters-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># Plot the tSNE with old cluster IDs</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(<span class="dt">object =</span> pbmc,
         <span class="dt">do.label =</span> <span class="ot">TRUE</span>,
         <span class="dt">pt.size =</span> <span class="fl">0.5</span>,
         <span class="dt">group.by =</span> <span class="st">"cluster.id"</span>)</code></pre></div>
<p><img src="img/class-7-rename_clusters-2.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="dealing-with-proliferating-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#dealing-with-proliferating-cells" class="anchor"></a>Dealing with proliferating cells</h2>
<p>PBMCs are poor examples for this analysis, but the commands are shown here to demonstrate how other, more complex variables are regressed out of these analyses.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Inspect the classifying genes</span>
cc.genes
<span class="co">#&gt; $s.genes</span>
<span class="co">#&gt;  [1] "MCM5"     "PCNA"     "TYMS"     "FEN1"     "MCM2"     "MCM4"    </span>
<span class="co">#&gt;  [7] "RRM1"     "UNG"      "GINS2"    "MCM6"     "CDCA7"    "DTL"     </span>
<span class="co">#&gt; [13] "PRIM1"    "UHRF1"    "MLF1IP"   "HELLS"    "RFC2"     "RPA2"    </span>
<span class="co">#&gt; [19] "NASP"     "RAD51AP1" "GMNN"     "WDR76"    "SLBP"     "CCNE2"   </span>
<span class="co">#&gt; [25] "UBR7"     "POLD3"    "MSH2"     "ATAD2"    "RAD51"    "RRM2"    </span>
<span class="co">#&gt; [31] "CDC45"    "CDC6"     "EXO1"     "TIPIN"    "DSCC1"    "BLM"     </span>
<span class="co">#&gt; [37] "CASP8AP2" "USP1"     "CLSPN"    "POLA1"    "CHAF1B"   "BRIP1"   </span>
<span class="co">#&gt; [43] "E2F8"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $g2m.genes</span>
<span class="co">#&gt;  [1] "HMGB2"   "CDK1"    "NUSAP1"  "UBE2C"   "BIRC5"   "TPX2"    "TOP2A"  </span>
<span class="co">#&gt;  [8] "NDC80"   "CKS2"    "NUF2"    "CKS1B"   "MKI67"   "TMPO"    "CENPF"  </span>
<span class="co">#&gt; [15] "TACC3"   "FAM64A"  "SMC4"    "CCNB2"   "CKAP2L"  "CKAP2"   "AURKB"  </span>
<span class="co">#&gt; [22] "BUB1"    "KIF11"   "ANP32E"  "TUBB4B"  "GTSE1"   "KIF20B"  "HJURP"  </span>
<span class="co">#&gt; [29] "CDCA3"   "HN1"     "CDC20"   "TTK"     "CDC25C"  "KIF2C"   "RANGAP1"</span>
<span class="co">#&gt; [36] "NCAPD2"  "DLGAP5"  "CDCA2"   "CDCA8"   "ECT2"    "KIF23"   "HMMR"   </span>
<span class="co">#&gt; [43] "AURKA"   "PSRC1"   "ANLN"    "LBR"     "CKAP5"   "CENPE"   "CTCF"   </span>
<span class="co">#&gt; [50] "NEK2"    "G2E3"    "GAS2L3"  "CBX5"    "CENPA"</span>

<span class="co"># Classify the cells</span>
pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/CellCycleScoring">CellCycleScoring</a></span>(pbmc,
                         <span class="dt">s.genes =</span> cc.genes<span class="op">$</span>s.genes,
                         <span class="dt">g2m.genes =</span> cc.genes<span class="op">$</span>g2m.genes, 
                         <span class="dt">set.ident =</span> <span class="ot">TRUE</span>)

<span class="co"># view cell cycle scores and phase assignments</span>
<span class="kw">head</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>meta.data)
<span class="co">#&gt;                nGene nUMI orig.ident percent.mito res.0.6 cluster.id</span>
<span class="co">#&gt; AAACATACAACCAC   781 2421   10x_PBMC  0.030177759       0          0</span>
<span class="co">#&gt; AAACATTGAGCTAC  1352 4903   10x_PBMC  0.037935958       2          2</span>
<span class="co">#&gt; AAACATTGATCAGC  1131 3149   10x_PBMC  0.008897363       0          0</span>
<span class="co">#&gt; AAACCGTGCTTCCG   960 2639   10x_PBMC  0.017430845       1          1</span>
<span class="co">#&gt; AAACCGTGTATGCG   522  981   10x_PBMC  0.012244898       5          5</span>
<span class="co">#&gt; AAACGCACTGGTAC   782 2164   10x_PBMC  0.016643551       0          0</span>
<span class="co">#&gt;                     S.Score   G2M.Score Phase       old.ident</span>
<span class="co">#&gt; AAACATACAACCAC  0.088671114 -0.03096637     S     CD4 T cells</span>
<span class="co">#&gt; AAACATTGAGCTAC -0.028912641 -0.04570735    G1         B cells</span>
<span class="co">#&gt; AAACATTGATCAGC -0.007559156  0.07088689   G2M     CD4 T cells</span>
<span class="co">#&gt; AAACCGTGCTTCCG  0.033761944  0.02260974     S CD14+ Monocytes</span>
<span class="co">#&gt; AAACCGTGTATGCG -0.042941375  0.04006888   G2M        NK cells</span>
<span class="co">#&gt; AAACGCACTGGTAC -0.040543204 -0.07054163    G1     CD4 T cells</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Cell cycle stage can then be regressed out using `Scale Data`</span>
pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/ScaleData">ScaleData</a></span>(pbmc,
                  <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>),
                  <span class="dt">display.progress =</span> <span class="ot">FALSE</span>)

<span class="co"># Alternatively, retain cycling vs. stationary information, but regress out differences in phase</span>
pbmc<span class="op">@</span>meta.data<span class="op">$</span>CC.Difference &lt;-<span class="st"> </span>pbmc<span class="op">@</span>meta.data<span class="op">$</span>S.Score <span class="op">-</span><span class="st"> </span>pbmc<span class="op">@</span>meta.data<span class="op">$</span>G2M.Score

pbmc &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Seurat/topics/ScaleData">ScaleData</a></span>(pbmc,
                  <span class="dt">vars.to.regress =</span> <span class="st">"CC.Difference"</span>,
                  <span class="dt">display.progress =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#single-cell-rna-seq-analysis-in-r-with-seurat">Single cell RNA-seq analysis in R with Seurat</a><ul class="nav nav-pills nav-stacked">
<li><a href="#after-updating-eda-a-new-package-named-seurat-should-be-installed">After updating eda, a new package named <code>Seurat</code> should be installed</a></li>
      <li><a href="#single-cell-rna-seq-analysis-in-r-starts-with-the-same-data-as-bulk-rna-seq-a-count-matrix">Single cell RNA-seq analysis in R starts with the same data as bulk RNA-seq: a count matrix</a></li>
      <li><a href="#seurat-uses-a-custom-object-to-store-counts-and-data-similar-to-the-summarizedexperiment-deseqdataset">Seurat uses a custom object to store counts and data (similar to the SummarizedExperiment &amp; DESeqDataSet)</a></li>
      <li><a href="#filtering-and-qc-for-scrna-seq">Filtering and QC for scRNA-seq</a></li>
      <li><a href="#normalizing-the-data-detecting-variable-genes-and-removing-uninteresting-variation">Normalizing the data, detecting variable genes, and removing uninteresting variation</a></li>
      <li><a href="#performing-dimensionality-reduction-and-clustering-pca">Performing dimensionality reduction and clustering: PCA</a></li>
      <li><a href="#performing-dimensionality-reduction-and-clustering-clustering-and-tsne">Performing dimensionality reduction and clustering: Clustering and tSNE</a></li>
      <li><a href="#identifying-marker-genes-differential-expression">Identifying marker genes (differential expression)</a></li>
      <li><a href="#manipulating-seurat-objects-renaming-clusters">Manipulating Seurat objects: renaming clusters</a></li>
      <li><a href="#dealing-with-proliferating-cells">Dealing with proliferating cells</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
