<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 6: RNA-seq Analysis part 1 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Genomic Interval Analysis part 2</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: RNA-Seq Analysis part 1</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li>
      <a href="../articles/exercises-4.html">Exercises 4</a>
    </li>
    <li>
      <a href="../articles/exercises-5.html">Exercises 5</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 6: RNA-seq Analysis part 1</h1>
                        <h4 class="author">Austin Gillen</h4>
            
            <h4 class="date">2017-12-12</h4>
          </div>

    
    
<div class="contents">
<div id="rna-seq-analysis-in-r" class="section level1">
<h1 class="hasAnchor">
<a href="#rna-seq-analysis-in-r" class="anchor"></a>RNA-seq analysis in R</h1>
<ul>
<li>Introduce matrices and matrix manipulation in R</li>
<li>Build a DESeqDataSet and calculate differential gene expression</li>
<li>Plot results and QC metrics</li>
</ul>
<div id="before-we-get-started-installing-bioconductor-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#before-we-get-started-installing-bioconductor-packages" class="anchor"></a>Before we get started: Installing Bioconductor packages</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"DESeq2"</span>)

<span class="co"># Also install two packages from CRAN</span>
<span class="kw">install.packages</span>(<span class="kw">c</span>(<span class="st">"pheatmap"</span>,<span class="st">"RColorBrewer"</span>))

<span class="kw"><a href="../reference/update_eda.html">update_eda</a></span>()

<span class="kw">library</span>(DESeq2)
<span class="kw">library</span>(pheatmap)
<span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(cowplot)
<span class="kw">library</span>(eda)</code></pre></div>
</div>
<div id="matrices-in-r-are-two-dimensional-objects-that-contain-elements-of-one-atomic-type" class="section level2">
<h2 class="hasAnchor">
<a href="#matrices-in-r-are-two-dimensional-objects-that-contain-elements-of-one-atomic-type" class="anchor"></a>Matrices in R are two dimensional objects that contain elements of one atomic type</h2>
<p>Matrices can contain characters or logical values, but we’re really only interested in numeric matrices:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Build 4x3 matrix with row-wise series of integers</span>
M &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>),
            <span class="dt">nrow =</span> <span class="dv">4</span>,
            <span class="dt">byrow =</span> <span class="ot">TRUE</span>)

M
<span class="co">#&gt;      [,1] [,2] [,3]</span>
<span class="co">#&gt; [1,]    1    2    3</span>
<span class="co">#&gt; [2,]    4    5    6</span>
<span class="co">#&gt; [3,]    7    8    9</span>
<span class="co">#&gt; [4,]   10   11   12</span></code></pre></div>
</div>
<div id="elements-of-a-matrix-can-be-accessed-using-column-and-row-indices-or-names" class="section level2">
<h2 class="hasAnchor">
<a href="#elements-of-a-matrix-can-be-accessed-using-column-and-row-indices-or-names" class="anchor"></a>Elements of a matrix can be accessed using column and row indices or names</h2>
<p>To return the value in the first row, second column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">M[<span class="dv">1</span>,<span class="dv">2</span>]
<span class="co">#&gt; [1] 2</span></code></pre></div>
<p>With named rows and columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Assign column names</span>
<span class="kw">colnames</span>(M) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"col1"</span>,
            <span class="st">"col2"</span>,
            <span class="st">"col3"</span>)

<span class="co"># Assign row names</span>
<span class="kw">row.names</span>(M) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"row1"</span>,
            <span class="st">"row2"</span>,
            <span class="st">"row3"</span>,
            <span class="st">"row4"</span>)

M[<span class="st">"row1"</span>,<span class="st">"col2"</span>]
<span class="co">#&gt; [1] 2</span></code></pre></div>
<p>Index ranges and character vectors of row/column names also work:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">M[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>]
<span class="co">#&gt;      col2 col3</span>
<span class="co">#&gt; row1    2    3</span>
<span class="co">#&gt; row2    5    6</span>

M[<span class="kw">c</span>(<span class="st">"row1"</span>,<span class="st">"row2"</span>),<span class="kw">c</span>(<span class="st">"col2"</span>,<span class="st">"col3"</span>)]
<span class="co">#&gt;      col2 col3</span>
<span class="co">#&gt; row1    2    3</span>
<span class="co">#&gt; row2    5    6</span></code></pre></div>
<p>As does negative selection:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Exclude rows 1 and 4</span>
M[<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>),]
<span class="co">#&gt;      col1 col2 col3</span>
<span class="co">#&gt; row2    4    5    6</span>
<span class="co">#&gt; row3    7    8    9</span></code></pre></div>
<p>Full rows and columns can be extracted as well:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># print 1st row</span>
M[<span class="dv">1</span>,]
<span class="co">#&gt; col1 col2 col3 </span>
<span class="co">#&gt;    1    2    3</span>

<span class="co"># print 2nd column</span>
M[,<span class="dv">2</span>]
<span class="co">#&gt; row1 row2 row3 row4 </span>
<span class="co">#&gt;    2    5    8   11</span>

<span class="co"># print 2nd and 3rd rows</span>
M[<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>,]
<span class="co">#&gt;      col1 col2 col3</span>
<span class="co">#&gt; row2    4    5    6</span>
<span class="co">#&gt; row3    7    8    9</span></code></pre></div>
</div>
<div id="mathematical-operations-can-be-performed-on-numeric-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#mathematical-operations-can-be-performed-on-numeric-matrices" class="anchor"></a>Mathematical operations can be performed on numeric matrices</h2>
<p>For example, taking the base 2 logarithm of each value:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">M_log &lt;-<span class="st"> </span><span class="kw">log2</span>(M)

M_log
<span class="co">#&gt;          col1     col2     col3</span>
<span class="co">#&gt; row1 0.000000 1.000000 1.584963</span>
<span class="co">#&gt; row2 2.000000 2.321928 2.584963</span>
<span class="co">#&gt; row3 2.807355 3.000000 3.169925</span>
<span class="co">#&gt; row4 3.321928 3.459432 3.584963</span></code></pre></div>
<p>Or normalizing all columns relative to an arbitrary reference column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">M_norm &lt;-<span class="st"> </span>M<span class="op">/</span>M[,<span class="dv">1</span>]

M_norm
<span class="co">#&gt;      col1     col2     col3</span>
<span class="co">#&gt; row1    1 2.000000 3.000000</span>
<span class="co">#&gt; row2    1 1.250000 1.500000</span>
<span class="co">#&gt; row3    1 1.142857 1.285714</span>
<span class="co">#&gt; row4    1 1.100000 1.200000</span></code></pre></div>
</div>
</div>
<div id="analyzing-rna-seq-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-rna-seq-data" class="anchor"></a>Analyzing RNA-Seq Data</h1>
<div id="simplified-rna-sequencing-analysis-workflow-genome-alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#simplified-rna-sequencing-analysis-workflow-genome-alignment" class="anchor"></a>Simplified RNA sequencing analysis workflow (genome alignment)</h2>
<table class="table">
<colgroup>
<col width="29%">
<col width="22%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Step</th>
<th>Tools</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Align reads to genome</td>
<td>STAR, HISAT2</td>
<td>Align reads to genomic sequence, retaining strand information if present</td>
</tr>
<tr class="even">
<td>Assign reads to features</td>
<td>Subread, StringTie</td>
<td>Assign aligned reads to genes and/or transcripts</td>
</tr>
<tr class="odd">
<td>Calculate differential expression</td>
<td>DESeq2, edgeR, Ballgown</td>
<td>Fit count matrices to model to determine differential expression</td>
</tr>
</tbody>
</table>
</div>
<div id="simplified-rna-sequencing-analysis-workflow-quasi-mapping" class="section level2">
<h2 class="hasAnchor">
<a href="#simplified-rna-sequencing-analysis-workflow-quasi-mapping" class="anchor"></a>Simplified RNA sequencing analysis workflow (quasi-mapping)</h2>
<table class="table">
<colgroup>
<col width="29%">
<col width="22%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Step</th>
<th>Tools</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Estimate transcript abundance</td>
<td>Salmon / Kallisto</td>
<td>Estimate transcript abundance directly on transcriptome sequences</td>
</tr>
<tr class="even">
<td>Calculate differential expression</td>
<td>DESeq2, edgeR, Ballgown</td>
<td>Fit count matrices to model to determine differential expression</td>
</tr>
</tbody>
</table>
</div>
<div id="inspect-rna-seq-count-and-phenotype-data-from-file-that-is-included-in-eda-package" class="section level2">
<h2 class="hasAnchor">
<a href="#inspect-rna-seq-count-and-phenotype-data-from-file-that-is-included-in-eda-package" class="anchor"></a>Inspect RNA-seq count and phenotype data from file that is included in eda package</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># View first few rows of count matrix</span>
<span class="kw">head</span>(drug_resistant_counts)
<span class="co">#&gt;           parent1 parent2 parent3 A_resistant1 A_resistant2 A_resistant3</span>
<span class="co">#&gt; DDX11L1         0       0       0            0            0            0</span>
<span class="co">#&gt; WASH7P          3       7      13            7           12            9</span>
<span class="co">#&gt; MIR6859-1       1       1       1            2            1            2</span>
<span class="co">#&gt; MIR1302-2       0       0       0            0            0            0</span>
<span class="co">#&gt; FAM138A         0       0       0            0            0            0</span>
<span class="co">#&gt; OR4F5           0       0       0            0            0            0</span>
<span class="co">#&gt;           B_resistant1 B_resistant2 B_resistant3 C_resistant1 C_resistant2</span>
<span class="co">#&gt; DDX11L1              0            0            0            0            0</span>
<span class="co">#&gt; WASH7P              10            9           11           24            7</span>
<span class="co">#&gt; MIR6859-1            2            2            2            6            0</span>
<span class="co">#&gt; MIR1302-2            0            0            0            0            0</span>
<span class="co">#&gt; FAM138A              0            0            0            0            0</span>
<span class="co">#&gt; OR4F5                0            0            0            0            0</span>
<span class="co">#&gt;           C_resistant3</span>
<span class="co">#&gt; DDX11L1              0</span>
<span class="co">#&gt; WASH7P               9</span>
<span class="co">#&gt; MIR6859-1            1</span>
<span class="co">#&gt; MIR1302-2            0</span>
<span class="co">#&gt; FAM138A              0</span>
<span class="co">#&gt; OR4F5                0</span>

<span class="co"># View phenotype data</span>
drug_resistant_coldata
<span class="co">#&gt;              drug</span>
<span class="co">#&gt; parent1      none</span>
<span class="co">#&gt; parent2      none</span>
<span class="co">#&gt; parent3      none</span>
<span class="co">#&gt; A_resistant1    A</span>
<span class="co">#&gt; A_resistant2    A</span>
<span class="co">#&gt; A_resistant3    A</span>
<span class="co">#&gt; B_resistant1    B</span>
<span class="co">#&gt; B_resistant2    B</span>
<span class="co">#&gt; B_resistant3    B</span>
<span class="co">#&gt; C_resistant1    C</span>
<span class="co">#&gt; C_resistant2    C</span>
<span class="co">#&gt; C_resistant3    C</span></code></pre></div>
</div>
<div id="deseq2-uses-deseqdatasets-a-subset-of-the-bioconductor-summarizedexperiment-object-to-store-data" class="section level2">
<h2 class="hasAnchor">
<a href="#deseq2-uses-deseqdatasets-a-subset-of-the-bioconductor-summarizedexperiment-object-to-store-data" class="anchor"></a>DESeq2 uses DESeqDataSets, a subset of the Bioconductor SummarizedExperiment object, to store data</h2>
<p>SummarizedExperiment objects provide a structured, self-contained way to store phenotype, gene and assay data, along with analysis intermediates and parameters.</p>
<div class="figure">
<img src="img/class-6-se.png">
</div>
</div>
<div id="build-a-deseqdataset-from-the-drug_resistant-data" class="section level2">
<h2 class="hasAnchor">
<a href="#build-a-deseqdataset-from-the-drug_resistant-data" class="anchor"></a>Build a DESeqDataSet from the drug_resistant data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create DESeqDataSet from drug_resistant data</span>
dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/DESeqDataSet">DESeqDataSetFromMatrix</a></span>(<span class="dt">countData =</span> drug_resistant_counts,
                                         <span class="dt">colData =</span> drug_resistant_coldata,
                                         <span class="dt">design =</span> <span class="op">~</span>drug)
<span class="co"># View DESeqDataSet</span>
dds
<span class="co">#&gt; class: DESeqDataSet </span>
<span class="co">#&gt; dim: 33121 12 </span>
<span class="co">#&gt; metadata(1): version</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(33121): DDX11L1 WASH7P ... VAMP7-2 IL9R-2</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(12): parent1 parent2 ... C_resistant2 C_resistant3</span>
<span class="co">#&gt; colData names(1): drug</span>

<span class="co"># View design</span>
dds<span class="op">@</span>design
<span class="co">#&gt; ~drug</span>

<span class="co"># View colData</span>
<span class="kw">colData</span>(dds)
<span class="co">#&gt; DataFrame with 12 rows and 1 column</span>
<span class="co">#&gt;                  drug</span>
<span class="co">#&gt;              &lt;factor&gt;</span>
<span class="co">#&gt; parent1          none</span>
<span class="co">#&gt; parent2          none</span>
<span class="co">#&gt; parent3          none</span>
<span class="co">#&gt; A_resistant1        A</span>
<span class="co">#&gt; A_resistant2        A</span>
<span class="co">#&gt; ...               ...</span>
<span class="co">#&gt; B_resistant2        B</span>
<span class="co">#&gt; B_resistant3        B</span>
<span class="co">#&gt; C_resistant1        C</span>
<span class="co">#&gt; C_resistant2        C</span>
<span class="co">#&gt; C_resistant3        C</span>

<span class="co"># View counts</span>
<span class="kw">head</span>(<span class="kw">assay</span>(dds))
<span class="co">#&gt;           parent1 parent2 parent3 A_resistant1 A_resistant2 A_resistant3</span>
<span class="co">#&gt; DDX11L1         0       0       0            0            0            0</span>
<span class="co">#&gt; WASH7P          3       7      13            7           12            9</span>
<span class="co">#&gt; MIR6859-1       1       1       1            2            1            2</span>
<span class="co">#&gt; MIR1302-2       0       0       0            0            0            0</span>
<span class="co">#&gt; FAM138A         0       0       0            0            0            0</span>
<span class="co">#&gt; OR4F5           0       0       0            0            0            0</span>
<span class="co">#&gt;           B_resistant1 B_resistant2 B_resistant3 C_resistant1 C_resistant2</span>
<span class="co">#&gt; DDX11L1              0            0            0            0            0</span>
<span class="co">#&gt; WASH7P              10            9           11           24            7</span>
<span class="co">#&gt; MIR6859-1            2            2            2            6            0</span>
<span class="co">#&gt; MIR1302-2            0            0            0            0            0</span>
<span class="co">#&gt; FAM138A              0            0            0            0            0</span>
<span class="co">#&gt; OR4F5                0            0            0            0            0</span>
<span class="co">#&gt;           C_resistant3</span>
<span class="co">#&gt; DDX11L1              0</span>
<span class="co">#&gt; WASH7P               9</span>
<span class="co">#&gt; MIR6859-1            1</span>
<span class="co">#&gt; MIR1302-2            0</span>
<span class="co">#&gt; FAM138A              0</span>
<span class="co">#&gt; OR4F5                0</span></code></pre></div>
</div>
<div id="filtering-genes-with-little-or-no-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-genes-with-little-or-no-expression" class="anchor"></a>Filtering genes with little or no expression</h2>
<p>DESeq2 performs internal filtering to remove genes with low expression, but explicitly filtering extreme cases beforehand can speed up processing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(dds)
<span class="co">#&gt; [1] 33121</span>

<span class="co"># remove genes with 0-1 total reads</span>
dds &lt;-<span class="st"> </span>dds[<span class="kw">rowSums</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/counts">counts</a></span>(dds)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>,]

<span class="kw">nrow</span>(dds)
<span class="co">#&gt; [1] 21178</span></code></pre></div>
</div>
<div id="stabilizing-the-variance-across-the-mean-for-visualizing-count-data" class="section level2">
<h2 class="hasAnchor">
<a href="#stabilizing-the-variance-across-the-mean-for-visualizing-count-data" class="anchor"></a>Stabilizing the variance across the mean for visualizing count data</h2>
<p>Many analysis methods for multidimensional data rely on variance being consistent across mean values. Raw count data has the greatest variance at high mean values (largest raw values), while log2-transformed data has the greatest variance at low mean values (largest fold-changes). To get around this, DESeq2 provides two different functions - <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/vst">vst()</a></code> and <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/rlog">rlog()</a></code> - that stabilize variance across the mean.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># rlog transform</span>
rld &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/rlog">rlog</a></span>(dds,
            <span class="dt">blind =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="sample-distances-are-useful-to-assess-the-overall-similarity-between-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-distances-are-useful-to-assess-the-overall-similarity-between-samples" class="anchor"></a>Sample distances are useful to assess the overall similarity between samples</h2>
<p>Sample distance plots are used to assess overall similarity between samples: which samples are similar to each other, which are different? We calculate the euclidean distance between samples using <code>dist()</code> on the rlog-transformed data to ensure roughly equal contributions from all genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Transpose count matrix and calculate distances using dist()</span>
sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(rld)))

<span class="co"># Convert distance dataframe to matrix</span>
sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists)

<span class="co"># Choose continuous palette from RColorBrewer</span>
colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">rev</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer">brewer.pal</a></span>(<span class="dv">9</span>, <span class="st">"Blues"</span>)))(<span class="dv">255</span>)

<span class="co"># Plot sample distance heatmap with pheatmap</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(sampleDistMatrix,
         <span class="dt">col =</span> colors)</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="principle-component-analysis-pca-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#principle-component-analysis-pca-plots" class="anchor"></a>Principle Component Analysis (PCA) plots</h2>
<p>Another method for visualizing relationships in multidimensional data is principle component analysis. PCA identifies clusters of variables (in this case, genes) that explain the majority of the variance present. PCA plots are x-y plots of the first two principle components; they spread the samples out in the two directions that explain most of the differences.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Generate PCA plot with DESeq2 plotPCA() wrapper</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/plotPCA">plotPCA</a></span>(rld, <span class="dt">intgroup =</span> <span class="st">"drug"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_color_brewer</span>(<span class="dt">palette =</span> <span class="st">"Set1"</span>)</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="calculating-differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#calculating-differential-expression" class="anchor"></a>Calculating differential expression</h2>
<p>DESeq2 provides a wrapper function <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/DESeq">DESeq()</a></code> that takes a DESeqDataSet and calculates differential expression. Progress through analysis steps are printed as the function runs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Run DESeq2</span>
dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/DESeq">DESeq</a></span>(dds)
<span class="co">#&gt; estimating size factors</span>
<span class="co">#&gt; estimating dispersions</span>
<span class="co">#&gt; gene-wise dispersion estimates</span>
<span class="co">#&gt; mean-dispersion relationship</span>
<span class="co">#&gt; final dispersion estimates</span>
<span class="co">#&gt; fitting model and testing</span></code></pre></div>
</div>
<div id="extracting-results-after-differential-expression-is-calculated" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-results-after-differential-expression-is-calculated" class="anchor"></a>Extracting results after differential expression is calculated</h2>
<p>DESeq2 results are extracted with the <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/results">results()</a></code> function. Specific sample-sample comparisons are specified with the <code>contrast</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Extract default results (No Drug vs. Drug A)</span>
res &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/results">results</a></span>(dds)

<span class="co"># View results summary</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/summary">summary</a></span>(res)
<span class="co">#&gt; </span>
<span class="co">#&gt; out of 21178 with nonzero total read count</span>
<span class="co">#&gt; adjusted p-value &lt; 0.1</span>
<span class="co">#&gt; LFC &gt; 0 (up)     : 3165, 15% </span>
<span class="co">#&gt; LFC &lt; 0 (down)   : 3337, 16% </span>
<span class="co">#&gt; outliers [1]     : 25, 0.12% </span>
<span class="co">#&gt; low counts [2]   : 3696, 17% </span>
<span class="co">#&gt; (mean count &lt; 2)</span>
<span class="co">#&gt; [1] see 'cooksCutoff' argument of ?results</span>
<span class="co">#&gt; [2] see 'independentFiltering' argument of ?results</span>

<span class="co"># Extract results for No Drug vs. Drug C</span>
res_none_C &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/results">results</a></span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">"drug"</span>, <span class="st">"none"</span>, <span class="st">"C"</span>))

<span class="co"># Convert dataframe to tibble for use with tidyverse tools</span>
res_tibble &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(res) <span class="op">%&gt;%</span>
<span class="st">  </span>tibble<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/rownames">rownames_to_column</a></span>(<span class="dt">var =</span> <span class="st">"gene"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">as.tibble</span>()</code></pre></div>
</div>
<div id="ma-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#ma-plots" class="anchor"></a>MA-plots</h2>
<p>MA plots - plots of mean expression (x-axis) vs. log2-fold change (y-axis) - are useful for visualizing the distribution of gene expression in your comparisons.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Return results with moderated log2 fold-changes.</span>
res_ma &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/lfcShrink">lfcShrink</a></span>(dds, <span class="dt">contrast=</span><span class="kw">c</span>(<span class="st">"drug"</span>,<span class="st">"none"</span>,<span class="st">"A"</span>), <span class="dt">res=</span>res)

<span class="co"># Plot MA</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/plotMA">plotMA</a></span>(res_ma, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">6</span>,<span class="dv">6</span>))</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="plotting-differential-expression-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-differential-expression-results" class="anchor"></a>Plotting differential expression results</h2>
<p>DESeq2 can plot normalized counts using the <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/plotCounts">plotCounts()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot differential expression</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/plotCounts">plotCounts</a></span>(dds,
           <span class="dt">gene =</span> <span class="st">"ESR1"</span>,
           <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">"drug"</span>))</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
dds_plot &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/plotCounts">plotCounts</a></span>(dds,
           <span class="dt">gene =</span> <span class="st">"ESR1"</span>,
           <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">"drug"</span>),
           <span class="dt">returnData =</span> <span class="ot">TRUE</span>)

<span class="co"># Clean up plot with ggplot2</span>
<span class="kw">ggplot</span>(dds_plot, <span class="kw">aes</span>(<span class="dt">x =</span> drug, <span class="dt">y =</span> count, <span class="dt">color =</span> drug)) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_y_log10</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">cex =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-9-2.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="plotting-heatmaps-of-differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-heatmaps-of-differentially-expressed-genes" class="anchor"></a>Plotting heatmaps of differentially expressed genes</h2>
<p>The pheatmaps package can be used to plot differential gene expression across samples/replicates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract the 20 most significantly differentially expressed genes</span>
top_genes &lt;-<span class="st"> </span>res_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(padj) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">20</span>)

<span class="co"># subset rlog matrix based on gene list</span>
mat  &lt;-<span class="st"> </span><span class="kw">assay</span>(rld)[ top_genes<span class="op">$</span>gene, ]

<span class="co"># normalize expression to per-gene average</span>
mat  &lt;-<span class="st"> </span>mat <span class="op">-</span><span class="st"> </span><span class="kw">rowMeans</span>(mat)

<span class="co"># plot heatmap</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(mat)</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h2>
<p>The exercises for today’s class are presented <a href="https://rnabioco.github.io/eda/articles/exercises-5.html">here</a>, Here is a <a href="https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-5.Rmd">link</a> to the raw text of these exercises</p>
</div>
<div id="quiz" class="section level2">
<h2 class="hasAnchor">
<a href="#quiz" class="anchor"></a>Quiz</h2>
<p>Create an RMarkdown document and answer the questions from the Exercises section. Write your answers in the text, and print the final tibbles/plots that your code produced that gave you the answer. Submit your final document as “Quiz 5” by <strong>Friday at 10 PM</strong>.</p>
<p><strong>Your submitted document must knit to HTML without errors</strong>. I.e., when you click the “Knit” button, the document should build and display and HTML page.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#rna-seq-analysis-in-r">RNA-seq analysis in R</a><ul class="nav nav-pills nav-stacked">
<li><a href="#before-we-get-started-installing-bioconductor-packages">Before we get started: Installing Bioconductor packages</a></li>
      <li><a href="#matrices-in-r-are-two-dimensional-objects-that-contain-elements-of-one-atomic-type">Matrices in R are two dimensional objects that contain elements of one atomic type</a></li>
      <li><a href="#elements-of-a-matrix-can-be-accessed-using-column-and-row-indices-or-names">Elements of a matrix can be accessed using column and row indices or names</a></li>
      <li><a href="#mathematical-operations-can-be-performed-on-numeric-matrices">Mathematical operations can be performed on numeric matrices</a></li>
      </ul>
</li>
      <li>
<a href="#analyzing-rna-seq-data">Analyzing RNA-Seq Data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#simplified-rna-sequencing-analysis-workflow-genome-alignment">Simplified RNA sequencing analysis workflow (genome alignment)</a></li>
      <li><a href="#simplified-rna-sequencing-analysis-workflow-quasi-mapping">Simplified RNA sequencing analysis workflow (quasi-mapping)</a></li>
      <li><a href="#inspect-rna-seq-count-and-phenotype-data-from-file-that-is-included-in-eda-package">Inspect RNA-seq count and phenotype data from file that is included in eda package</a></li>
      <li><a href="#deseq2-uses-deseqdatasets-a-subset-of-the-bioconductor-summarizedexperiment-object-to-store-data">DESeq2 uses DESeqDataSets, a subset of the Bioconductor SummarizedExperiment object, to store data</a></li>
      <li><a href="#build-a-deseqdataset-from-the-drug_resistant-data">Build a DESeqDataSet from the drug_resistant data</a></li>
      <li><a href="#filtering-genes-with-little-or-no-expression">Filtering genes with little or no expression</a></li>
      <li><a href="#stabilizing-the-variance-across-the-mean-for-visualizing-count-data">Stabilizing the variance across the mean for visualizing count data</a></li>
      <li><a href="#sample-distances-are-useful-to-assess-the-overall-similarity-between-samples">Sample distances are useful to assess the overall similarity between samples</a></li>
      <li><a href="#principle-component-analysis-pca-plots">Principle Component Analysis (PCA) plots</a></li>
      <li><a href="#calculating-differential-expression">Calculating differential expression</a></li>
      <li><a href="#extracting-results-after-differential-expression-is-calculated">Extracting results after differential expression is calculated</a></li>
      <li><a href="#ma-plots">MA-plots</a></li>
      <li><a href="#plotting-differential-expression-results">Plotting differential expression results</a></li>
      <li><a href="#plotting-heatmaps-of-differentially-expressed-genes">Plotting heatmaps of differentially expressed genes</a></li>
      <li><a href="#exercises">Exercises</a></li>
      <li><a href="#quiz">Quiz</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
