<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 4: Genomic Interval Analysis part 1 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 4: Genomic Interval Analysis part 1</h1>
                        <h4 class="author">Jay Hesselberth</h4>
            
            <h4 class="date">2017-12-06</h4>
          </div>

    
    
<div class="contents">
<div id="genomic-interval-data" class="section level1">
<h1 class="hasAnchor">
<a href="#genomic-interval-data" class="anchor"></a>Genomic Interval Data</h1>
<ul>
<li>Introduce genome-wide data with the UCSC genome browser and ENCODE chromatin data</li>
<li>Summarizing genomic datatsets as intervals and signals</li>
<li>Intersecting gene models, SNP databases, and Chip-Seq data</li>
</ul>
</div>
<div id="visualizing-genomics-data-with-the-ucsc-genome-browser" class="section level1">
<h1 class="hasAnchor">
<a href="#visualizing-genomics-data-with-the-ucsc-genome-browser" class="anchor"></a>Visualizing genomics data with the UCSC genome browser</h1>
<p>The <a href="http://genome.ucsc.edu/index.html">UCSC genome browser</a> provides a web interface for viewing, downloading, and querying genomics data. Numerous datasets are available to view, including:</p>
<ol style="list-style-type: decimal">
<li>Gene annotations (<a href="http://www.gencodegenes.org/releases/current.html">GENCODE</a>)</li>
<li>Gene expression data (<a href="https://gtexportal.org/home/">GTEx</a>)</li>
<li>SNP databases (<a href="http://www.internationalgenome.org/home">1000 genomes</a>)</li>
<li>Functional genomics data (<a href="https://www.encodeproject.org/">ENCODE</a>)</li>
<li>User-uploaded datasets</li>
</ol>
</div>
<div id="genomics-data-summarizes-to-intervals" class="section level1">
<h1 class="hasAnchor">
<a href="#genomics-data-summarizes-to-intervals" class="anchor"></a>Genomics data summarizes to intervals</h1>
<p>The genome can be described as a coordinate system:</p>
<div class="figure">
<img src="img/class-4-karyotype.png">
</div>
<div id="nucleotides-are-represented-in-1-d-positional-coordinates" class="section level2">
<h2 class="hasAnchor">
<a href="#nucleotides-are-represented-in-1-d-positional-coordinates" class="anchor"></a>Nucleotides are represented in 1-D positional coordinates:</h2>
<pre><code>chr1 ATGCTAGCTGATCGTAGCTG...
     0123456789...
     
chr2 ATGCTAGCTGATCGTAGCTG...
     0123456789...
     
chr3 ATGCTAGCTGATCGTAGCTG...
     0123456789...    </code></pre>
</div>
<div id="intervals-describe-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#intervals-describe-regions" class="anchor"></a>Intervals describe regions</h2>
<pre><code>chr1 ATGCTAGCTGATCGTAGCTG...

# first 20 nucleotides
#chrom  start end
chr1    0     20

# whole chromosomes
chr1  0 249250621
chr2  0 243199373
chr3  0 198022430</code></pre>
</div>
<div id="read-alignments-are-intervals" class="section level2">
<h2 class="hasAnchor">
<a href="#read-alignments-are-intervals" class="anchor"></a>Read alignments are intervals</h2>
<p>After aligning genomics data, each alignment can be represented as an interval covering a particular genomic sequence.</p>
<pre><code>read 1:           ------            
read 2:     -----  
read 3:    ----
      chr1 ATGCTAGCTGATCGTAGCTG...
      
      
chr1  0   4
chr1  1   6
chr1  7   13</code></pre>
</div>
</div>
<div id="bed-format" class="section level1">
<h1 class="hasAnchor">
<a href="#bed-format" class="anchor"></a>BED format</h1>
<p>The <a href="https://genome.ucsc.edu/FAQ/FAQformat#format1"><code>BED</code></a> file format defines genomic regions of interest:</p>
<pre><code>#BED3 (3-columns)

#chrom start end
chr1   0     4
chr1   1     6
chr1   7     13</code></pre>
<p>and can define genomic features of interest:</p>
<ul>
<li>read alignments</li>
<li>gene coordinates</li>
<li>positions of genetic variation (i.e. SNPs, indels)</li>
</ul>
<pre><code>#BED6
#chrom start end name  score strand

chr1  0  4  ELAVL1_binding_site 100.00  +
chr1  1  6  microRNA            10      -
chr1  7  13 alternative_exon    1       .

interval 1:       -----&gt;            
interval 2:  &lt;----  
interval 3:&lt;--&gt;
      chr1:ATGCTAGCTGATCGTAGCTG...
</code></pre>
</div>
<div id="obtaining-interval-data" class="section level1">
<h1 class="hasAnchor">
<a href="#obtaining-interval-data" class="anchor"></a>Obtaining interval data</h1>
<p><code>BED</code> files can be downloaded from many sources:</p>
<ol style="list-style-type: decimal">
<li><p>The <a href="https://genome.ucsc.edu/cgi-bin/hgTables">UCSC table browser</a> provides an easy interface for downloading many types of genomic data as <code>BED</code> (and other) format files.</p></li>
<li><p>The ENCODE project provides a data access portal to download experimental data from numerous types of high-throughput experiments (<code>BED</code> files <a href="https://www.encodeproject.org/matrix/?type=Experiment&amp;files.file_type=bed+narrowPeak">here</a>).</p></li>
</ol>
<div id="bedgraph-files-define-genomic-signals" class="section level2">
<h2 class="hasAnchor">
<a href="#bedgraph-files-define-genomic-signals" class="anchor"></a>Bedgraph files define genomic signals</h2>
<p>The <a href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html"><code>bedGraph</code></a> format is an extension of the <code>BED</code> format commonly used to summarize single nucleotide read coverage from <code>ChIP-Seq</code>, <code>CLIP-Seq</code>, <code>mRNA-Seq</code>, and other genome-wide data.</p>
<pre><code>read 1:           ------            
read 2:     -----  
read 3:    ----
      chr1 ATGCTAGCTGATCGTAGCTG...
Coverage   12221101111110000000

chrom start end value
chr1  0  1  1
chr1  1  4  2
chr1  4  6  1
chr1  7 13  1</code></pre>
</div>
</div>
<div id="exploring-chip-seq-data" class="section level1">
<h1 class="hasAnchor">
<a href="#exploring-chip-seq-data" class="anchor"></a>Exploring ChIP-Seq Data</h1>
<div id="simplified-sequencing-analysis-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#simplified-sequencing-analysis-workflow" class="anchor"></a>Simplified sequencing analysis workflow</h2>
<table style="width:88%;" class="table">
<colgroup>
<col width="22%">
<col width="22%">
<col width="43%">
</colgroup>
<thead><tr class="header">
<th>Step</th>
<th>File Format</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Get read sequences</td>
<td><a href="https://en.wikipedia.org/wiki/FASTQ_format"><code>FASTQ</code></a></td>
<td>Sequencing Reads</td>
</tr>
<tr class="even">
<td>Align to genome</td>
<td><a href="http://samtools.github.io/hts-specs/SAMv1.pdf"><code>BAM/SAM</code></a></td>
<td>
<strong>S</strong>equence <strong>A</strong>lignment <strong>M</strong>ap Alignment information for each read (BAM = binary file)</td>
</tr>
<tr class="odd">
<td>Summarize Coverage</td>
<td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1.8"><code>bedGraph/bigWig</code></a></td>
<td>Count number of reads covering every position in genome i.e. read coverage (bigWig = binary file)</td>
</tr>
<tr class="even">
<td>Define enriched regions</td>
<td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1"><code>BED</code></a></td>
<td>Find regions with enriched coverage. i.e.(“Peaks” of signal)</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="img/class-4-alignments.jpg">
</div>
</div>
<div id="chip-seq-peaks-in-human-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#chip-seq-peaks-in-human-cells" class="anchor"></a>ChIP-Seq Peaks in Human Cells</h2>
<p>The <a href="https://www.encodeproject.org"><code>ENCODE</code></a> project produced hundreds of <a href="https://en.wikipedia.org/wiki/ChIP-sequencing">ChIP-Seq</a> datasets across dozens of humans cell lines. A compendium of ChIP-Seq experiments from 161 DNA-binding proteins were aggregated into a single BED file, available from the UCSC genome browser. We will use this dataset (see <a href="http://genome.ucsc.edu/cgi-bin/hgTables?db=hg19&amp;hgta_group=regulation&amp;hgta_track=wgEncodeRegTfbsClusteredV3&amp;hgta_table=wgEncodeRegTfbsClusteredV3&amp;hgta_doSchema=describe+table+schema">description</a>) to explore interval analysis methods.</p>
<p>First we will discuss a few methods for loading the data into R.</p>
</div>
<div id="download-the-data-to-your-computer-using-r" class="section level2">
<h2 class="hasAnchor">
<a href="#download-the-data-to-your-computer-using-r" class="anchor"></a>Download the data to your computer using R</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
url &lt;-<span class="st"> "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz"</span>

<span class="kw">download.file</span>(url, <span class="st">"wgEncodeRegTfbsClusteredV3.bed.gz"</span>)

tf_bed &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">"wgEncodeRegTfbsClusteredV3.bed.gz"</span>, 
                   <span class="dt">col_names =</span> F)</code></pre></div>
</div>
<div id="load-the-data-into-memory-directly-from-url-link" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-data-into-memory-directly-from-url-link" class="anchor"></a>Load the data into memory directly from url link</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
url &lt;-<span class="st"> "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz"</span>

tf_bed &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(url, <span class="dt">col_names =</span> F)</code></pre></div>
</div>
<div id="load-data-from-file-that-is-included-in-eda-package" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-from-file-that-is-included-in-eda-package" class="anchor"></a>Load data from file that is included in eda package</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(eda)

file_name &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, 
                         <span class="st">"wgEncodeRegTfbsClusteredV3.bed.gz"</span>, 
                         <span class="dt">package =</span> <span class="st">"eda"</span>)

tf_bed &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(file_name, <span class="dt">col_names =</span> F)

tf_bed
<span class="co">#&gt; # A tibble: 4,380,444 x 4</span>
<span class="co">#&gt;       X1     X2     X3     X4</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  10073  10329 ZBTB33</span>
<span class="co">#&gt;  2  chr1  10149  10413  CEBPB</span>
<span class="co">#&gt;  3  chr1  16110  16390   CTCF</span>
<span class="co">#&gt;  4  chr1  29198  29688   TAF1</span>
<span class="co">#&gt;  5  chr1  29275  29591  GABPA</span>
<span class="co">#&gt;  6  chr1  89795  90051   USF1</span>
<span class="co">#&gt;  7  chr1  91156  91580   CTCF</span>
<span class="co">#&gt;  8  chr1 104859 105089   CTCF</span>
<span class="co">#&gt;  9  chr1 138850 139274   CTCF</span>
<span class="co">#&gt; 10  chr1 235541 235877    SP1</span>
<span class="co">#&gt; # ... with 4,380,434 more rows</span></code></pre></div>
</div>
<div id="fixing-column-names" class="section level2">
<h2 class="hasAnchor">
<a href="#fixing-column-names" class="anchor"></a>Fixing column names</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_names &lt;-<span class="st"> </span><span class="kw">c</span>(
  <span class="st">"chrom"</span>,
  <span class="st">"start"</span>,
  <span class="st">"end"</span>,
  <span class="st">"TF"</span>
)

<span class="kw">colnames</span>(tf_bed) &lt;-<span class="st"> </span>new_names
tf_bed
<span class="co">#&gt; # A tibble: 4,380,444 x 4</span>
<span class="co">#&gt;    chrom  start    end     TF</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  10073  10329 ZBTB33</span>
<span class="co">#&gt;  2  chr1  10149  10413  CEBPB</span>
<span class="co">#&gt;  3  chr1  16110  16390   CTCF</span>
<span class="co">#&gt;  4  chr1  29198  29688   TAF1</span>
<span class="co">#&gt;  5  chr1  29275  29591  GABPA</span>
<span class="co">#&gt;  6  chr1  89795  90051   USF1</span>
<span class="co">#&gt;  7  chr1  91156  91580   CTCF</span>
<span class="co">#&gt;  8  chr1 104859 105089   CTCF</span>
<span class="co">#&gt;  9  chr1 138850 139274   CTCF</span>
<span class="co">#&gt; 10  chr1 235541 235877    SP1</span>
<span class="co">#&gt; # ... with 4,380,434 more rows</span></code></pre></div>
</div>
</div>
<div id="in-class-exercises" class="section level1">
<h1 class="hasAnchor">
<a href="#in-class-exercises" class="anchor"></a>In class exercises</h1>
<ol style="list-style-type: decimal">
<li><p>Which TF has the most binding-sites (AKA peaks)?</p></li>
<li><p>Which TF covers the most basepairs in the genome?</p></li>
<li><p>Which TF has the broadest or the narrowest binding-sites on average?</p></li>
</ol>
</div>
<div id="integrating-multiple-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#integrating-multiple-datasets" class="anchor"></a>Integrating multiple datasets</h1>
<p>We can now use the BED format as a common file to investigate multiple genome-wide datatypes. Imagine that we had two bed files named (<code>x</code> and <code>y</code>). How would we go about finding intervals in <code>x</code> that overlap with <code>y</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x
<span class="co">#&gt; # A tibble: 3 x 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    25    50</span>
<span class="co">#&gt; 2  chr1    40    80</span>
<span class="co">#&gt; 3  chr1   100   125</span>
y
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    30    75</span></code></pre></div>
<p>This is not very easy to do with dplyr or base R, and would be very slow.</p>
</div>
<div id="how-to-efficiently-analyze-intervals-and-signals" class="section level1">
<h1 class="hasAnchor">
<a href="#how-to-efficiently-analyze-intervals-and-signals" class="anchor"></a>How to efficiently analyze intervals and signals?</h1>
<p>To efficiently query interval data a set of algorithms are used to speed up the querying process. Examples here: <a href="https://en.wikipedia.org/wiki/Interval_tree">Interval Tree</a> or <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0973-5">ChromSweep</a>.</p>
<p>These algorithms have been implemented in a set of tools specific for analyzing genomic interval data.</p>
<div id="command-line-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#command-line-tools" class="anchor"></a>Command-line tools</h2>
<ul>
<li>
<a href="http://bedtools.readthedocs.io/en/latest/"><code>BEDTools</code></a> (the most popular and original tool-suite)</li>
<li>
<a href="https://bedops.readthedocs.io/en/latest/"><code>BedOPS</code></a> (similar to BEDTools but offers some unique functionality)</li>
</ul>
</div>
<div id="r-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#r-packages" class="anchor"></a>R-Packages</h2>
<ul>
<li>
<a href="https://rnabioco.github.io/valr/"><code>valr</code></a> (Interval analysis with a <code>tidyverse</code> style syntax)</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html"><code>GRanges</code></a> (Interval analysis used in many <a href="http://www.bioconductor.org/"><code>BioConductor</code></a> packages)</li>
</ul>
</div>
</div>
<div id="introduction-to-valr" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-valr" class="anchor"></a>Introduction to valr</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#install.packages("valr")</span>
<span class="kw">library</span>(valr)</code></pre></div>
<p><a href="https://rnabioco.github.io/valr/"><code>valr</code></a> was developed to simplify genomic interval analysis and replace the need for learning the command line. We will use <code>valr</code> today to introduce genomic interval analysis.</p>
<div id="reading-in-bed-data" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-bed-data" class="anchor"></a>Reading in BED data</h2>
<p><code>valr</code> assigns common column names to facilitate comparisons between tibbles. All tibbles will have <code>chrom</code>, <code>start</code>, and <code>end</code> columns, and some tibbles from multi-column formats will have additional pre-determined column names. See the <a href="https://rnabioco.github.io/valr/reference/read_bed.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed()</a></code></a> documentation for additional details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># get path to bed file</span>
bed_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>,
                        <span class="st">"6fields.bed.gz"</span>,
                        <span class="dt">package =</span> <span class="st">"valr"</span>)

<span class="co"># read in bed, indicate the number of columns with n_fields</span>
bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(bed_file, <span class="dt">n_fields =</span> <span class="dv">6</span>)

bed
<span class="co">#&gt; # A tibble: 10 x 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  11873  14409      DDX11L1     3      +</span>
<span class="co">#&gt;  2  chr1  14361  19759       WASH7P    10      -</span>
<span class="co">#&gt;  3  chr1  14406  29370       WASH7P     7      -</span>
<span class="co">#&gt;  4  chr1  34610  36081      FAM138F     3      -</span>
<span class="co">#&gt;  5  chr1  69090  70008        OR4F5     1      +</span>
<span class="co">#&gt;  6  chr1 134772 140566    LOC729737     3      -</span>
<span class="co">#&gt;  7  chr1 321083 321115     DQ597235     1      +</span>
<span class="co">#&gt;  8  chr1 321145 321207     DQ599768     1      +</span>
<span class="co">#&gt;  9  chr1 322036 326938 LOC100133331     3      +</span>
<span class="co">#&gt; 10  chr1 327545 328439    LOC388312     1      +</span></code></pre></div>
</div>
<div id="find-intersecting-intervals-with-bed_intersect" class="section level2">
<h2 class="hasAnchor">
<a href="#find-intersecting-intervals-with-bed_intersect" class="anchor"></a>Find intersecting intervals with bed_intersect()</h2>
<p>One of the most common operations in interval analysis is to determine if intervals in 1 BED file overlap with intervals in another file. This operation can be used to answer many questions including:</p>
<ol style="list-style-type: decimal">
<li>Does a SNP overlap with a protein-coding region?</li>
<li>Which genes does a transcription factor bind to?</li>
<li>Does an RNA binding protein bind to 3’UTRs?</li>
</ol>
<p>The <a href="https://rnabioco.github.io/valr/reference/bed_intersect.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect()</a></code></a> function in valr finds intervals that overlap in two datasets and reports the intersecting intervals.</p>
<p>Imagine that we have a set of two bed files that we have read into R as <code>x</code> and <code>y</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x
<span class="co">#&gt; # A tibble: 3 x 6</span>
<span class="co">#&gt;   chrom start   end  name score strand</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1    25    50     A     0      +</span>
<span class="co">#&gt; 2  chr1    40    80     B     0      -</span>
<span class="co">#&gt; 3  chr1   100   125     C     0      -</span>
y
<span class="co">#&gt; # A tibble: 1 x 6</span>
<span class="co">#&gt;   chrom start   end  name score strand</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1    30    75     C     0      -</span></code></pre></div>
<p>We can find intersecting intervals by passing <code>x</code> ans <code>y</code> to <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect()</a></code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 2 x 12</span>
<span class="co">#&gt;   chrom start.x end.x name.x score.x strand.x start.y end.y name.y score.y</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1      25    50      A       0        +      30    75      C       0</span>
<span class="co">#&gt; 2  chr1      40    80      B       0        -      30    75      C       0</span>
<span class="co">#&gt; # ... with 2 more variables: strand.y &lt;chr&gt;, .overlap &lt;int&gt;</span>

<span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect</a></span>(x, y, <span class="dt">suffix =</span> <span class="kw">c</span>(<span class="st">"_xfile"</span>, <span class="st">"_yfile"</span>))
<span class="co">#&gt; # A tibble: 2 x 12</span>
<span class="co">#&gt;   chrom start_xfile end_xfile name_xfile score_xfile strand_xfile</span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;      &lt;chr&gt;       &lt;dbl&gt;        &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1          25        50          A           0            +</span>
<span class="co">#&gt; 2  chr1          40        80          B           0            -</span>
<span class="co">#&gt; # ... with 6 more variables: start_yfile &lt;dbl&gt;, end_yfile &lt;dbl&gt;,</span>
<span class="co">#&gt; #   name_yfile &lt;chr&gt;, score_yfile &lt;dbl&gt;, strand_yfile &lt;chr&gt;,</span>
<span class="co">#&gt; #   .overlap &lt;int&gt;</span>

x <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect</a></span>(y)
<span class="co">#&gt; # A tibble: 2 x 12</span>
<span class="co">#&gt;   chrom start.x end.x name.x score.x strand.x start.y end.y name.y score.y</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1      25    50      A       0        +      30    75      C       0</span>
<span class="co">#&gt; 2  chr1      40    80      B       0        -      30    75      C       0</span>
<span class="co">#&gt; # ... with 2 more variables: strand.y &lt;chr&gt;, .overlap &lt;int&gt;</span></code></pre></div>
<p><code>valr</code> functions respect groups established by <code>group_by()</code>. For example to only intersect intervals if they are on the same strand, <code>group_by()</code> strand prior to intersecting.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">group_by</span>(x, strand)
y &lt;-<span class="st"> </span><span class="kw">group_by</span>(y, strand)
<span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 1 x 12</span>
<span class="co">#&gt;   chrom start.x end.x name.x score.x strand.x start.y end.y name.y score.y</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;chr&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1      40    80      B       0        -      30    75      C       0</span>
<span class="co">#&gt; # ... with 2 more variables: strand.y &lt;chr&gt;, .overlap &lt;int&gt;</span></code></pre></div>
<p><code>valr</code> includes many functions for manipulating, comparing, and randomizing interval data. The <a href="https://rnabioco.github.io/valr/reference/index.html">documentation</a> contains a list of all of the available functions.</p>
</div>
<div id="how-many-snps-in-tf-binding-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#how-many-snps-in-tf-binding-sites" class="anchor"></a>How many SNPs in TF binding sites?</h2>
<p>First obtain a bed file of SNPs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
snp_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, 
                        <span class="st">"hg19.snps147.chr22.bed.gz"</span>, 
                        <span class="dt">package =</span> <span class="st">"valr"</span>)

snps &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(snp_file, 
                 <span class="dt">n_fields =</span> <span class="dv">6</span>)
snps
<span class="co">#&gt; # A tibble: 10,000 x 6</span>
<span class="co">#&gt;    chrom    start      end        name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;       &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 16053247 16053248 rs587721086     0      +</span>
<span class="co">#&gt;  2 chr22 16053443 16053444  rs80167676     0      +</span>
<span class="co">#&gt;  3 chr22 16055964 16055965 rs587706951     0      +</span>
<span class="co">#&gt;  4 chr22 16069373 16069374   rs2154787     0      +</span>
<span class="co">#&gt;  5 chr22 16069782 16069783   rs1963212     0      +</span>
<span class="co">#&gt;  6 chr22 16100513 16100514   rs8140563     0      +</span>
<span class="co">#&gt;  7 chr22 16114622 16114623 rs558374111     0      +</span>
<span class="co">#&gt;  8 chr22 16115181 16115182 rs567521927     0      +</span>
<span class="co">#&gt;  9 chr22 16115377 16115378 rs538733846     0      +</span>
<span class="co">#&gt; 10 chr22 16121639 16121640 rs544560983     0      +</span>
<span class="co">#&gt; # ... with 9,990 more rows</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snps_in_tfs &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect</a></span>(tf_bed, snps) 

snps_in_tfs
<span class="co">#&gt; # A tibble: 8,851 x 10</span>
<span class="co">#&gt;    chrom  start.x    end.x  TF.x  start.y    end.y      name.y score.y</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;       &lt;chr&gt;   &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 16872262 16872686  CTCF 16872280 16872281    rs131546       0</span>
<span class="co">#&gt;  2 chr22 16921734 16922262  CTCF 16921912 16921913 rs575577088       0</span>
<span class="co">#&gt;  3 chr22 16921781 16922105 CEBPB 16921912 16921913 rs575577088       0</span>
<span class="co">#&gt;  4 chr22 16921808 16922078  SMC3 16921912 16921913 rs575577088       0</span>
<span class="co">#&gt;  5 chr22 17024669 17024969 NR3C1 17024865 17024866   rs2379966       0</span>
<span class="co">#&gt;  6 chr22 17044937 17045357  TAF1 17045002 17045003 rs536197102       0</span>
<span class="co">#&gt;  7 chr22 17083377 17084311 RBBP5 17083589 17083590   rs2236640       0</span>
<span class="co">#&gt;  8 chr22 17083465 17083650   MAX 17083589 17083590   rs2236640       0</span>
<span class="co">#&gt;  9 chr22 17083520 17084322  UBTF 17083589 17083590   rs2236640       0</span>
<span class="co">#&gt; 10 chr22 17083536 17084046 SAP30 17083589 17083590   rs2236640       0</span>
<span class="co">#&gt; # ... with 8,841 more rows, and 2 more variables: strand.y &lt;chr&gt;,</span>
<span class="co">#&gt; #   .overlap &lt;int&gt;</span></code></pre></div>
</div>
<div id="how-close-are-snps-to-transcription-start-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#how-close-are-snps-to-transcription-start-sites" class="anchor"></a>How close are SNPs to transcription start sites?</h2>
<p>The <a href="https://rnabioco.github.io/valr/reference/bed_intersect.html"><code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_closest">bed_closest()</a></code></a> function finds the closest intervals between two sets of intervals.</p>
<p>First read in a BED file that contains genes in the human genome (hg19 build).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, 
                        <span class="st">"hg19_genes.bed.gz"</span>, 
                        <span class="dt">package =</span> <span class="st">"eda"</span>)

genes &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(gene_file, 
                 <span class="dt">n_fields =</span> <span class="dv">6</span>)
genes
<span class="co">#&gt; # A tibble: 55,006 x 6</span>
<span class="co">#&gt;    chrom start   end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1 11868 14362 LOC102725121     3      +</span>
<span class="co">#&gt;  2  chr1 11873 14409      DDX11L1     3      +</span>
<span class="co">#&gt;  3  chr1 14361 29370       WASH7P    11      -</span>
<span class="co">#&gt;  4  chr1 17368 17436    MIR6859-1     1      -</span>
<span class="co">#&gt;  5  chr1 17368 17436    MIR6859-2     1      -</span>
<span class="co">#&gt;  6  chr1 17368 17436    MIR6859-3     1      -</span>
<span class="co">#&gt;  7  chr1 17368 17436    MIR6859-4     1      -</span>
<span class="co">#&gt;  8  chr1 30365 30503    MIR1302-2     1      +</span>
<span class="co">#&gt;  9  chr1 30365 30503    MIR1302-9     1      +</span>
<span class="co">#&gt; 10  chr1 30365 30503   MIR1302-10     1      +</span>
<span class="co">#&gt; # ... with 54,996 more rows</span></code></pre></div>
<p>Next we’ll generate a set of transcription start sites for each gene. The start site depends on the strand of the interval. Here for simplicity we will only look at intervals from the plus strand. If you want to keep the minus strand you need to use <code>ifelse()</code> in the <code>mutate()</code> call to conditionally define the start and end based on strand.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss &lt;-<span class="st"> </span><span class="kw">filter</span>(genes, strand <span class="op">==</span><span class="st"> "+"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
              
tss
<span class="co">#&gt; # A tibble: 28,035 x 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1  chr1  11868  11869 LOC102725121     3      +</span>
<span class="co">#&gt;  2  chr1  11873  11874      DDX11L1     3      +</span>
<span class="co">#&gt;  3  chr1  30365  30366    MIR1302-2     1      +</span>
<span class="co">#&gt;  4  chr1  30365  30366    MIR1302-9     1      +</span>
<span class="co">#&gt;  5  chr1  30365  30366   MIR1302-10     1      +</span>
<span class="co">#&gt;  6  chr1  30365  30366   MIR1302-11     1      +</span>
<span class="co">#&gt;  7  chr1  69090  69091        OR4F5     1      +</span>
<span class="co">#&gt;  8  chr1 323891 323892 LOC100132287     3      +</span>
<span class="co">#&gt;  9  chr1 323891 323892 LOC100132062     3      +</span>
<span class="co">#&gt; 10  chr1 323891 323892 LOC100133331     4      +</span>
<span class="co">#&gt; # ... with 28,025 more rows</span></code></pre></div>
<p>Lastly, we will use the <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_closest">bed_closest()</a></code> tool to find the closest <code>TSS</code> to each <code>SNP</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_closest">bed_closest</a></span>(snps, tss) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(chrom<span class="op">:</span>strand.x, name.y, .overlap, .dist)
<span class="co">#&gt; # A tibble: 14,019 x 9</span>
<span class="co">#&gt;    chrom  start.x    end.x      name.x score.x strand.x  name.y .overlap</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;       &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;    &lt;int&gt;</span>
<span class="co">#&gt;  1 chr22 16053247 16053248 rs587721086       0        + BMS1P22        0</span>
<span class="co">#&gt;  2 chr22 16053443 16053444  rs80167676       0        + BMS1P22        0</span>
<span class="co">#&gt;  3 chr22 16055964 16055965 rs587706951       0        + BMS1P22        0</span>
<span class="co">#&gt;  4 chr22 16069373 16069374   rs2154787       0        + BMS1P22        0</span>
<span class="co">#&gt;  5 chr22 16069782 16069783   rs1963212       0        + BMS1P22        0</span>
<span class="co">#&gt;  6 chr22 16100513 16100514   rs8140563       0        + BMS1P22        0</span>
<span class="co">#&gt;  7 chr22 16114622 16114623 rs558374111       0        + BMS1P22        0</span>
<span class="co">#&gt;  8 chr22 16115181 16115182 rs567521927       0        + BMS1P22        0</span>
<span class="co">#&gt;  9 chr22 16115377 16115378 rs538733846       0        + BMS1P22        0</span>
<span class="co">#&gt; 10 chr22 16121639 16121640 rs544560983       0        + BMS1P22        0</span>
<span class="co">#&gt; # ... with 14,009 more rows, and 1 more variables: .dist &lt;int&gt;</span>

<span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_closest">bed_closest</a></span>(snps, tss) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">abs</span>(.dist) <span class="op">&lt;</span><span class="st"> </span><span class="dv">50</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(chrom<span class="op">:</span>strand.x, name.y, .overlap, .dist)
<span class="co">#&gt; # A tibble: 14 x 9</span>
<span class="co">#&gt;    chrom  start.x    end.x      name.x score.x strand.x       name.y</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;       &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;        &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 21043817 21043818 rs373084233       0        +    POM121L4P</span>
<span class="co">#&gt;  2 chr22 21319416 21319417 rs112080977       0        +        AIFM3</span>
<span class="co">#&gt;  3 chr22 21319416 21319417 rs112080977       0        +        AIFM3</span>
<span class="co">#&gt;  4 chr22 23229961 23229962   rs6003365       0        +        IGLL5</span>
<span class="co">#&gt;  5 chr22 23229961 23229962   rs6003365       0        +        IGLL5</span>
<span class="co">#&gt;  6 chr22 25423930 25423931  rs11912058       0        +     KIAA1671</span>
<span class="co">#&gt;  7 chr22 26908525 26908525 rs147381057       0        + LOC100507599</span>
<span class="co">#&gt;  8 chr22 31090785 31090786  rs79717894       0        +        OSBP2</span>
<span class="co">#&gt;  9 chr22 38005057 38005058  rs13053965       0        +         GGA1</span>
<span class="co">#&gt; 10 chr22 38142198 38142199 rs117955278       0        +       TRIOBP</span>
<span class="co">#&gt; 11 chr22 38142198 38142199 rs117955278       0        +       TRIOBP</span>
<span class="co">#&gt; 12 chr22 38203871 38203872   rs5995499       0        +         GCAT</span>
<span class="co">#&gt; 13 chr22 38203871 38203872   rs5995499       0        +         GCAT</span>
<span class="co">#&gt; 14 chr22 46156445 46156446  rs60308683       0        +      MIR4762</span>
<span class="co">#&gt; # ... with 2 more variables: .overlap &lt;int&gt;, .dist &lt;int&gt;</span></code></pre></div>
</div>
</div>
<div id="exercises" class="section level1">
<h1 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h1>
<ol style="list-style-type: decimal">
<li><p>The Encode project has performed <code>eCLIP</code> experiments to map RNA binding protein (RBP) binding sites for &gt; 100 RBPs. Go to the <a href="https://www.encodeproject.org/matrix/?type=Experiment">Encode website</a>, search for <code>eCLIP</code> experiments, select an RBP eClip experiment, and load a <code>BED</code> file (bed narrowPeak format) with the binding sites into R using just the url. Use the <a href="https://rnabioco.github.io/valr/reference/read_bed.html"><code>read_narrowPeak()</code></a> function from <code>valr</code>. Make sure that the file was generated using the <code>hg19</code> genome build, and that you do <strong>not</strong> select the <code>eCLIP mock input</code> samples, which are negative controls. Display the bed tibble that you pulled from the Encode database and give a one to two sentence summary about the known functions of the RBP.</p></li>
<li><p>Use <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect()</a></code> and some dplyr functions to determine how many <code>eCLIP</code> peaks are found in genes and how many are intergenic. (see the documentation on the <code>invert</code> argument for <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_intersect">bed_intersect()</a></code>). Use the <code>hg19_genes.bed.gz</code> bed file as your gene bed file (location shown below). Explain in the interpretation whether or not the RBP primarily binds to known (geneic) or unannotated (intergeneic) regions.</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genes.bed.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
<span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(genes, <span class="dt">n_fields =</span> <span class="dv">6</span>)</code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><p>Parse the output from question 2 to determine which gene has the most binding sites for your <code>RBP</code>.</p></li>
<li><p>Use another function from <code>valr</code> (overview and documentation <a href="https://rnabioco.github.io/valr/articles/valr.html">here</a> to study a relationship between your eClip binding sites and genes in the <code>gene</code> bed file. Explain your question and answer. Some <code>valr</code> functions require a <code>genome</code> file, which indicates the lengths of the chromosomes.You can access this file as shown below.</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genome.txt.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_genome">read_genome</a></span>(genome_file)

genome</code></pre></div>
</div>
<div id="quiz" class="section level1">
<h1 class="hasAnchor">
<a href="#quiz" class="anchor"></a>Quiz</h1>
<p>Create an RMarkdown document and answer the questions from the Exercises section. Write your answers in the text, and print the final tibbles that your code produced that gave you the answer. Submit your final document as “Quiz 3” by Friday at 10 PM.</p>
<p><strong>Your submitted document must knit to HTML without errors</strong>. I.e., when you click the “Knit” button, the document should build and display and HTML page.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#genomic-interval-data">Genomic Interval Data</a></li>
      <li><a href="#visualizing-genomics-data-with-the-ucsc-genome-browser">Visualizing genomics data with the UCSC genome browser</a></li>
      <li>
<a href="#genomics-data-summarizes-to-intervals">Genomics data summarizes to intervals</a><ul class="nav nav-pills nav-stacked">
<li><a href="#nucleotides-are-represented-in-1-d-positional-coordinates">Nucleotides are represented in 1-D positional coordinates:</a></li>
      <li><a href="#intervals-describe-regions">Intervals describe regions</a></li>
      <li><a href="#read-alignments-are-intervals">Read alignments are intervals</a></li>
      </ul>
</li>
      <li><a href="#bed-format">BED format</a></li>
      <li>
<a href="#obtaining-interval-data">Obtaining interval data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#bedgraph-files-define-genomic-signals">Bedgraph files define genomic signals</a></li>
      </ul>
</li>
      <li>
<a href="#exploring-chip-seq-data">Exploring ChIP-Seq Data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#simplified-sequencing-analysis-workflow">Simplified sequencing analysis workflow</a></li>
      <li><a href="#chip-seq-peaks-in-human-cells">ChIP-Seq Peaks in Human Cells</a></li>
      <li><a href="#download-the-data-to-your-computer-using-r">Download the data to your computer using R</a></li>
      <li><a href="#load-the-data-into-memory-directly-from-url-link">Load the data into memory directly from url link</a></li>
      <li><a href="#load-data-from-file-that-is-included-in-eda-package">Load data from file that is included in eda package</a></li>
      <li><a href="#fixing-column-names">Fixing column names</a></li>
      </ul>
</li>
      <li><a href="#in-class-exercises">In class exercises</a></li>
      <li><a href="#integrating-multiple-datasets">Integrating multiple datasets</a></li>
      <li>
<a href="#how-to-efficiently-analyze-intervals-and-signals">How to efficiently analyze intervals and signals?</a><ul class="nav nav-pills nav-stacked">
<li><a href="#command-line-tools">Command-line tools</a></li>
      <li><a href="#r-packages">R-Packages</a></li>
      </ul>
</li>
      <li>
<a href="#introduction-to-valr">Introduction to valr</a><ul class="nav nav-pills nav-stacked">
<li><a href="#reading-in-bed-data">Reading in BED data</a></li>
      <li><a href="#find-intersecting-intervals-with-bed_intersect">Find intersecting intervals with bed_intersect()</a></li>
      <li><a href="#how-many-snps-in-tf-binding-sites">How many SNPs in TF binding sites?</a></li>
      <li><a href="#how-close-are-snps-to-transcription-start-sites">How close are SNPs to transcription start sites?</a></li>
      </ul>
</li>
      <li><a href="#exercises">Exercises</a></li>
      <li><a href="#quiz">Quiz</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
