<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 1: Overview and data wrangling • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Wrangling</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Gene Expression Vignette</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Chromatin Vignette</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: Single-Cell Analysis</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Submitting Assignments</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/eda">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 1: Overview and data wrangling</h1>
                        <h4 class="author">Jay Hesselberth</h4>
            
            <h4 class="date">2017-11-11</h4>
          </div>

    
    
<div class="contents">
<section id="goals-for-the-class" class="level1"><h1>Goals for the class</h1>
<ul>
<li><p><em>Learn the basics of data import, processing, plotting, and report generation.</em></p></li>
<li><p><em>Reach for RStudio instead of MS Excel / Prism etc as your go-to analysis software.</em> Why? Reproducibility. Workflows in standard spreadsheet programs require human interaction (i.e., selecting cells, entering formulae) and are prone to mistakes. An RMarkdown document is fully reproducible.</p></li>
</ul></section><section id="rstudio-overview" class="level1"><h1>RStudio Overview</h1>
<p>We’ll cover the basic steps of data import and analysis within RStudio and will highlight some features of the software as we go.</p>
<section id="import" class="level2"><h2>Import</h2>
<p>There are two main ways to import data within RStudio. One is interactive, using the “Import Dataset” button in the Environment panel.</p>
<p>The other is to use functions in <code>readr</code> or <code>readxl</code> to load data on the console or within an RMarkdown report.</p>
</section><section id="processing" class="level2"><h2>Processing</h2>
<p>The <code>tidyverse</code> is a collection of libraries that enable efficient data processing and visualization. You will probably start many analyses by loading the library:</p>
<pre class="sourceCode r" id="cb1"><code class="sourceCode r"><div class="sourceLine" id="cb1-1" data-line-number="1">
<span class="kw">library</span>(tidyverse)</div>
<div class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&gt; ── Attaching packages ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.2.0 ──</span></div>
<div class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; ✔ ggplot2 2.2.1     ✔ purrr   0.2.4</span></div>
<div class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; ✔ tibble  1.3.4     ✔ dplyr   0.7.4</span></div>
<div class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; ✔ tidyr   0.7.2     ✔ stringr 1.2.0</span></div>
<div class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; ✔ readr   1.1.1     ✔ forcats 0.2.0</span></div>
<div class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; ── Conflicts ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──</span></div>
<div class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></div>
<div class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></div></code></pre>
</section><section id="plotting" class="level2"><h2>Plotting</h2>
<p>The <code>tidyverse</code> contains <code>ggplot2</code>, which you covered in your pre-requisites. Let’s make a simple plot in RStudio. Note that the plot appears in the “Plots” panel. From there, you can Zoom, Export to a file, or Publish on RPubs.</p>
<pre class="sourceCode r" id="cb2"><code class="sourceCode r"><div class="sourceLine" id="cb2-1" data-line-number="1">
<span class="kw">library</span>(tidyverse)</div>
<div class="sourceLine" id="cb2-2" data-line-number="2"></div>
<div class="sourceLine" id="cb2-3" data-line-number="3">
<span class="kw">ggplot</span>(mtcars, <span class="kw">aes</span>(<span class="dt">x =</span> hp, <span class="dt">y =</span> mpg, <span class="dt">color =</span> <span class="kw">factor</span>(cyl))) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</div></code></pre>
<p><img src="img/class-1-plot-1.png" width="672" style="display: block; margin: auto;"></p>
</section><section id="reports" class="level2"><h2>Reports</h2>
<p>Finally we’ll create an RMarkdown document. Select File &gt; New File &gt; RMarkdown… and fill the Title and Author fields. Then click OK to create an RMarkdown document.</p>
</section></section><section id="real-world-example-data-collected-in-the-96-well-plate-format." class="level1"><h1>Real world example: Data collected in the 96-well plate format.</h1>
<p>We’ll use tidy data principles to analyze some qPCR data.</p>
<section id="formatting-data-for-efficient-analysis" class="level2"><h2>Formatting data for efficient analysis</h2>
<p>Reformatting your data into a “tidy” format is a useful practice that allows you efficiently manipulate and analyze data. We will start with some toy examples (using the <code>anscombe</code> data set) and then will reformat a commonly used data type from a quantitative PCR experiment.</p>
<ul>
<li>columns are variables</li>
<li>rows are observations</li>
<li>related data should be stored in separate tables</li>
</ul></section><section id="the-tidyr-package" class="level2"><h2>The <code>tidyr</code> package</h2>
<p>The <code>tidyr</code> package has some useful functions for rearranging data for efficient analysis. See the <a href="https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf">cheat sheet</a> for reference.</p>
<section id="gather-and-spread" class="level3"><h3>
<code>gather</code> and <code>spread</code>
</h3>
</section><section id="separate-extract-and-unite" class="level3"><h3>
<code>separate</code>, <code>extract</code>, and <code>unite</code>
</h3>
</section><section id="crossing-and-complete" class="level3"><h3>
<code>crossing</code> and <code>complete</code>
</h3>
<p>The <code>crossing</code> method is useful to generate combinations of variables.</p>
<pre class="sourceCode r" id="cb3"><code class="sourceCode r"><div class="sourceLine" id="cb3-1" data-line-number="1">sample &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'wt'</span>, <span class="st">'mut'</span>)</div>
<div class="sourceLine" id="cb3-2" data-line-number="2">gene &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'IFN'</span>, <span class="st">'ACTIN'</span>)</div>
<div class="sourceLine" id="cb3-3" data-line-number="3">rt &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'+'</span>, <span class="st">'-'</span>)</div>
<div class="sourceLine" id="cb3-4" data-line-number="4">rep &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>
</div>
<div class="sourceLine" id="cb3-5" data-line-number="5"></div>
<div class="sourceLine" id="cb3-6" data-line-number="6">tidyr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidyr/topics/expand">crossing</a></span>(sample, gene, rep, rt)</div>
<div class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; # A tibble: 24 x 4</span></div>
<div class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt;    sample  gene   rep    rt</span></div>
<div class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt;     &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;chr&gt;</span></div>
<div class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt;  1    mut ACTIN     1     -</span></div>
<div class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt;  2    mut ACTIN     1     +</span></div>
<div class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt;  3    mut ACTIN     2     -</span></div>
<div class="sourceLine" id="cb3-13" data-line-number="13"><span class="co">#&gt;  4    mut ACTIN     2     +</span></div>
<div class="sourceLine" id="cb3-14" data-line-number="14"><span class="co">#&gt;  5    mut ACTIN     3     -</span></div>
<div class="sourceLine" id="cb3-15" data-line-number="15"><span class="co">#&gt;  6    mut ACTIN     3     +</span></div>
<div class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">#&gt;  7    mut   IFN     1     -</span></div>
<div class="sourceLine" id="cb3-17" data-line-number="17"><span class="co">#&gt;  8    mut   IFN     1     +</span></div>
<div class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">#&gt;  9    mut   IFN     2     -</span></div>
<div class="sourceLine" id="cb3-19" data-line-number="19"><span class="co">#&gt; 10    mut   IFN     2     +</span></div>
<div class="sourceLine" id="cb3-20" data-line-number="20"><span class="co">#&gt; # ... with 14 more rows</span></div></code></pre>
</section></section><section id="naming-samples." class="level2"><h2>Naming samples.</h2>
<p>All variables should be systematically listed in your sample names, i.e. <code>name_rep_time_RT</code>. Systematic naming makes it easy to programmtically extract relevant information.</p>
<p>Take an example like this, where the sample names are a combination of a genotype (<code>WT</code> and <code>MT</code>), a time point (0,4,8,24 hour), and a replicate (1,2,3), separated by a hyphen.</p>
<pre><code>#&gt; # A tibble: 24 x 2
#&gt;    sample.name value
#&gt;  *       &lt;chr&gt; &lt;int&gt;
#&gt;  1     MT-t0-1    81
#&gt;  2     MT-t0-2    35
#&gt;  3     MT-t0-3    89
#&gt;  4    MT-t24-1    32
#&gt;  5    MT-t24-2    24
#&gt;  6    MT-t24-3    62
#&gt;  7     MT-t4-1    67
#&gt;  8     MT-t4-2    14
#&gt;  9     MT-t4-3    49
#&gt; 10     MT-t8-1    94
#&gt; # ... with 14 more rows</code></pre>
<p>Becauase the samples are named systematically, it is easy to separate this information into individual columns</p>
<pre class="sourceCode r" id="cb5"><code class="sourceCode r"><div class="sourceLine" id="cb5-1" data-line-number="1">sample_info &lt;-<span class="st"> </span>samples <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb5-2" data-line-number="2">
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidyr/topics/separate">separate</a></span>(</div>
<div class="sourceLine" id="cb5-3" data-line-number="3">    sample.name,</div>
<div class="sourceLine" id="cb5-4" data-line-number="4">    <span class="dt">into =</span> <span class="kw">c</span>(<span class="st">'sample'</span>, <span class="st">'hour'</span>, <span class="st">'rep'</span>),</div>
<div class="sourceLine" id="cb5-5" data-line-number="5">    <span class="dt">sep =</span> <span class="st">"-"</span>
</div>
<div class="sourceLine" id="cb5-6" data-line-number="6">  )</div></code></pre>
<p>Now we can use <code>dplyr</code> and <code>tidyr</code> functions to manipulate the data.</p>
<pre class="sourceCode r" id="cb6"><code class="sourceCode r"><div class="sourceLine" id="cb6-1" data-line-number="1">sample_info <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(sample, hour) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="kw">mean</span>(value))</div>
<div class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; # A tibble: 8 x 3</span></div>
<div class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; # Groups:   sample [?]</span></div>
<div class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt;   sample  hour `mean(value)`</span></div>
<div class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;         &lt;dbl&gt;</span></div>
<div class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; 1     MT    t0      68.33333</span></div>
<div class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; 2     MT   t24      39.33333</span></div>
<div class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; 3     MT    t4      43.33333</span></div>
<div class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt; 4     MT    t8      51.33333</span></div>
<div class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; 5     WT    t0      52.66667</span></div>
<div class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt; 6     WT   t24      76.33333</span></div>
<div class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt; 7     WT    t4      55.00000</span></div>
<div class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt; 8     WT    t8      42.33333</span></div>
<div class="sourceLine" id="cb6-14" data-line-number="14"></div>
<div class="sourceLine" id="cb6-15" data-line-number="15">sample_info <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">spread</span>(hour, value) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">t24_norm =</span> t24 <span class="op">-</span><span class="st"> </span>t0)</div>
<div class="sourceLine" id="cb6-16" data-line-number="16"><span class="co">#&gt; # A tibble: 6 x 7</span></div>
<div class="sourceLine" id="cb6-17" data-line-number="17"><span class="co">#&gt;   sample   rep    t0   t24    t4    t8 t24_norm</span></div>
<div class="sourceLine" id="cb6-18" data-line-number="18"><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span></div>
<div class="sourceLine" id="cb6-19" data-line-number="19"><span class="co">#&gt; 1     MT     1    81    32    67    94      -49</span></div>
<div class="sourceLine" id="cb6-20" data-line-number="20"><span class="co">#&gt; 2     MT     2    35    24    14    27      -11</span></div>
<div class="sourceLine" id="cb6-21" data-line-number="21"><span class="co">#&gt; 3     MT     3    89    62    49    33      -27</span></div>
<div class="sourceLine" id="cb6-22" data-line-number="22"><span class="co">#&gt; 4     WT     1    97    79     3    72      -18</span></div>
<div class="sourceLine" id="cb6-23" data-line-number="23"><span class="co">#&gt; 5     WT     2    45    80    97    12       35</span></div>
<div class="sourceLine" id="cb6-24" data-line-number="24"><span class="co">#&gt; 6     WT     3    16    70    65    43       54</span></div></code></pre>
<section id="resources" class="level3"><h3>Resources</h3>
<p>This <a href="https://speakerdeck.com/jennybc/how-to-name-files">slide deck</a> from Jenny Bryan has useful tips for naming files that are also applicable to sample names.</p>
</section></section><section id="organizing-samples." class="level2"><h2>Organizing samples.</h2>
<p>How should you organize samples on your plate? Should -RT and +RT samples alternate? Where should I put my “no template” controls?</p>
<pre class="sourceCode r" id="cb7"><code class="sourceCode r"><div class="sourceLine" id="cb7-1" data-line-number="1">
<span class="kw">library</span>(eda)</div>
<div class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">#&gt; Loading required package: valr</span></div>
<div class="sourceLine" id="cb7-3" data-line-number="3"></div>
<div class="sourceLine" id="cb7-4" data-line-number="4">qpcr</div>
<div class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt; # A tibble: 145 x 6</span></div>
<div class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt;     well fluor  target content       sample       cq</span></div>
<div class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;        &lt;chr&gt;    &lt;dbl&gt;</span></div>
<div class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt;  1   A01  SYBR IFNbeta Unkn-01      Control 31.82551</span></div>
<div class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt;  2   A02  SYBR IFNbeta Unkn-02 Control_noRT 36.14470</span></div>
<div class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">#&gt;  3   A03  SYBR IFNbeta Unkn-03         5ppp 18.45676</span></div>
<div class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">#&gt;  4   A04  SYBR IFNbeta Unkn-04    5ppp_noRT      NaN</span></div>
<div class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">#&gt;  5   A05  SYBR IFNbeta Unkn-05          3p1 19.01628</span></div>
<div class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">#&gt;  6   A06  SYBR IFNbeta Unkn-06     3p1_noRT      NaN</span></div>
<div class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">#&gt;  7   A07  SYBR IFNbeta Unkn-07          3p2 19.10928</span></div>
<div class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">#&gt;  8   A08  SYBR IFNbeta Unkn-08     3p2_noRT      NaN</span></div>
<div class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">#&gt;  9   A09  SYBR IFNbeta Unkn-09           2p 22.79299</span></div>
<div class="sourceLine" id="cb7-17" data-line-number="17"><span class="co">#&gt; 10   A10  SYBR IFNbeta Unkn-10      2p_noRT      NaN</span></div>
<div class="sourceLine" id="cb7-18" data-line-number="18"><span class="co">#&gt; # ... with 135 more rows</span></div></code></pre>
<!-- From Rachel: Basically, I calculate the average Ct between all the sample replicates for the gene of interest primer set and the internal control. Then you subtract the gene of interest average from the internal control average and use that change in CT to calculate the fold change relative to actin using the deal-delta CT method. -->
<pre class="sourceCode r" id="cb8"><code class="sourceCode r"><div class="sourceLine" id="cb8-1" data-line-number="1">qpcr_tidy &lt;-<span class="st"> </span>qpcr <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb8-2" data-line-number="2">
<span class="st">  </span><span class="kw">separate</span>(sample, <span class="dt">into =</span> <span class="kw">c</span>(<span class="st">'ligand'</span>, <span class="st">'RT'</span>), <span class="dt">sep =</span> <span class="st">'_'</span>, <span class="dt">fill =</span> <span class="st">'right'</span>) <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb8-3" data-line-number="3">
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>fluor, <span class="op">-</span>content, <span class="op">-</span>well) <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb8-4" data-line-number="4">
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">RT =</span> <span class="kw">is.na</span>(RT))</div></code></pre>
</section></section><section id="exercises" class="level1"><h1>Exercises</h1>
<ol type="1">
<li>Reformat the <code>datasaurus</code> dataset so that the following plotting code works:</li>
</ol>
<pre class="sourceCode r" id="cb9"><code class="sourceCode r"><div class="sourceLine" id="cb9-1" data-line-number="1">datasaurus_tidy <span class="op">%&gt;%</span>
</div>
<div class="sourceLine" id="cb9-2" data-line-number="2">
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span>
</div>
<div class="sourceLine" id="cb9-3" data-line-number="3">
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span>
</div>
<div class="sourceLine" id="cb9-4" data-line-number="4">
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>dataset)</div></code></pre>
<p>What does this plot tell you?</p>
<ol start="2" type="1">
<li>
<li>
<li>
</ol></section><section id="too-easy" class="level1"><h1>Too Easy?</h1>
<p>If you found those exercises too easy, try these:</p>
<ol type="1">
<li>Tidy the <code>anscombe</code> data set. Your tidy data should look like:</li>
</ol>
<pre><code>#&gt; # A tibble: 44 x 3
#&gt;      set     x     y
#&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
#&gt;  1     1     4  4.26
#&gt;  2     1     5  5.68
#&gt;  3     1     6  7.24
#&gt;  4     1     7  4.82
#&gt;  5     1     8  6.95
#&gt;  6     1     9  8.81
#&gt;  7     1    10  8.04
#&gt;  8     1    11  8.33
#&gt;  9     1    12 10.84
#&gt; 10     1    13  7.58
#&gt; # ... with 34 more rows</code></pre>
<ol start="2" type="1">
<li><p>Calculate summary statistics in your tidy <code>anscombe</code> data. Calculate <code>n</code>, <code>mean</code> for each <code>x</code> and <code>y</code> variable in the sets. Calculate the <code>cor</code> between <code>x</code> and <code>y</code> for each set. What do you infer from these summary statistics?</p></li>
<li><p>Plot the tidy <code>anscombe</code> data. Plot each <code>x</code>/<code>y</code> point on separate facets for each set, and plot a smoothing line. Examine the plots. How do they compare to the summary statistics?</p></li>
</ol>
<pre class="sourceCode r" id="cb11"><code class="sourceCode r"><div class="sourceLine" id="cb11-1" data-line-number="1">
<span class="kw">ggplot</span>(anscombe_tidy, <span class="kw">aes</span>(x, y)) <span class="op">+</span>
</div>
<div class="sourceLine" id="cb11-2" data-line-number="2">
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span>
</div>
<div class="sourceLine" id="cb11-3" data-line-number="3">
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>set) <span class="op">+</span>
</div>
<div class="sourceLine" id="cb11-4" data-line-number="4">
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">'lm'</span>, <span class="dt">se =</span> <span class="ot">FALSE</span>)</div></code></pre>
</section><section id="quiz" class="level1"><h1>Quiz</h1>
<p>Create an RMarkdown document and answer the questions from the Exercises section. Submit your final document as “Quiz 1” by Thurs at 10 PM.</p>
<p><strong>Your submitted document must knit to HTML without errors</strong>. I.e., when you click the “Knit” button, the document should build and display and HTML page.</p>
</section>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#goals-for-the-class">Goals for the class</a></li>
      <li>
<a href="#rstudio-overview">RStudio Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#import">Import</a></li>
      <li><a href="#processing">Processing</a></li>
      <li><a href="#plotting">Plotting</a></li>
      <li><a href="#reports">Reports</a></li>
      </ul>
</li>
      <li>
<a href="#real-world-example-data-collected-in-the-96-well-plate-format-">Real world example: Data collected in the 96-well plate format.</a><ul class="nav nav-pills nav-stacked">
<li><a href="#formatting-data-for-efficient-analysis">Formatting data for efficient analysis</a></li>
      <li><a href="#the-tidyr-package">The <code>tidyr</code> package</a></li>
      <li><a href="#naming-samples-">Naming samples.</a></li>
      <li><a href="#organizing-samples-">Organizing samples.</a></li>
      </ul>
</li>
      <li><a href="#exercises">Exercises</a></li>
      <li><a href="#too-easy">Too Easy?</a></li>
      <li><a href="#quiz">Quiz</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
