<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 1: Overview and data wrangling • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Wrangling</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Gene Expression Vignette</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Chromatin Vignette</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: Single-Cell Analysis</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Submitting Assignments</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/eda">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 1: Overview and data wrangling</h1>
                        <h4 class="author">Jay Hesselberth</h4>
            
            <h4 class="date">2017-10-22</h4>
          </div>

    
    
<div class="contents">
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>Content to start with. Update workstations if needed (devtools?).</p>
<div id="formatting-data-for-efficient-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#formatting-data-for-efficient-analysis" class="anchor"></a>Formatting data for efficient analysis</h2>
<p>Reformatting your data into a “tidy” format is a useful practice that allows you efficiently manipulate and analyze data. We will start with some toy examples (using the <code>anscombe</code> data set) and then will reformat a commonly used data type from a quantitative PCR experiment.</p>
<ul>
<li>columns are variables</li>
<li>rows are observations</li>
<li>related data should be stored in separate tables</li>
</ul>
</div>
<div id="the-tidyr-package" class="section level2">
<h2 class="hasAnchor">
<a href="#the-tidyr-package" class="anchor"></a>The <code>tidyr</code> package</h2>
<ul>
<li>gather, spread</li>
<li>separate, extract</li>
</ul>
</div>
<div id="real-world-examples-data-collected-in-the-96-well-plate-format-" class="section level2">
<h2 class="hasAnchor">
<a href="#real-world-examples-data-collected-in-the-96-well-plate-format-" class="anchor"></a>Real world examples: Data collected in the 96-well plate format.</h2>
<p>We’ll use tidy data principles to analyze some qPCR data.</p>
</div>
<div id="naming-samples-" class="section level2">
<h2 class="hasAnchor">
<a href="#naming-samples-" class="anchor"></a>Naming samples.</h2>
<p>This <a href="https://speakerdeck.com/jennybc/how-to-name-files">slide deck</a> from Jenny Bryan has useful tips for naming files that are also applicable to sample names.</p>
<p>All variables should be systematically listed in your sample names, i.e. <code>name_rep_time_RT</code>. Systematic naming makes it easy to programmtically extract relevant information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="co">#&gt; Loading tidyverse: ggplot2</span>
<span class="co">#&gt; Loading tidyverse: tibble</span>
<span class="co">#&gt; Loading tidyverse: tidyr</span>
<span class="co">#&gt; Loading tidyverse: readr</span>
<span class="co">#&gt; Loading tidyverse: purrr</span>
<span class="co">#&gt; Loading tidyverse: dplyr</span>
<span class="co">#&gt; Conflicts with tidy packages ----------------------------------------------</span>
<span class="co">#&gt; filter(): dplyr, stats</span>
<span class="co">#&gt; lag():    dplyr, stats</span>
<span class="kw">library</span>(stringr)

<span class="co"># for reproducible `sample`</span>
<span class="kw">set.seed</span>(<span class="dv">47681</span>)

samples &lt;-
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/tidyr/topics/expand">crossing</a></span>(
    <span class="dt">name =</span> <span class="kw">c</span>(<span class="st">'WT'</span>, <span class="st">'MT'</span>),
    <span class="dt">hours =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">8</span>, <span class="dv">24</span>),
    <span class="dt">reps =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(
    <span class="dt">value =</span> <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>, <span class="kw">n</span>(), <span class="dt">replace =</span> <span class="ot">TRUE</span>),
    <span class="dt">.id =</span> <span class="kw">row_number</span>())</code></pre></div>
</div>
<div id="organizing-samples-" class="section level2">
<h2 class="hasAnchor">
<a href="#organizing-samples-" class="anchor"></a>Organizing samples.</h2>
<p>How should you organize samples on your plate? Should -RT and +RT samples alternate? Where should I put my “no template” controls?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(eda)
<span class="co">#&gt; Loading required package: valr</span>

qpcr
<span class="co">#&gt; # A tibble: 145 x 6</span>
<span class="co">#&gt;     well fluor  target content       sample       cq</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;        &lt;chr&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1   A01  SYBR IFNbeta Unkn-01      Control 31.82551</span>
<span class="co">#&gt;  2   A02  SYBR IFNbeta Unkn-02 Control_noRT 36.14470</span>
<span class="co">#&gt;  3   A03  SYBR IFNbeta Unkn-03         5ppp 18.45676</span>
<span class="co">#&gt;  4   A04  SYBR IFNbeta Unkn-04    5ppp_noRT      NaN</span>
<span class="co">#&gt;  5   A05  SYBR IFNbeta Unkn-05          3p1 19.01628</span>
<span class="co">#&gt;  6   A06  SYBR IFNbeta Unkn-06     3p1_noRT      NaN</span>
<span class="co">#&gt;  7   A07  SYBR IFNbeta Unkn-07          3p2 19.10928</span>
<span class="co">#&gt;  8   A08  SYBR IFNbeta Unkn-08     3p2_noRT      NaN</span>
<span class="co">#&gt;  9   A09  SYBR IFNbeta Unkn-09           2p 22.79299</span>
<span class="co">#&gt; 10   A10  SYBR IFNbeta Unkn-10      2p_noRT      NaN</span>
<span class="co">#&gt; # ... with 135 more rows</span></code></pre></div>
<!-- From Rachel: Basically, I calculate the average Ct between all the sample replicates for the gene of interest primer set and the internal control. Then you subtract the gene of interest average from the internal control average and use that change in CT to calculate the fold change relative to actin using the deal-delta CT method. -->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">qpcr_tidy &lt;-<span class="st"> </span>qpcr <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">separate</span>(sample, <span class="dt">into =</span> <span class="kw">c</span>(<span class="st">'ligand'</span>, <span class="st">'RT'</span>), <span class="dt">sep =</span> <span class="st">'_'</span>, <span class="dt">fill =</span> <span class="st">'right'</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>fluor, <span class="op">-</span>content, <span class="op">-</span>well) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">RT =</span> <span class="kw">is.na</span>(RT))</code></pre></div>
</div>
<div id="exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h2>
</div>
</div>
<div id="quiz" class="section level1">
<h1 class="hasAnchor">
<a href="#quiz" class="anchor"></a>Quiz</h1>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#formatting-data-for-efficient-analysis">Formatting data for efficient analysis</a></li>
      <li><a href="#the-tidyr-package">The <code>tidyr</code> package</a></li>
      <li><a href="#real-world-examples-data-collected-in-the-96-well-plate-format-">Real world examples: Data collected in the 96-well plate format.</a></li>
      <li><a href="#naming-samples-">Naming samples.</a></li>
      <li><a href="#organizing-samples-">Organizing samples.</a></li>
      <li><a href="#exercises">Exercises</a></li>
      </ul>
</li>
      <li><a href="#quiz">Quiz</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
