<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exercises 4 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Genomic Interval Analysis part 2</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: RNA-Seq Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-7.html">Class 7: RNA-Seq Analysis part 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li>
      <a href="../articles/exercises-4.html">Exercises 4</a>
    </li>
    <li>
      <a href="../articles/exercises-5.html">Exercises 5</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Exercises 4</h1>
                        <h4 class="author">Your Name Here</h4>
            
            <h4 class="date">12/11/2017</h4>
          </div>

    
    
<div class="contents">
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>Here is a <a href="https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-4.Rmd">link</a> to the text of these exercises.</p>
</div>
<div id="question-1" class="section level1">
<h1 class="hasAnchor">
<a href="#question-1" class="anchor"></a>Question 1</h1>
<p>What is the difference between <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> and <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_random">bed_random()</a></code>? Explain your answer. No code is necessary, but you are welcome to show some example usage if it helps your explanation.</p>
<div id="interpretation" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-2" class="section level1">
<h1 class="hasAnchor">
<a href="#question-2" class="anchor"></a>Question 2</h1>
<p>Write a function that computes the final tibble needed for plotting a metagene. Explain in your interpretation why writing a function is useful for this task.</p>
<p>The function should take three arguments. - <code>genes_bed</code> is a bed tibble that contains positions of human genes. (<code>chrom</code>, <code>start</code>, <code>end</code>, <code>name</code>, <code>score</code>, <code>strand</code>) - <code>chipseq_bedgraph</code> is a bedgraph tibble (<code>chrom</code>, <code>start</code>, <code>end</code>, <code>value</code>) - <code>genome_bed</code> is a genome tibble (<code>chrom</code>, <code>size</code>)</p>
<p>Run your function using the following bed tibbles as input arguments</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(eda)
<span class="kw">library</span>(tidyverse)</code></pre></div>
<pre><code>## ── Attaching packages ──────────────────────────────────────────────────────────────────────────────────── tidyverse 1.2.1 ──</code></pre>
<pre><code>## ✔ ggplot2 2.2.1     ✔ purrr   0.2.4
## ✔ tibble  1.3.4     ✔ dplyr   0.7.4
## ✔ tidyr   0.7.2     ✔ stringr 1.2.0
## ✔ readr   1.1.1     ✔ forcats 0.2.0</code></pre>
<pre><code>## ── Conflicts ─────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># genes_bed </span>
genes_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genes.bed.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(genes_file, <span class="dt">n_fields =</span> <span class="dv">6</span>)

<span class="co"># chipseq_bedgraph</span>
bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/valr_example">valr_example</a></span>(<span class="st">"hela.h3k4.chip.bg.gz"</span>),
                <span class="dt">n_fields =</span> <span class="dv">4</span>)
bg &lt;-<span class="st"> </span><span class="kw">select</span>(bed, <span class="kw">everything</span>(), <span class="dt">value =</span> name)
bg &lt;-<span class="st"> </span><span class="kw">mutate</span>(bg, 
              <span class="dt">value =</span> <span class="kw">as.numeric</span>(value))

<span class="co"># genome_bed </span>
genome_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genome.txt.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_genome">read_genome</a></span>(genome_file)</code></pre></div>
<p>Your output tibble should look like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       46.97059</span>
<span class="co">#&gt;  2       2       56.48529</span>
<span class="co">#&gt;  3       3       47.42667</span>
<span class="co">#&gt;  4       4       53.97297</span>
<span class="co">#&gt;  5       5       48.36842</span>
<span class="co">#&gt;  6       6       56.05333</span>
<span class="co">#&gt;  7       7       47.08861</span>
<span class="co">#&gt;  8       8       55.15190</span>
<span class="co">#&gt;  9       9       47.33750</span>
<span class="co">#&gt; 10      10       55.94937</span>
<span class="co">#&gt; # ... with 191 more rows</span></code></pre></div>
<div id="strategy" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-1" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-1" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-3" class="section level1">
<h1 class="hasAnchor">
<a href="#question-3" class="anchor"></a>Question 3</h1>
<p>Write a function that plots a metagene. The function should take one argument, the output from question 2, and produce a plot. Run the function on the output of question 2, and show your plot.</p>
<p>Here’s some pseudo-code to get you started. I’ve also included some code that fixes the x-axis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plot_metagene &lt;-<span class="st"> </span><span class="cf">function</span>(dat) {
  
  region_size &lt;-<span class="st"> </span><span class="dv">1000</span>
  win_size &lt;-<span class="st"> </span><span class="dv">10</span>

  x_labels &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span>region_size, region_size, <span class="dt">by =</span> win_size <span class="op">*</span><span class="st"> </span><span class="dv">25</span>)
  x_breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">201</span>, <span class="dt">by =</span> <span class="dv">25</span>)

  <span class="kw">ggplot</span>(..., <span class="kw">aes</span>(..., ...)) <span class="op">+</span>
<span class="st">  </span>... <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks)  <span class="op">+</span>
}

<span class="kw">plot_metagene</span>(metagene_data)</code></pre></div>
<div id="strategy-1" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy-1" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-2" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-2" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-4" class="section level1">
<h1 class="hasAnchor">
<a href="#question-4" class="anchor"></a>Question 4</h1>
<p>Use the <code>brauer_gene_exp</code> table to answer the following questions.</p>
<ul>
<li><p>Are there any rows with <code>NA</code> values in <code>brauer_gene_exp</code>? Show your work. Hint: calculate the number of rows before and after removing <code>NA</code> values.</p></li>
<li><p>Which <code>systematic_name</code> and <code>nutrient</code> combination has the highest expression value? Interpret this result in the section below, keeping in mind that the <code>nutrient</code> condition indicates when the nutrient is <em>omitted</em> from the growth medium. Look up this gene at the <a href="http://yeastgenome.org">Saccharomyces Genome Database</a>.</p></li>
</ul>
<div id="strategy-2" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy-2" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-3" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-3" class="anchor"></a>Interpretation</h2>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li>
<a href="#question-1">Question 1</a><ul class="nav nav-pills nav-stacked">
<li><a href="#interpretation">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-2">Question 2</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy">Strategy</a></li>
      <li><a href="#interpretation-1">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-3">Question 3</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy-1">Strategy</a></li>
      <li><a href="#interpretation-2">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-4">Question 4</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy-2">Strategy</a></li>
      <li><a href="#interpretation-3">Interpretation</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
