<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 9: Tips, Data I/O, and Resources • pbda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Class 9: Tips, Data I/O, and Resources">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pbda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/practical-data-analysis">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Class 9: Tips, Data I/O, and Resources</h1>
                        <h4 class="author">Kent Riemondy</h4>
            
            <h4 class="date">2019-12-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/master/vignettes/class-9.Rmd"><code>vignettes/class-9.Rmd</code></a></small>
      <div class="hidden name"><code>class-9.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a></code></pre></div>
<div id="rmarkdown" class="section level1">
<h1 class="hasAnchor">
<a href="#rmarkdown" class="anchor"></a>Rmarkdown</h1>
<p>Read the <a href="https://bookdown.org/yihui/rmarkdown/">Guide to RMarkdown</a> for an exhaustive description of the various formats and options for using RMarkdown documents.</p>
<div id="caching" class="section level2">
<h2 class="hasAnchor">
<a href="#caching" class="anchor"></a>caching</h2>
<p>You can speed up knitting of your RMarkdowns by using caching to store the results from each chunk.</p>
<p><code>{r, cache = TRUE}</code></p>
<p>Or set globally at top of document</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">knitr<span class="op">::</span>opts_chunk<span class="op">$</span><span class="kw">set</span>(<span class="dt">cache =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="mixing-r-code-with-other-languages" class="section level2">
<h2 class="hasAnchor">
<a href="#mixing-r-code-with-other-languages" class="anchor"></a>Mixing R code with other languages</h2>
<p>In RMarkdown you can run other languages (<code>python</code>, <code>C++</code>, <code>C</code>, <code>shell</code>, <code>julia</code>, etc.) in each code chunk.</p>
<div id="shell" class="section level3">
<h3 class="hasAnchor">
<a href="#shell" class="anchor"></a>Shell</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="bu">echo</span> hello world</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">#&gt; hello world</span></a></code></pre></div>
</div>
<div id="python" class="section level3">
<h3 class="hasAnchor">
<a href="#python" class="anchor"></a>Python</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="im">import</span> datetime </a>
<a class="sourceLine" id="cb4-2" data-line-number="2">now <span class="op">=</span> datetime.datetime.now()</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="bu">print</span>(<span class="st">"hello the date and time is currently </span><span class="sc">{}</span><span class="st">"</span>.<span class="bu">format</span>(now))</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt; hello the date and time is currently 2019-12-10 21:50:17.093179</span></a></code></pre></div>
</div>
</div>
</div>
<div id="rdata-files" class="section level1">
<h1 class="hasAnchor">
<a href="#rdata-files" class="anchor"></a>.Rdata files</h1>
<p>When you close R you are prompted to save your workspace. This will store a <code>.RData</code> file that will contain all of the R objects that you generated in your current session. Upon startup of a “new” session this <code>.Rdata</code> file will populate your environment with old R objects.</p>
<p>This can cause many problems, including slow startup time, conflicts between your current objects and old objects, and irreproducible analyses.</p>
<p>I recommend never storing your <code>.RData</code>. In RStudio you can turn off this option under preferences.</p>
</div>
<div id="reading-and-writing-data" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-and-writing-data" class="anchor"></a>Reading and writing data</h1>
<div id="readr" class="section level2">
<h2 class="hasAnchor">
<a href="#readr" class="anchor"></a>readr</h2>
<p>The <code>readr</code> package (part of the tidyverse) provides a suite of functions to read data into R. These functions generally are faster and have better default behavior than the base R equivalents.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">read_tsv</span>()</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">read_csv</span>()</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">read_delim</span>()</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw">read_lines</span>()</a></code></pre></div>
<p>The <code>readr</code> functions take a path to the file as an argument or a URL directly.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># read directly from a URL</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">read_tsv</span>(<span class="st">"https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/inst/extdata/david.txt"</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co"># alternatively download the file, then read it into R.</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span>(<span class="st">"https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/inst/extdata/david.txt"</span>, <span class="st">"david.txt"</span>)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="kw">read_tsv</span>(<span class="st">"david.txt"</span>)</a></code></pre></div>
<p>Here is example to read data stored internally from the <code>pbda</code> package.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">path_to_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"david.txt"</span>, <span class="dt">package =</span> <span class="st">"pbda"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">path_to_file</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#&gt; [1] "/home/travis/R/Library/pbda/extdata/david.txt"</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="kw">read_tsv</span>(path_to_file) </a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt; Parsed with column specification:</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">#&gt; cols(</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt;   Category = col_character(),</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt;   Term = col_character(),</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">#&gt;   Count = col_double(),</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">#&gt;   `%` = col_double(),</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">#&gt;   PValue = col_double(),</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">#&gt;   Genes = col_character(),</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">#&gt;   `List Total` = col_double(),</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">#&gt;   `Pop Hits` = col_double(),</span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">#&gt;   `Pop Total` = col_double(),</span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co">#&gt;   `Fold Enrichment` = col_double(),</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18"><span class="co">#&gt;   Bonferroni = col_double(),</span></a>
<a class="sourceLine" id="cb7-19" data-line-number="19"><span class="co">#&gt;   Benjamini = col_double(),</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20"><span class="co">#&gt;   FDR = col_double()</span></a>
<a class="sourceLine" id="cb7-21" data-line-number="21"><span class="co">#&gt; )</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22"><span class="co">#&gt; # A tibble: 70 x 13</span></a>
<a class="sourceLine" id="cb7-23" data-line-number="23"><span class="co">#&gt;    Category Term  Count     `%`  PValue Genes `List Total` `Pop Hits`</span></a>
<a class="sourceLine" id="cb7-24" data-line-number="24"><span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb7-25" data-line-number="25"><span class="co">#&gt;  1 KEGG_PA… hsa0…    21 0.00744 4.37e-5 GSTA…          717         74</span></a>
<a class="sourceLine" id="cb7-26" data-line-number="26"><span class="co">#&gt;  2 KEGG_PA… hsa0…    33 0.0117  5.29e-5 HLA-…          717        150</span></a>
<a class="sourceLine" id="cb7-27" data-line-number="27"><span class="co">#&gt;  3 KEGG_PA… hsa0…    18 0.00638 2.04e-4 CYP2…          717         64</span></a>
<a class="sourceLine" id="cb7-28" data-line-number="28"><span class="co">#&gt;  4 KEGG_PA… hsa0…    11 0.00390 2.27e-4 UGT1…          717         27</span></a>
<a class="sourceLine" id="cb7-29" data-line-number="29"><span class="co">#&gt;  5 KEGG_PA… hsa0…    38 0.0135  3.28e-4 TNF,…          717        200</span></a>
<a class="sourceLine" id="cb7-30" data-line-number="30"><span class="co">#&gt;  6 KEGG_PA… hsa0…    18 0.00638 4.45e-4 GSTA…          717         68</span></a>
<a class="sourceLine" id="cb7-31" data-line-number="31"><span class="co">#&gt;  7 KEGG_PA… hsa0…    63 0.0223  4.50e-4 ADCY…          717        393</span></a>
<a class="sourceLine" id="cb7-32" data-line-number="32"><span class="co">#&gt;  8 KEGG_PA… hsa0…    14 0.00496 5.90e-4 UGT1…          717         46</span></a>
<a class="sourceLine" id="cb7-33" data-line-number="33"><span class="co">#&gt;  9 KEGG_PA… hsa0…    22 0.00779 6.09e-4 PRKC…          717         95</span></a>
<a class="sourceLine" id="cb7-34" data-line-number="34"><span class="co">#&gt; 10 KEGG_PA… hsa0…    11 0.00390 1.40e-3 UGT1…          717         33</span></a>
<a class="sourceLine" id="cb7-35" data-line-number="35"><span class="co">#&gt; # … with 60 more rows, and 5 more variables: `Pop Total` &lt;dbl&gt;, `Fold</span></a>
<a class="sourceLine" id="cb7-36" data-line-number="36"><span class="co">#&gt; #   Enrichment` &lt;dbl&gt;, Bonferroni &lt;dbl&gt;, Benjamini &lt;dbl&gt;, FDR &lt;dbl&gt;</span></a></code></pre></div>
<p>Often files are not formatted perfectly for reading into R. Below are some common problems, and the solutions</p>
<ol style="list-style-type: decimal">
<li>File has a header line that should not be read. (use the <code>skip</code> argument to skip <code>n</code> lines)</li>
<li>File has column names that are poorly formatted (use the <code>col_names = FALSE</code> to skip them, then name them yourself)</li>
<li>A column in the data is interpreted as a number but then a character is found. (use the <code>col_types</code> argument to specify the data type for each column)</li>
<li>The file is a mess. (Use <code>read_lines()</code> to read in each line into a character vector and convert to a tibble yourself)</li>
</ol>
<p>Similar to the read functions there are also utilities to write to a file.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">write_tsv</span>(mtcars, <span class="st">"mtcars.txt"</span>)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">write_csv</span>(mtcars, <span class="st">"mtcars.csv"</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co"># supplying a .gz will compress the data</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="kw">write_tsv</span>(mtcars, <span class="st">"mtcars.txt.gz"</span>)</a></code></pre></div>
</div>
<div id="base-r-functions-are-sometimes-helpful" class="section level2">
<h2 class="hasAnchor">
<a href="#base-r-functions-are-sometimes-helpful" class="anchor"></a>base R functions are sometimes helpful</h2>
<p>They can be slow, and by default will convert any character vector into a factor, which is often not desired. Set <code>stringsAsFactors = FALSE</code> to avoid this behavior.</p>
<p>Reading:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="st">"your_file.csv"</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"your_file.txt"</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>Writing:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span>(mtcars, <span class="st">"mtcars.txt"</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span>((mtcars, <span class="st">"mtcars.csv"</span>)</a></code></pre></div>
</div>
<div id="writing-and-reading-binary--rds-or--rda-files" class="section level2">
<h2 class="hasAnchor">
<a href="#writing-and-reading-binary--rds-or--rda-files" class="anchor"></a>Writing and reading binary <code>.rds</code> or <code>.rda</code> files</h2>
<p>Often it is useful to save a R object as a file directly. For example the seurat objects can be saved to disk as an <code>.rds</code> file which can be directly loaded into R at a later date. These files can also be shared with others.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># save mtcars as a binary file</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(mtcars, <span class="st">"mtcars.rds"</span>)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co"># read mtcars into R as an R object called new_data</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5">new_data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="st">"mtcars.rds"</span>)</a></code></pre></div>
<p><code>.rda</code> files are binary files that can be read into R as prenamed objects. They can store many objects in a single file.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">simple_data &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">a =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">simple_data2 &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">simple_data3 &lt;-<span class="st"> </span><span class="kw">bind_cols</span>(simple_data, simple_data2)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co"># pass R objects to the save function</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="kw"><a href="https://rdrr.io/r/base/save.html">save</a></span>(simple_data, </a>
<a class="sourceLine" id="cb12-7" data-line-number="7">  simple_data2, </a>
<a class="sourceLine" id="cb12-8" data-line-number="8">  simple_data3, </a>
<a class="sourceLine" id="cb12-9" data-line-number="9">  <span class="dt">file =</span> <span class="st">"collection_of_r_objects.rda"</span>)</a>
<a class="sourceLine" id="cb12-10" data-line-number="10"></a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="co"># in a new R session use load() to reload these objects into your environment</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="kw"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="st">"collection_of_r_objects.rda"</span>)</a>
<a class="sourceLine" id="cb12-13" data-line-number="13"></a>
<a class="sourceLine" id="cb12-14" data-line-number="14">simple_data</a>
<a class="sourceLine" id="cb12-15" data-line-number="15"><span class="co">#&gt; # A tibble: 10 x 1</span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16"><span class="co">#&gt;        a</span></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"><span class="co">#&gt;    &lt;int&gt;</span></a>
<a class="sourceLine" id="cb12-18" data-line-number="18"><span class="co">#&gt;  1     1</span></a>
<a class="sourceLine" id="cb12-19" data-line-number="19"><span class="co">#&gt;  2     2</span></a>
<a class="sourceLine" id="cb12-20" data-line-number="20"><span class="co">#&gt;  3     3</span></a>
<a class="sourceLine" id="cb12-21" data-line-number="21"><span class="co">#&gt;  4     4</span></a>
<a class="sourceLine" id="cb12-22" data-line-number="22"><span class="co">#&gt;  5     5</span></a>
<a class="sourceLine" id="cb12-23" data-line-number="23"><span class="co">#&gt;  6     6</span></a>
<a class="sourceLine" id="cb12-24" data-line-number="24"><span class="co">#&gt;  7     7</span></a>
<a class="sourceLine" id="cb12-25" data-line-number="25"><span class="co">#&gt;  8     8</span></a>
<a class="sourceLine" id="cb12-26" data-line-number="26"><span class="co">#&gt;  9     9</span></a>
<a class="sourceLine" id="cb12-27" data-line-number="27"><span class="co">#&gt; 10    10</span></a></code></pre></div>
</div>
<div id="writing-and-reading-xlsx-files-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#writing-and-reading-xlsx-files-in-r" class="anchor"></a>Writing and reading xlsx files in R</h2>
<p>There are two packages useful for working with excel files.</p>
<p><code>readxl</code> is really good for reading xlsx files into R. <code>openxlsx</code> is pretty good at writing xlsx files.</p>
</div>
</div>
<div id="organizing-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#organizing-analyses" class="anchor"></a>Organizing Analyses</h1>
<p>Read this paper: <a href="https://doi.org/10.1371/journal.pcbi.1000424">A Quick Guide to Organizing Computational Biology Projects</a></p>
<p>Here’s how I generally organize my projects.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="ex">NEWS.md</span>   # markdown document for tracking progress</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="ex">data</span>      # data directory for storing raw and processed data</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="ex">dbases</span>    # database files downloaded from other places</a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="ex">docs</span>      # publication documents and random project files</a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="ex">your-project.Rproj</span> <span class="co"># Make an Rproject in Rstudio for every project</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6"><span class="ex">results</span>   # store all RMarkdown here</a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="ex">src</span>       # store useful scripts here</a></code></pre></div>
<p>In your results directory make a new directory with a proper date naming scheme for each analyses. In each directory have an RMarkdown.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="ex">2018-12-06_dataqc</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="ex">2018-12-08_exploratory_analysis</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="ex">2018-12-10_summary_plots</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="ex">2018-12-14_publication_figures</span></a></code></pre></div>
</div>
<div id="gitgithub" class="section level1">
<h1 class="hasAnchor">
<a href="#gitgithub" class="anchor"></a>Git/Github</h1>
<p><strong>Use it.</strong> It’s a great way to organize your projects, provide visibility, and foster reproducibility.</p>
<p>As a student github will give you 5 private repositories for free. However using github with public repositories is preferred if the code or results aren’t sensitive.</p>
<p><a href="https://education.github.com/students" class="uri">https://education.github.com/students</a></p>
</div>
<div id="finding-help" class="section level1">
<h1 class="hasAnchor">
<a href="#finding-help" class="anchor"></a>Finding help</h1>
<div id="hopefully-not-your-experience" class="section level2">
<h2 class="hasAnchor">
<a href="#hopefully-not-your-experience" class="anchor"></a>Hopefully not your experience</h2>
<p><img src="https://imgs.xkcd.com/comics/wisdom_of_the_ancients.png"></p>
</div>
<div id="online-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#online-resources" class="anchor"></a>Online resources</h2>
<p>The <a href="https://community.rstudio.com/">R studio community forums</a> are a great resource for asking quesitons about tidyverse related packages.</p>
<p><a href="https://stackoverflow.com/">StackOverflow</a> provides user contributes questions and answers on a variety of topics.</p>
<p>For help with bioconductor packages visit the <a href="https://support.bioconductor.org/">support page</a></p>
<p>Find out if others are having similar issues by searching the issue on the <a href="https://github.com/">Github</a> page for the package that you are using.</p>
</div>
<div id="general-r-coding-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#general-r-coding-resources" class="anchor"></a>General R coding resources</h2>
<p>Hadley Wickman has written two very good (free) ebooks on using R.</p>
<p><a href="https://r4ds.had.co.nz/">R for Data Science</a></p>
<p><a href="http://adv-r.had.co.nz/">Advanced R</a></p>
</div>
<div id="bioinformatic-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#bioinformatic-resources" class="anchor"></a>Bioinformatic resources</h2>
<p>For general bioinformatics advice I found the following text very useful.</p>
<p><a href="http://vincebuffalo.org/book/">Bioinformatics Data Skills</a> by Vince Buffalo</p>
<p>For statistics related to bioinformatics this free course is excellent:</p>
<p><a href="http://genomicsclass.github.io/book/">PH525x series - Biomedical Data Science</a></p>
<p>For more detailed descriptions of single cell RNA-Seq analysis:</p>
<p><a href="http://hemberg-lab.github.io/scRNA.seq.course/" class="uri">http://hemberg-lab.github.io/scRNA.seq.course/</a></p>
</div>
<div id="offline-help" class="section level2">
<h2 class="hasAnchor">
<a href="#offline-help" class="anchor"></a>Offline help</h2>
<p>We (the RBI fellows) hold standing office hours from 1-2:30p on Thursdays in RC1S rm. 9101. Bring us your broken code and we’ll try to sort it out. Send us an email to let us know you will be stopping by (<code>rbi.fellows@ucdenver.edu</code>).</p>
</div>
<div id="lastly-sometimes-code-is-broken" class="section level2">
<h2 class="hasAnchor">
<a href="#lastly-sometimes-code-is-broken" class="anchor"></a>Lastly sometimes code is broken</h2>
<p>Go to github and search the issues to see if the problem has been reported or fixed. If not submit an issue that describes the problem. The <a href="https://reprex.tidyverse.org/">reprex package</a> makes it easy to produce well-formatted R reproducible examples that demonstrate the problem. Often developers will be thankful for your help with making their software better.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#rmarkdown">Rmarkdown</a><ul class="nav nav-pills nav-stacked">
<li><a href="#caching">caching</a></li>
      <li><a href="#mixing-r-code-with-other-languages">Mixing R code with other languages</a></li>
      </ul>
</li>
      <li><a href="#rdata-files">.Rdata files</a></li>
      <li>
<a href="#reading-and-writing-data">Reading and writing data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#readr">readr</a></li>
      <li><a href="#base-r-functions-are-sometimes-helpful">base R functions are sometimes helpful</a></li>
      <li><a href="#writing-and-reading-binary--rds-or--rda-files">Writing and reading binary <code>.rds</code> or <code>.rda</code> files</a></li>
      <li><a href="#writing-and-reading-xlsx-files-in-r">Writing and reading xlsx files in R</a></li>
      </ul>
</li>
      <li><a href="#organizing-analyses">Organizing Analyses</a></li>
      <li><a href="#gitgithub">Git/Github</a></li>
      <li>
<a href="#finding-help">Finding help</a><ul class="nav nav-pills nav-stacked">
<li><a href="#hopefully-not-your-experience">Hopefully not your experience</a></li>
      <li><a href="#online-resources">Online resources</a></li>
      <li><a href="#general-r-coding-resources">General R coding resources</a></li>
      <li><a href="#bioinformatic-resources">Bioinformatic resources</a></li>
      <li><a href="#offline-help">Offline help</a></li>
      <li><a href="#lastly-sometimes-code-is-broken">Lastly sometimes code is broken</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy, Rui Fu, Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
