<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 9: Exploring data with PCA • pbda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Class 9: Exploring data with PCA">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pbda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    </li>
    <li>
      <a href="../articles/class-6.html">Class 6: Writing Functions</a>
    </li>
    <li>
      <a href="../articles/class-7.html">Class 7: Iteration</a>
    </li>
    <li>
      <a href="../articles/class-8.html">Class 8: Heatmaps and Clustering</a>
    </li>
    <li>
      <a href="../articles/class-9.html">Class 9: Exploring data with PCA</a>
    </li>
    <li>
      <a href="../articles/class-10.html">Class 10: Jumping into more complex R, tips, resources</a>
    </li>
    <li>
      <a href="../articles/class-11.html">Class 11: RNA-seq Analysis</a>
    </li>
    <li>
      <a href="../articles/class-12.html">Class 12: Working with gene lists</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/practical-data-analysis">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Class 9: Exploring data with PCA</h1>
                        <h4 class="author">Kent Riemondy</h4>
            
            <h4 class="date">2019-12-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/master/vignettes/class-9.Rmd"><code>vignettes/class-9.Rmd</code></a></small>
      <div class="hidden name"><code>class-9.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ComplexHeatmap)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggrepel)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pbda)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(cowplot)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="kw">theme_set</span>(<span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>())</a></code></pre></div>
<p>Goals: Use PCA to identify sources of variation and summarize relationship between variables/samples in a large dataset.</p>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>As we’ve learned throughout the class, you need to visualize your data to identify key patterns. However, high dimensional datasets are difficult to visualize in a meaningful way. For example how do you find a pattern in a large sequencing dataset with 100 samples, and 20,000 genes?</p>
<p>Heatmaps provide one way to visualize selectively gene sets for a relatively small number of samples, but discerning overall patterns between samples can still be difficult, even with clustering.</p>
<p>Principle Component Analysis is a commonly used technique to examine the structure of large datasets.</p>
<p>Resources:</p>
<ul>
<li><a href="http://setosa.io/ev/principal-component-analysis/">PCA Explained Visually</a></li>
<li><a href="https://en.wikipedia.org/wiki/Principal_component_analysis">PCA formal explanation</a></li>
<li><a href="https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues">PCA informal explanation</a></li>
</ul>
<div id="background-on-pca" class="section level3">
<h3 class="hasAnchor">
<a href="#background-on-pca" class="anchor"></a>Background on PCA</h3>
<p>PCA takes a set of variables and generates a new set of variables. The new variables are selected to capture the maximal variance in the data.</p>
<p>Each new variable is known as a principal component. The first principal component represents a linear combination of the original variables that captures the maximum variance in the dataset. Basically find a vector through the dataset that captures the maximum variance.</p>
<p>The second component is also a linear combination of the original variables but is uncorrelated to the first component.</p>
<p>In this manner PCA generates a new set of variables that are ordered based on the variance captured in the data. It’s for this reason that is so useful for exploratory analysis and dimensionality reduction.</p>
<p>Consider the following experiment where we measure gene expression for 2 genes in 100 samples. Because there are only 2 genes, we can visualize all of the data in a 2D scatterplot.</p>
<p>First, let’s generate some example data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># generate some correlated variables, scale, and make tidy </span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">42</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">sim &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/MASS/man/mvrnorm.html">mvrnorm</a></span>(<span class="dv">100</span>,</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">                     <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">0</span>),</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">                     <span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="fl">0.95</span>, <span class="fl">0.95</span>, <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">                     <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">                     <span class="dv">2</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="st">  </span><span class="kw">rownames_to_column</span>(<span class="st">"id"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="st">  </span><span class="kw">arrange</span>(V1, V2) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="st">  </span><span class="kw">rename</span>(<span class="dt">gene_1 =</span> V1, <span class="dt">gene_2 =</span> V2) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">id =</span> <span class="kw">row_number</span>())</a>
<a class="sourceLine" id="cb2-15" data-line-number="15"></a>
<a class="sourceLine" id="cb2-16" data-line-number="16">sim</a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt; # A tibble: 100 x 3</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt;       id gene_1 gene_2</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt;    &lt;int&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt;  1     1  -3.00  -2.12</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt;  2     2  -2.82  -2.93</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt;  3     3  -2.28  -2.38</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt;  4     4  -2.17  -2.54</span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co">#&gt;  5     5  -1.53  -1.31</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#&gt;  6     6  -1.52  -1.93</span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26"><span class="co">#&gt;  7     7  -1.52  -1.81</span></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co">#&gt;  8     8  -1.42  -1.99</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="co">#&gt;  9     9  -1.30  -1.41</span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="co">#&gt; 10    10  -1.29  -1.38</span></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="co">#&gt; # … with 90 more rows</span></a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># select points to label for visualization</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">max_min_points &lt;-<span class="st"> </span><span class="kw">slice</span>(sim, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">95</span><span class="op">:</span><span class="dv">100</span>))</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co"># make a plot</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">og_plt &lt;-<span class="st"> </span><span class="kw">ggplot</span>(sim, <span class="kw">aes</span>(gene_<span class="dv">1</span>, gene_<span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>(<span class="dt">data =</span> max_min_points, </a>
<a class="sourceLine" id="cb3-8" data-line-number="8">                  <span class="kw">aes</span>(<span class="dt">label =</span> id)) <span class="op">+</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="st">  </span><span class="co"># scale y axis to same as x</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(sim<span class="op">$</span>gene_<span class="dv">1</span>),</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">       <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(sim<span class="op">$</span>gene_<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb3-12" data-line-number="12"></a>
<a class="sourceLine" id="cb3-13" data-line-number="13">og_plt</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Next we’ll run PCA on this simulated data using the <code>prcomp</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># run pca</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">pc &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(sim[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)], </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">             <span class="dt">center =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb4-4" data-line-number="4">             <span class="dt">scale. =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><code>prcomp</code> returns an object of class <code>prcomp</code> with a few slots. The <code>x</code> slot contains the principal components in a matrix.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(pc)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; [1] "sdev"     "rotation" "center"   "scale"    "x"</span></a></code></pre></div>
<p>For ease of plotting I will define a function to make a plot.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># function to make a pretty labeled plot</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">plot_labeled_pca &lt;-<span class="st"> </span><span class="cf">function</span>(input_data, </a>
<a class="sourceLine" id="cb6-4" data-line-number="4">                             pca_obj){</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">  </a>
<a class="sourceLine" id="cb6-6" data-line-number="6">  <span class="co">#tidy the pc matrix</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7">  pc_x &lt;-<span class="st"> </span>pca_obj<span class="op">$</span>x <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">id =</span> input_data<span class="op">$</span>id) </a>
<a class="sourceLine" id="cb6-10" data-line-number="10"> </a>
<a class="sourceLine" id="cb6-11" data-line-number="11">  <span class="co"># select points to label</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12">  max_min_points &lt;-<span class="st"> </span><span class="kw">slice</span>(pc_x, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">95</span><span class="op">:</span><span class="dv">100</span>))</a>
<a class="sourceLine" id="cb6-13" data-line-number="13"></a>
<a class="sourceLine" id="cb6-14" data-line-number="14">  <span class="co"># compute variance explained</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15">  percent_var &lt;-<span class="st"> </span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span>pca_obj<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="st">                    </span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(pca_obj<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>))  <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">signif</a></span>(<span class="dv">3</span>)</a>
<a class="sourceLine" id="cb6-18" data-line-number="18"></a>
<a class="sourceLine" id="cb6-19" data-line-number="19">  <span class="co">#plot the data</span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20">  pca_plt &lt;-<span class="st"> </span><span class="kw">ggplot</span>(pc_x, <span class="kw">aes</span>(PC1, PC2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-21" data-line-number="21"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb6-22" data-line-number="22"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>(<span class="dt">data =</span> max_min_points, </a>
<a class="sourceLine" id="cb6-23" data-line-number="23">                  <span class="kw">aes</span>(<span class="dt">label =</span> id)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-24" data-line-number="24"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="kw">str_c</span>(<span class="st">"PC1 "</span>, percent_var[<span class="dv">1</span>], <span class="st">"%"</span>),</a>
<a class="sourceLine" id="cb6-25" data-line-number="25">         <span class="dt">y =</span> <span class="kw">str_c</span>(<span class="st">"PC2 "</span>, percent_var[<span class="dv">2</span>], <span class="st">"%"</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-26" data-line-number="26"><span class="st">    </span><span class="co"># make axes symmetric</span></a>
<a class="sourceLine" id="cb6-27" data-line-number="27"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(pc_x<span class="op">$</span>PC1), </a>
<a class="sourceLine" id="cb6-28" data-line-number="28">           <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(pc_x<span class="op">$</span>PC1)))</a>
<a class="sourceLine" id="cb6-29" data-line-number="29">  </a>
<a class="sourceLine" id="cb6-30" data-line-number="30">  pca_plt </a>
<a class="sourceLine" id="cb6-31" data-line-number="31">}</a></code></pre></div>
<p>If we plot out the new variables PC1 and PC2 you can see that now the data has been flattened and spread out along the PC1 axis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">pca_plt &lt;-<span class="st"> </span><span class="kw">plot_labeled_pca</span>(sim, pc)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">pca_plt</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-5-1.png" width="480" style="display: block; margin: auto;"></p>
<p>As we can see, PCA has refactored the data such that PC1 now represents a vector through the data that captures the most variance. The second PC, now represents a distinct type of variance represented as variance perpendicular to the y = x line.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(og_plt, pca_plt)</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Another way to demonstrate what PCA is doing is to show the information captured by each PC. (see post <a href="https://stats.stackexchange.com/questions/229092/how-to-reverse-pca-and-reconstruct-original-variables-from-several-principal-com?noredirect=1&amp;lq=1">here</a>)</p>
<p><img src="img/pca.png" width="700px" style="display: block; margin: auto;"></p>
</div>
<div id="example-with-more-dimensions" class="section level3">
<h3 class="hasAnchor">
<a href="#example-with-more-dimensions" class="anchor"></a>Example with more dimensions</h3>
<p>What if we had more dimensions that are just noise? How does this impact PCA?</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># generate a matrix of random numbers </span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co"># drawn from a normal dist. to simulate noise</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">20181214</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">more_genes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb9-5" data-line-number="5">                           <span class="dt">mean =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb9-6" data-line-number="6">                           <span class="dt">sd =</span> <span class="fl">0.25</span>), </a>
<a class="sourceLine" id="cb9-7" data-line-number="7">                     <span class="dt">nrow =</span> <span class="dv">100</span>, </a>
<a class="sourceLine" id="cb9-8" data-line-number="8">                     <span class="dt">ncol =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb9-9" data-line-number="9"></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co"># add column names</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(more_genes) &lt;-<span class="st"> </span><span class="kw">str_c</span>(<span class="st">"gene_noise_"</span>,</a>
<a class="sourceLine" id="cb9-12" data-line-number="12">                              <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(more_genes))</a>
<a class="sourceLine" id="cb9-13" data-line-number="13"></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co"># add the matrix to the simulation data_frame by column</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15">sim_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(sim, more_genes) </a>
<a class="sourceLine" id="cb9-16" data-line-number="16"></a>
<a class="sourceLine" id="cb9-17" data-line-number="17">sim_<span class="dv">2</span>[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"><span class="co">#&gt;   id    gene_1    gene_2 gene_noise_1 gene_noise_2 gene_noise_3</span></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="co">#&gt; 1  1 -2.995217 -2.123704  0.089215287  -0.40819808  -0.48037882</span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="co">#&gt; 2  2 -2.821900 -2.933205 -0.339377410  -0.03582506   0.41886793</span></a>
<a class="sourceLine" id="cb9-21" data-line-number="21"><span class="co">#&gt; 3  3 -2.278693 -2.375272  0.131539067  -0.42624492   0.28643147</span></a>
<a class="sourceLine" id="cb9-22" data-line-number="22"><span class="co">#&gt; 4  4 -2.167279 -2.535480  0.098178061  -0.20509451  -0.06692551</span></a>
<a class="sourceLine" id="cb9-23" data-line-number="23"><span class="co">#&gt; 5  5 -1.528019 -1.310610  0.007170167   0.04159706   0.03533156</span></a></code></pre></div>
<p>The <code><a href="https://rdrr.io/r/graphics/plot.html">plot()</a></code> function can be used to quickly generate many exploratory plots. If you pass two vectors of the same length, you will get a scatterplot.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(sim_<span class="dv">2</span><span class="op">$</span>gene_<span class="dv">1</span>, sim_<span class="dv">2</span><span class="op">$</span>gene_<span class="dv">2</span>)</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-9-1.png" width="480" style="display: block; margin: auto;"></p>
<p>However, if you supply <code>plot</code> with a data.frame with only numeric values, it will plot all of the cross comparisons of the columns, which is useful in this example to quickly highlight that <code>gene_1</code> and <code>gene_2</code> are the only correlated or interesting variables.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(sim_<span class="dv">2</span>[, <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>])</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Now if we had 2000 dimensions (i.e. genes) we couldn’t resonably investigate all of the scatterplots for each gene compared to each other gene. Even with a small real dataset with 10 genes, it’s hard to identify important relationships.</p>
<p>Next we’ll run PCA on this new data</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">pc2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(sim_<span class="dv">2</span>[, <span class="dv">2</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(sim_<span class="dv">2</span>)], </a>
<a class="sourceLine" id="cb12-2" data-line-number="2">              <span class="dt">center =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb12-3" data-line-number="3">              <span class="dt">scale. =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">pc2_plt &lt;-<span class="st"> </span><span class="kw">plot_labeled_pca</span>(sim_<span class="dv">2</span>, pc2) </a>
<a class="sourceLine" id="cb13-2" data-line-number="2"></a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(og_plt, pc2_plt)</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="how-much-variation-is-captured-by-each-pc" class="section level3">
<h3 class="hasAnchor">
<a href="#how-much-variation-is-captured-by-each-pc" class="anchor"></a>How much variation is captured by each PC?</h3>
<p>The <code>sdev</code> slot contains the standard deviation of the principal components. The square of this value is the variance explained by the PC.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">sds &lt;-<span class="st"> </span>pc2<span class="op">$</span>sdev</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="co"># get variance per pc</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4">var_per_pc &lt;-<span class="st"> </span>sds<span class="op">^</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"></a>
<a class="sourceLine" id="cb14-6" data-line-number="6"><span class="co">#calculate percent variance per pc</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7">percent_var_per_pc &lt;-<span class="st"> </span><span class="dv">100</span> <span class="op">*</span><span class="st"> </span>(var_per_pc <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(var_per_pc))</a>
<a class="sourceLine" id="cb14-8" data-line-number="8"></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="co">#plot a vector</span></a>
<a class="sourceLine" id="cb14-10" data-line-number="10"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(percent_var_per_pc)</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-13-1.png" width="480" style="display: block; margin: auto;"></p>
<p>As we can see most of the variation in the data is captured in the first PC. This means that if we dropped the remaining PCs from the data we would lose very little information. It is for this reason that PCA is so useful for dimensionality reduction for big datasets like single cell RNA-seq.</p>
</div>
<div id="how-to-find-features-genes-associated-with-each-pc" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-find-features-genes-associated-with-each-pc" class="anchor"></a>How to find features (genes) associated with each PC</h3>
<p>As we have learned each PC is linear combination of the features (i.e genes) in the data. How do we determine which features are most important for each PC?</p>
<p>The <code>rotation</code> slot of the <code>prcomp</code> object contains the rotations (also known as loadings) for each PC. These loadings can be interpreted as the amount each feature contributes to a principle component. Strongly positive loadings mean that the feature is strongly positivlye correlated with the PC. Strongly negative loadings indicate that it is strongly negatively correlated. Loadings at zero indicate no contribution.</p>
<p>By ranking the loadings we can identify which genes impact each PC.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">pc2<span class="op">$</span>rotation[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="co">#&gt;                         PC1         PC2         PC3</span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="co">#&gt; gene_1         0.7053063021 -0.04871949  0.01920815</span></a>
<a class="sourceLine" id="cb15-4" data-line-number="4"><span class="co">#&gt; gene_2         0.7053462316  0.03364238 -0.02973091</span></a>
<a class="sourceLine" id="cb15-5" data-line-number="5"><span class="co">#&gt; gene_noise_1   0.0230621826  0.02051417 -0.08251792</span></a>
<a class="sourceLine" id="cb15-6" data-line-number="6"><span class="co">#&gt; gene_noise_2   0.0319279005  0.21472325  0.07504765</span></a>
<a class="sourceLine" id="cb15-7" data-line-number="7"><span class="co">#&gt; gene_noise_3  -0.0143038106 -0.11483800 -0.03152142</span></a>
<a class="sourceLine" id="cb15-8" data-line-number="8"><span class="co">#&gt; gene_noise_4  -0.0236241033  0.22100695  0.02117113</span></a>
<a class="sourceLine" id="cb15-9" data-line-number="9"><span class="co">#&gt; gene_noise_5  -0.0246378278 -0.18849671  0.63608275</span></a>
<a class="sourceLine" id="cb15-10" data-line-number="10"><span class="co">#&gt; gene_noise_6  -0.0037009439 -0.05526442  0.16294670</span></a>
<a class="sourceLine" id="cb15-11" data-line-number="11"><span class="co">#&gt; gene_noise_7  -0.0268014930  0.05081115 -0.69522057</span></a>
<a class="sourceLine" id="cb15-12" data-line-number="12"><span class="co">#&gt; gene_noise_8   0.0368993691  0.05205190  0.13542228</span></a>
<a class="sourceLine" id="cb15-13" data-line-number="13"><span class="co">#&gt; gene_noise_9   0.0002263975 -0.81355912 -0.21193486</span></a>
<a class="sourceLine" id="cb15-14" data-line-number="14"><span class="co">#&gt; gene_noise_10 -0.0038867809 -0.42682676  0.08440284</span></a></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co"># for PC1</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">pc2<span class="op">$</span>rotation[, <span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="dv">3</span>)</a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="co">#&gt;       gene_2       gene_1 gene_noise_8 </span></a>
<a class="sourceLine" id="cb16-6" data-line-number="6"><span class="co">#&gt;   0.70534623   0.70530630   0.03689937</span></a></code></pre></div>
<p>Sidebar: sorting vectors using <code><a href="https://rdrr.io/r/base/order.html">order()</a></code></p>
<p>order return the sort order of a vector.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="co">#&gt; [1] 1 2 3 5</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co">#&gt; [1] 3 1 4 2</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">3</span>), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="co">#&gt; [1] 2 4 1 3</span></a></code></pre></div>
<p>Useful shortcut to get vector sorted on the absolute values:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">vec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">-9</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">-3</span>, <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"></a>
<a class="sourceLine" id="cb18-3" data-line-number="3">vec[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(vec), </a>
<a class="sourceLine" id="cb18-4" data-line-number="4">          <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)]</a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="co">#&gt; [1] -10  10  -9   5   5  -3   2   0</span></a></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">pc1 &lt;-<span class="st"> </span>pc2<span class="op">$</span>rotation[, <span class="dv">1</span>] </a>
<a class="sourceLine" id="cb19-2" data-line-number="2"></a>
<a class="sourceLine" id="cb19-3" data-line-number="3">pc1[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(pc1), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)] <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>()</a>
<a class="sourceLine" id="cb19-5" data-line-number="5"><span class="co">#&gt;       gene_2       gene_1 gene_noise_8 gene_noise_2 gene_noise_7 gene_noise_5 </span></a>
<a class="sourceLine" id="cb19-6" data-line-number="6"><span class="co">#&gt;   0.70534623   0.70530630   0.03689937   0.03192790  -0.02680149  -0.02463783</span></a></code></pre></div>
<p>Using this approach we can quickly identify that <code>gene_1</code> and <code>gene_2</code> are driving the variance in the data.</p>
</div>
</div>
<div id="run-pca-on-a-single-cell-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#run-pca-on-a-single-cell-dataset" class="anchor"></a>Run PCA on a single cell dataset</h2>
<div id="step-1-first-lets-filter-and-normalize-the-esc_mat-matrix-" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-first-lets-filter-and-normalize-the-esc_mat-matrix-" class="anchor"></a>Step 1: First let’s filter and normalize the <code>esc_mat</code> matrix.</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># remove lowly expressed features (genes with non-zero values in at least 10 samples)</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">filtered_mat &lt;-<span class="st"> </span>esc_mat[<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(esc_mat <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</a>
<a class="sourceLine" id="cb20-3" data-line-number="3"></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="co"># log normalize </span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5">norm_mat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(filtered_mat, <span class="dv">2</span>, <span class="cf">function</span>(x) x <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(x))</a>
<a class="sourceLine" id="cb20-6" data-line-number="6">norm_mat &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="op">*</span><span class="st"> </span>norm_mat</a>
<a class="sourceLine" id="cb20-7" data-line-number="7">norm_mat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(norm_mat <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a></code></pre></div>
</div>
<div id="step-2-scale-the-data-" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-scale-the-data-" class="anchor"></a>Step 2: Scale the data.</h3>
<p>This makes every variable equal in weight. For example if one variable ranges from -100 to 100, this variable will have more weight than one that ranges from -1 to 1. Often in most datasets scaling the variables is necessary to avoid projecting the PCA solely based on the abundance of a variable.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># scale and center each gene</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">zmat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/scale.html">scale</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(norm_mat)))</a></code></pre></div>
</div>
<div id="step-3-run-pca" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-run-pca" class="anchor"></a>Step 3: run PCA</h3>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># run PCA with genes as columns and cells as rows</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2">pc &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(zmat), </a>
<a class="sourceLine" id="cb22-3" data-line-number="3">             <span class="dt">center =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb22-4" data-line-number="4">             <span class="dt">scale. =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="step-4-examine-variance-explained-by-each-pc" class="section level3">
<h3 class="hasAnchor">
<a href="#step-4-examine-variance-explained-by-each-pc" class="anchor"></a>Step 4: Examine variance explained by each PC</h3>
<p>Exercise: make a ggplot plot that plots PC versus percent variance explained</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">pc_var &lt;-<span class="st"> </span>pc<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2">pc_var &lt;-<span class="st"> </span><span class="dv">100</span> <span class="op">*</span><span class="st"> </span>(pc_var <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(pc_var))</a>
<a class="sourceLine" id="cb23-3" data-line-number="3"></a>
<a class="sourceLine" id="cb23-4" data-line-number="4">plt_dat &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb23-5" data-line-number="5">  <span class="dt">var_explained =</span> pc_var,</a>
<a class="sourceLine" id="cb23-6" data-line-number="6">  <span class="dt">PC =</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(pc_var)</a>
<a class="sourceLine" id="cb23-7" data-line-number="7">)</a>
<a class="sourceLine" id="cb23-8" data-line-number="8"></a>
<a class="sourceLine" id="cb23-9" data-line-number="9"><span class="kw">ggplot</span>(plt_dat, <span class="kw">aes</span>(<span class="dt">x =</span> PC, </a>
<a class="sourceLine" id="cb23-10" data-line-number="10">                    <span class="dt">y =</span> var_explained)) <span class="op">+</span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="step-5-plot-pcs" class="section level3">
<h3 class="hasAnchor">
<a href="#step-5-plot-pcs" class="anchor"></a>Step 5: Plot PCs</h3>
<p>Exercise: Generate a ggplot plot that plots PC1 versus PC2 and color by developmental stage. (i.e. zy, 4cell, 8cell, etc.)</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">pc_mat &lt;-<span class="st"> </span>pc<span class="op">$</span>x</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggrepel)</a>
<a class="sourceLine" id="cb24-4" data-line-number="4"></a>
<a class="sourceLine" id="cb24-5" data-line-number="5">pc_mat <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb24-7" data-line-number="7"><span class="st">  </span><span class="kw">as_tibble</span>(<span class="dt">rownames =</span> <span class="st">"sample"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(PC1, PC2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb24-10" data-line-number="10"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>(<span class="kw">aes</span>(<span class="dt">label =</span> sample))</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-23-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"></a>
<a class="sourceLine" id="cb25-3" data-line-number="3">pc_mat <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-5" data-line-number="5"><span class="st">  </span><span class="kw">as_tibble</span>(<span class="dt">rownames =</span> <span class="st">"sample"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-6" data-line-number="6"><span class="st">  </span><span class="kw">separate</span>(sample, </a>
<a class="sourceLine" id="cb25-7" data-line-number="7">           <span class="dt">into =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"time"</span>, <span class="st">"rep"</span>),</a>
<a class="sourceLine" id="cb25-8" data-line-number="8">           <span class="dt">sep =</span> <span class="st">"_"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-9" data-line-number="9"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(PC1, PC2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb25-10" data-line-number="10"><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color =</span> time)) </a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-23-2.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="kmeans-cluster-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#kmeans-cluster-the-data" class="anchor"></a>kmeans cluster the data</h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">42</span>)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2">km &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(pc_mat, <span class="dt">centers =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3">km_clusters &lt;-<span class="st"> </span>km<span class="op">$</span>cluster</a>
<a class="sourceLine" id="cb26-4" data-line-number="4"></a>
<a class="sourceLine" id="cb26-5" data-line-number="5">tidy_clusters &lt;-<span class="st"> </span><span class="kw">tibble</span>(</a>
<a class="sourceLine" id="cb26-6" data-line-number="6">  <span class="dt">sample =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(km_clusters),</a>
<a class="sourceLine" id="cb26-7" data-line-number="7">  <span class="dt">cluster =</span> km_clusters</a>
<a class="sourceLine" id="cb26-8" data-line-number="8">)</a>
<a class="sourceLine" id="cb26-9" data-line-number="9">tidy_clusters</a>
<a class="sourceLine" id="cb26-10" data-line-number="10"><span class="co">#&gt; # A tibble: 72 x 2</span></a>
<a class="sourceLine" id="cb26-11" data-line-number="11"><span class="co">#&gt;    sample       cluster</span></a>
<a class="sourceLine" id="cb26-12" data-line-number="12"><span class="co">#&gt;    &lt;chr&gt;          &lt;int&gt;</span></a>
<a class="sourceLine" id="cb26-13" data-line-number="13"><span class="co">#&gt;  1 zy_1               4</span></a>
<a class="sourceLine" id="cb26-14" data-line-number="14"><span class="co">#&gt;  2 zy_2               4</span></a>
<a class="sourceLine" id="cb26-15" data-line-number="15"><span class="co">#&gt;  3 zy_3               4</span></a>
<a class="sourceLine" id="cb26-16" data-line-number="16"><span class="co">#&gt;  4 zy_4               4</span></a>
<a class="sourceLine" id="cb26-17" data-line-number="17"><span class="co">#&gt;  5 early2cell_1       4</span></a>
<a class="sourceLine" id="cb26-18" data-line-number="18"><span class="co">#&gt;  6 early2cell_2       4</span></a>
<a class="sourceLine" id="cb26-19" data-line-number="19"><span class="co">#&gt;  7 early2cell_3       4</span></a>
<a class="sourceLine" id="cb26-20" data-line-number="20"><span class="co">#&gt;  8 early2cell_4       4</span></a>
<a class="sourceLine" id="cb26-21" data-line-number="21"><span class="co">#&gt;  9 early2cell_5       4</span></a>
<a class="sourceLine" id="cb26-22" data-line-number="22"><span class="co">#&gt; 10 early2cell_6       4</span></a>
<a class="sourceLine" id="cb26-23" data-line-number="23"><span class="co">#&gt; # … with 62 more rows</span></a>
<a class="sourceLine" id="cb26-24" data-line-number="24"></a>
<a class="sourceLine" id="cb26-25" data-line-number="25">plt_dat &lt;-<span class="st"> </span>pc_mat <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-26" data-line-number="26"><span class="st">  </span><span class="kw">as_tibble</span>(<span class="dt">rownames =</span> <span class="st">"sample"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-27" data-line-number="27"><span class="st">  </span><span class="kw">left_join</span>(tidy_clusters, <span class="dt">by =</span> <span class="st">"sample"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-28" data-line-number="28"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">cluster =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(cluster))</a>
<a class="sourceLine" id="cb26-29" data-line-number="29"></a>
<a class="sourceLine" id="cb26-30" data-line-number="30"></a>
<a class="sourceLine" id="cb26-31" data-line-number="31"><span class="kw">ggplot</span>(plt_dat, <span class="kw">aes</span>(PC1, PC2)) <span class="op">+</span></a>
<a class="sourceLine" id="cb26-32" data-line-number="32"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color =</span> cluster))</a></code></pre></div>
<p><img src="class-9_files/figure-html/unnamed-chunk-24-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="step-6-find-genes-that-are-associated-with-each-pc" class="section level3">
<h3 class="hasAnchor">
<a href="#step-6-find-genes-that-are-associated-with-each-pc" class="anchor"></a>Step 6: Find genes that are associated with each PC</h3>
<p>For example we can see that PC1 separates that data along the developmental time. If we find genes with strongly positive loadings these should be genes that increase over time.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co"># find the top 25 genes that have the most positive and the most negative loadings for PC1</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"></a>
<a class="sourceLine" id="cb27-3" data-line-number="3">pc1 &lt;-<span class="st"> </span>pc<span class="op">$</span>rotation[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb27-4" data-line-number="4"></a>
<a class="sourceLine" id="cb27-5" data-line-number="5">top25 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(pc1,</a>
<a class="sourceLine" id="cb27-6" data-line-number="6">              <span class="dt">decreasing =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="dv">25</span>)</a>
<a class="sourceLine" id="cb27-7" data-line-number="7">bottom25 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(pc1) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="dv">25</span>)</a>
<a class="sourceLine" id="cb27-8" data-line-number="8"></a>
<a class="sourceLine" id="cb27-9" data-line-number="9">top_genes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(top25, bottom25)</a></code></pre></div>
<p>Exercise: Make a heatmap that plots the pc1_genes</p>
<p><img src="class-9_files/figure-html/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="additional-exercise" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-exercise" class="anchor"></a>Additional exercise</h2>
<p>There is another dataset called <code>tx_rates</code> in the <code>pbda</code> package. This dataset contains transcription rates of mouse dendritic cells after exposure to LPS. Run PCA on this dataset to investigate the relationships between the different sampled time points. First remove genes (rows) with all zero values and remove rows that have zero variance (<code><a href="https://rdrr.io/pkg/matrixStats/man/rowVars.html">matrixStats::rowVars</a></code>) then log transform the data prior to scaling.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#pca">PCA</a></li>
      <li><a href="#run-pca-on-a-single-cell-dataset">Run PCA on a single cell dataset</a></li>
      <li><a href="#additional-exercise">Additional exercise</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy, Rui Fu, Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
