<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 5: Working with gene lists • pbda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Class 5: Working with gene lists">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pbda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Exploring a biological data set</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: RNA-seq Analysis</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
    <li>
      <a href="../articles/problem-set-1.html">Problem Set 1</a>
    </li>
    <li>
      <a href="../articles/problem-set-2.html">Problem Set 2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Class 5: Working with gene lists</h1>
                        <h4 class="author">Rui Fu</h4>
            
            <h4 class="date">2018-12-06</h4>
      
      
      <div class="hidden name"><code>class-5.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># devtools::install_github('idpt7810/practical-data-analysis') run this before class</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(pbda)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)</a></code></pre></div>
<div id="did-experiment-ran-through-analysis-pipeline-what-now" class="section level2">
<h2 class="hasAnchor">
<a href="#did-experiment-ran-through-analysis-pipeline-what-now" class="anchor"></a>Did experiment, ran through analysis pipeline, what now?</h2>
<p>Handle gene lists</p>
<p>Try to connect prior knowledge to the results</p>
<p>Recall some basic R and dataframe manipulation commands</p>
</div>
<div id="retrieve-gene-lists-from-rna-seq-or-scrna-seq" class="section level2">
<h2 class="hasAnchor">
<a href="#retrieve-gene-lists-from-rna-seq-or-scrna-seq" class="anchor"></a>Retrieve gene lists from RNA-seq or scRNA-seq</h2>
<p>Recall how data is saved in objects</p>
<p>Or whatever file you already saved containing lists of gene names</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># from DEseq output, get a tibble back</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">loc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"res_class4.rds"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">res &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(loc) <span class="co"># loading a DESeq2 object automatically loads DESeq2</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(res)</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt; [1] "DESeqResults"</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; attr(,"package")</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; [1] "DESeq2"</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(res)</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt; log2 fold change (MLE): drug none vs A </span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt; Wald test p-value: drug none vs A </span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt; DataFrame with 6 rows and 6 columns</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt;                       baseMean     log2FoldChange             lfcSE</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt;                      &lt;numeric&gt;          &lt;numeric&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt; WASH7P        10.0030522829909 -0.121047340632252 0.697032782022802</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt; MIR6859-1     1.70729126552468 -0.496551060380745  1.50712758679313</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt; LOC101927589 0.237586926465382  -1.61061682077907  4.37050557280351</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt; LOC729737     164.730907819854  -1.56620710989126   0.2640457959487</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt; LOC100996442 0.456230877334472  -1.57579090799684  2.97748794222305</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt; LOC102723897  139.842022077039 -0.373915799642309 0.214064711031773</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt;                            stat               pvalue                 padj</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt;                       &lt;numeric&gt;            &lt;numeric&gt;            &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt; WASH7P       -0.173660900540389    0.862131967385253    0.922817938233145</span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co">#&gt; MIR6859-1    -0.329468496716531    0.741801601921924    0.844449335816827</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#&gt; LOC101927589 -0.368519566889816    0.712485857892975                   NA</span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26"><span class="co">#&gt; LOC729737     -5.93157374183509 3.00044749513025e-09 5.67484419528589e-08</span></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co">#&gt; LOC100996442 -0.529235025825269    0.596642421979017                   NA</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="co">#&gt; LOC102723897  -1.74674189799929   0.0806821191329157    0.181200019774001</span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"></a>
<a class="sourceLine" id="cb2-30" data-line-number="30">res_tibble &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(res) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31"><span class="st">  </span>tibble<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/tibble/topics/rownames">rownames_to_column</a></span>(<span class="dt">var =</span> <span class="st">"gene"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32"><span class="st">  </span><span class="kw">as.tibble</span>()</a>
<a class="sourceLine" id="cb2-33" data-line-number="33">res_tibble</a>
<a class="sourceLine" id="cb2-34" data-line-number="34"><span class="co">#&gt; # A tibble: 21,178 x 7</span></a>
<a class="sourceLine" id="cb2-35" data-line-number="35"><span class="co">#&gt;    gene        baseMean log2FoldChange lfcSE   stat     pvalue        padj</span></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="co">#&gt;    &lt;chr&gt;          &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;      &lt;dbl&gt;       &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb2-37" data-line-number="37"><span class="co">#&gt;  1 WASH7P        10.0           -0.121 0.697 -0.174    8.62e-1     9.23e-1</span></a>
<a class="sourceLine" id="cb2-38" data-line-number="38"><span class="co">#&gt;  2 MIR6859-1      1.71          -0.497 1.51  -0.329    7.42e-1     8.44e-1</span></a>
<a class="sourceLine" id="cb2-39" data-line-number="39"><span class="co">#&gt;  3 LOC1019275…    0.238         -1.61  4.37  -0.369    7.12e-1    NA      </span></a>
<a class="sourceLine" id="cb2-40" data-line-number="40"><span class="co">#&gt;  4 LOC729737    165.            -1.57  0.264 -5.93     3.00e-9     5.67e-8</span></a>
<a class="sourceLine" id="cb2-41" data-line-number="41"><span class="co">#&gt;  5 LOC1009964…    0.456         -1.58  2.98  -0.529    5.97e-1    NA      </span></a>
<a class="sourceLine" id="cb2-42" data-line-number="42"><span class="co">#&gt;  6 LOC1027238…  140.            -0.374 0.214 -1.75     8.07e-2     1.81e-1</span></a>
<a class="sourceLine" id="cb2-43" data-line-number="43"><span class="co">#&gt;  7 LOC1019291…    0.451         -1.58  3.00  -0.525    5.99e-1    NA      </span></a>
<a class="sourceLine" id="cb2-44" data-line-number="44"><span class="co">#&gt;  8 RP4-669L17…    1.62          -1.13  1.41  -0.801    4.23e-1     5.94e-1</span></a>
<a class="sourceLine" id="cb2-45" data-line-number="45"><span class="co">#&gt;  9 LOC1001348…    8.57           0.328 0.723  0.454    6.50e-1     7.80e-1</span></a>
<a class="sourceLine" id="cb2-46" data-line-number="46"><span class="co">#&gt; 10 MIR6723        0.616         -0.736 3.22  -0.229    8.19e-1    NA      </span></a>
<a class="sourceLine" id="cb2-47" data-line-number="47"><span class="co">#&gt; # ... with 21,168 more rows</span></a>
<a class="sourceLine" id="cb2-48" data-line-number="48"></a>
<a class="sourceLine" id="cb2-49" data-line-number="49"><span class="co"># res_tibble contains calculations for all genes, pick out "interesting" genes</span></a>
<a class="sourceLine" id="cb2-50" data-line-number="50">gene_tibble &lt;-<span class="st"> </span>res_tibble <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-51" data-line-number="51"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(padj) <span class="op">&amp;</span><span class="st"> </span>padj <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.01</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># p &lt;= 0.05 is common, but can be lowered, sometimes even 0.1 is acceptable</span></a>
<a class="sourceLine" id="cb2-52" data-line-number="52"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(log2FoldChange <span class="op">&gt;=</span><span class="st"> </span><span class="dv">1</span> <span class="op">|</span><span class="st"> </span>log2FoldChange <span class="op">&lt;=</span><span class="st"> </span><span class="dv">-1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-53" data-line-number="53"><span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(log2FoldChange))</a>
<a class="sourceLine" id="cb2-54" data-line-number="54">gene_tibble</a>
<a class="sourceLine" id="cb2-55" data-line-number="55"><span class="co">#&gt; # A tibble: 1,908 x 7</span></a>
<a class="sourceLine" id="cb2-56" data-line-number="56"><span class="co">#&gt;    gene         baseMean log2FoldChange lfcSE  stat    pvalue      padj</span></a>
<a class="sourceLine" id="cb2-57" data-line-number="57"><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb2-58" data-line-number="58"><span class="co">#&gt;  1 MAGEC2          133.           11.2  1.20   9.33 1.06e- 20 9.57e- 19</span></a>
<a class="sourceLine" id="cb2-59" data-line-number="59"><span class="co">#&gt;  2 PAGE5            76.5          10.7  1.20   8.90 5.57e- 19 4.23e- 17</span></a>
<a class="sourceLine" id="cb2-60" data-line-number="60"><span class="co">#&gt;  3 PGR            1732.            9.44 0.316 29.9  9.03e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb2-61" data-line-number="61"><span class="co">#&gt;  4 FAM133A          27.2           9.24 1.24   7.44 1.01e- 13 4.05e- 12</span></a>
<a class="sourceLine" id="cb2-62" data-line-number="62"><span class="co">#&gt;  5 ZFP3             61.4           9.15 1.21   7.54 4.64e- 14 1.96e- 12</span></a>
<a class="sourceLine" id="cb2-63" data-line-number="63"><span class="co">#&gt;  6 GREB1          2343.            9.08 0.303 29.9  8.27e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb2-64" data-line-number="64"><span class="co">#&gt;  7 LOC102724756     18.0           8.66 1.29   6.69 2.25e- 11 5.99e- 10</span></a>
<a class="sourceLine" id="cb2-65" data-line-number="65"><span class="co">#&gt;  8 CYP2B7P        1839.            8.58 0.336 25.5  1.94e-143 8.47e-140</span></a>
<a class="sourceLine" id="cb2-66" data-line-number="66"><span class="co">#&gt;  9 CTAG2           153.            8.31 0.675 12.3  7.76e- 35 2.46e- 32</span></a>
<a class="sourceLine" id="cb2-67" data-line-number="67"><span class="co">#&gt; 10 TMLHE           188.            8.25 1.05   7.88 3.34e- 15 1.62e- 13</span></a>
<a class="sourceLine" id="cb2-68" data-line-number="68"><span class="co">#&gt; # ... with 1,898 more rows</span></a>
<a class="sourceLine" id="cb2-69" data-line-number="69"></a>
<a class="sourceLine" id="cb2-70" data-line-number="70"><span class="co"># get a vector of gene names</span></a>
<a class="sourceLine" id="cb2-71" data-line-number="71">gene_vector &lt;-<span class="st"> </span>gene_tibble<span class="op">$</span>gene</a>
<a class="sourceLine" id="cb2-72" data-line-number="72">gene_vector &lt;-<span class="st"> </span>gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(gene) <span class="co"># at least 5 more ways to do this</span></a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># or load already saved file on disk</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">loc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"gene_tibble.csv"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">gene_tibble2 &lt;-<span class="st"> </span><span class="kw">read_csv</span>(loc) <span class="co"># reads out as tibble</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">write_csv</span>(gene_tibble, <span class="st">"gene_tibble.csv"</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co"># or write as txt file, without col name</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="kw">write_csv</span>(gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(gene), <span class="st">"gene_tibble.txt"</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>) <span class="co"># write from tibble</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="kw">write_lines</span>(gene_vector, <span class="st">"gene_tibble_v.txt"</span>) <span class="co"># write for vector</span></a></code></pre></div>
</div>
<div id="feed-gene-lists-to-other-software-outside-of-r-get-enrichment-score-back" class="section level2">
<h2 class="hasAnchor">
<a href="#feed-gene-lists-to-other-software-outside-of-r-get-enrichment-score-back" class="anchor"></a>Feed gene lists to other software outside of R, get enrichment score back</h2>
<p>DAVID <a href="https://david.ncifcrf.gov/" class="uri">https://david.ncifcrf.gov/</a></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># save output and load in R</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># for macs: curl -o david.txt -k https://david.ncifcrf.gov/data/download/chart_F3B3023831201544032516222.txt</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">loc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"david.txt"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">david &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(loc)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co"># View(david)</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co"># do whatever you want, for instance plot fold change of all genes in row 1</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8">g1 &lt;-<span class="st"> </span>david<span class="op">$</span>Genes[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">g1_split &lt;-<span class="st"> </span><span class="kw">str_split</span>(g1, <span class="st">", "</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(cowplot)</a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="kw">ggplot</span>(gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(gene <span class="op">%in%</span><span class="st"> </span>g1_split), </a>
<a class="sourceLine" id="cb4-12" data-line-number="12">       <span class="kw">aes</span>(<span class="dt">x =</span> gene, <span class="dt">y =</span> log2FoldChange, <span class="dt">fill =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(padj))) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> david<span class="op">$</span>Term[<span class="dv">1</span>]) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="st">  </span><span class="kw">scale_fill_gradient2</span>(<span class="dt">low =</span> <span class="st">"white"</span>, <span class="dt">high =</span> <span class="st">"black"</span>)</a></code></pre></div>
<p><img src="img/class-5-david-1.png" width="700" style="display: block; margin: auto;"></p>
<p>other options:</p>
<p>Panther <a href="http://www.geneontology.org/page/go-enrichment-analysis" class="uri">http://www.geneontology.org/page/go-enrichment-analysis</a></p>
<p>GSEA (requires ranked list, for instance by fold-change, and requires full list of genes instead of just “significant”) <a href="http://software.broadinstitute.org/gsea/index.jsp" class="uri">http://software.broadinstitute.org/gsea/index.jsp</a></p>
<p>or, GO and GSEA right in R (will require additional packages):</p>
<p>topGO: <a href="https://bioconductor.org/packages/release/bioc/html/topGO.html" class="uri">https://bioconductor.org/packages/release/bioc/html/topGO.html</a></p>
<p>fgsea: <a href="http://bioconductor.org/packages/release/bioc/html/fgsea.html" class="uri">http://bioconductor.org/packages/release/bioc/html/fgsea.html</a></p>
</div>
<div id="sometimes-transformation-of-names-is-needed---stringr-package-part-of-tidyverse" class="section level2">
<h2 class="hasAnchor">
<a href="#sometimes-transformation-of-names-is-needed---stringr-package-part-of-tidyverse" class="anchor"></a>Sometimes transformation of names is needed - <code>stringr</code> package, part of tidyverse</h2>
<p>Depending on the transcriptome and analysis options used in RNA-seq/scRNA-seq, the list of names can be messy.</p>
<p>For instance, trimming of gene/transcript version.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">loc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ensg_tibble.csv"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">ensg_tibble &lt;-<span class="st"> </span><span class="kw">read_csv</span>(loc)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">ensg_tibble</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; # A tibble: 5 x 2</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt;   gene                padj</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt;   &lt;chr&gt;              &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; 1 ENSG00000238164.4 0.0001</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt; 2 ENSG00000049249.6 0.0002</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; 3 ENSG00000226823.1 0.0001</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt; 4 ENSG00000266470.1 0.0002</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt; 5 ENSG00000229435.2 0.05</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"></a>
<a class="sourceLine" id="cb5-13" data-line-number="13">ensg_tibble2 &lt;-<span class="st"> </span>ensg_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">str_replace</span>(gene, <span class="dt">pattern =</span> <span class="st">"</span><span class="ch">\\</span><span class="st">..*"</span>, <span class="dt">replacement =</span> <span class="st">""</span>)) <span class="co"># regular expressions are useful and worth the time to learn, see ?base::regex</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14">ensg_tibble2</a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; # A tibble: 5 x 2</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt;   gene              padj</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; 1 ENSG00000238164 0.0001</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; 2 ENSG00000049249 0.0002</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; 3 ENSG00000226823 0.0001</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; 4 ENSG00000266470 0.0002</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; 5 ENSG00000229435 0.05</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co"># some other simple and useful functions</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25">gene_tibble_l &lt;-<span class="st"> </span>gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">str_to_lower</span>(gene))</a>
<a class="sourceLine" id="cb5-26" data-line-number="26">gene_tibble_l</a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; # A tibble: 1,908 x 7</span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"><span class="co">#&gt;    gene         baseMean log2FoldChange lfcSE  stat    pvalue      padj</span></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-30" data-line-number="30"><span class="co">#&gt;  1 magec2          133.           11.2  1.20   9.33 1.06e- 20 9.57e- 19</span></a>
<a class="sourceLine" id="cb5-31" data-line-number="31"><span class="co">#&gt;  2 page5            76.5          10.7  1.20   8.90 5.57e- 19 4.23e- 17</span></a>
<a class="sourceLine" id="cb5-32" data-line-number="32"><span class="co">#&gt;  3 pgr            1732.            9.44 0.316 29.9  9.03e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb5-33" data-line-number="33"><span class="co">#&gt;  4 fam133a          27.2           9.24 1.24   7.44 1.01e- 13 4.05e- 12</span></a>
<a class="sourceLine" id="cb5-34" data-line-number="34"><span class="co">#&gt;  5 zfp3             61.4           9.15 1.21   7.54 4.64e- 14 1.96e- 12</span></a>
<a class="sourceLine" id="cb5-35" data-line-number="35"><span class="co">#&gt;  6 greb1          2343.            9.08 0.303 29.9  8.27e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb5-36" data-line-number="36"><span class="co">#&gt;  7 loc102724756     18.0           8.66 1.29   6.69 2.25e- 11 5.99e- 10</span></a>
<a class="sourceLine" id="cb5-37" data-line-number="37"><span class="co">#&gt;  8 cyp2b7p        1839.            8.58 0.336 25.5  1.94e-143 8.47e-140</span></a>
<a class="sourceLine" id="cb5-38" data-line-number="38"><span class="co">#&gt;  9 ctag2           153.            8.31 0.675 12.3  7.76e- 35 2.46e- 32</span></a>
<a class="sourceLine" id="cb5-39" data-line-number="39"><span class="co">#&gt; 10 tmlhe           188.            8.25 1.05   7.88 3.34e- 15 1.62e- 13</span></a>
<a class="sourceLine" id="cb5-40" data-line-number="40"><span class="co">#&gt; # ... with 1,898 more rows</span></a>
<a class="sourceLine" id="cb5-41" data-line-number="41"></a>
<a class="sourceLine" id="cb5-42" data-line-number="42">gene_tibble_u &lt;-<span class="st"> </span>gene_tibble_l <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">str_to_upper</span>(gene))</a>
<a class="sourceLine" id="cb5-43" data-line-number="43">gene_tibble_u <span class="co"># note that this is not ideal for most biology situations, see below for better way</span></a>
<a class="sourceLine" id="cb5-44" data-line-number="44"><span class="co">#&gt; # A tibble: 1,908 x 7</span></a>
<a class="sourceLine" id="cb5-45" data-line-number="45"><span class="co">#&gt;    gene         baseMean log2FoldChange lfcSE  stat    pvalue      padj</span></a>
<a class="sourceLine" id="cb5-46" data-line-number="46"><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-47" data-line-number="47"><span class="co">#&gt;  1 MAGEC2          133.           11.2  1.20   9.33 1.06e- 20 9.57e- 19</span></a>
<a class="sourceLine" id="cb5-48" data-line-number="48"><span class="co">#&gt;  2 PAGE5            76.5          10.7  1.20   8.90 5.57e- 19 4.23e- 17</span></a>
<a class="sourceLine" id="cb5-49" data-line-number="49"><span class="co">#&gt;  3 PGR            1732.            9.44 0.316 29.9  9.03e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb5-50" data-line-number="50"><span class="co">#&gt;  4 FAM133A          27.2           9.24 1.24   7.44 1.01e- 13 4.05e- 12</span></a>
<a class="sourceLine" id="cb5-51" data-line-number="51"><span class="co">#&gt;  5 ZFP3             61.4           9.15 1.21   7.54 4.64e- 14 1.96e- 12</span></a>
<a class="sourceLine" id="cb5-52" data-line-number="52"><span class="co">#&gt;  6 GREB1          2343.            9.08 0.303 29.9  8.27e-197 7.88e-193</span></a>
<a class="sourceLine" id="cb5-53" data-line-number="53"><span class="co">#&gt;  7 LOC102724756     18.0           8.66 1.29   6.69 2.25e- 11 5.99e- 10</span></a>
<a class="sourceLine" id="cb5-54" data-line-number="54"><span class="co">#&gt;  8 CYP2B7P        1839.            8.58 0.336 25.5  1.94e-143 8.47e-140</span></a>
<a class="sourceLine" id="cb5-55" data-line-number="55"><span class="co">#&gt;  9 CTAG2           153.            8.31 0.675 12.3  7.76e- 35 2.46e- 32</span></a>
<a class="sourceLine" id="cb5-56" data-line-number="56"><span class="co">#&gt; 10 TMLHE           188.            8.25 1.05   7.88 3.34e- 15 1.62e- 13</span></a>
<a class="sourceLine" id="cb5-57" data-line-number="57"><span class="co">#&gt; # ... with 1,898 more rows</span></a>
<a class="sourceLine" id="cb5-58" data-line-number="58"></a>
<a class="sourceLine" id="cb5-59" data-line-number="59">gene_tibble3 &lt;-<span class="st"> </span>gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">len =</span> <span class="kw">str_length</span>(gene))</a>
<a class="sourceLine" id="cb5-60" data-line-number="60">gene_tibble3</a>
<a class="sourceLine" id="cb5-61" data-line-number="61"><span class="co">#&gt; # A tibble: 1,908 x 8</span></a>
<a class="sourceLine" id="cb5-62" data-line-number="62"><span class="co">#&gt;    gene      baseMean log2FoldChange lfcSE  stat    pvalue      padj   len</span></a>
<a class="sourceLine" id="cb5-63" data-line-number="63"><span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;int&gt;</span></a>
<a class="sourceLine" id="cb5-64" data-line-number="64"><span class="co">#&gt;  1 MAGEC2       133.           11.2  1.20   9.33 1.06e- 20 9.57e- 19     6</span></a>
<a class="sourceLine" id="cb5-65" data-line-number="65"><span class="co">#&gt;  2 PAGE5         76.5          10.7  1.20   8.90 5.57e- 19 4.23e- 17     5</span></a>
<a class="sourceLine" id="cb5-66" data-line-number="66"><span class="co">#&gt;  3 PGR         1732.            9.44 0.316 29.9  9.03e-197 7.88e-193     3</span></a>
<a class="sourceLine" id="cb5-67" data-line-number="67"><span class="co">#&gt;  4 FAM133A       27.2           9.24 1.24   7.44 1.01e- 13 4.05e- 12     7</span></a>
<a class="sourceLine" id="cb5-68" data-line-number="68"><span class="co">#&gt;  5 ZFP3          61.4           9.15 1.21   7.54 4.64e- 14 1.96e- 12     4</span></a>
<a class="sourceLine" id="cb5-69" data-line-number="69"><span class="co">#&gt;  6 GREB1       2343.            9.08 0.303 29.9  8.27e-197 7.88e-193     5</span></a>
<a class="sourceLine" id="cb5-70" data-line-number="70"><span class="co">#&gt;  7 LOC10272…     18.0           8.66 1.29   6.69 2.25e- 11 5.99e- 10    12</span></a>
<a class="sourceLine" id="cb5-71" data-line-number="71"><span class="co">#&gt;  8 CYP2B7P     1839.            8.58 0.336 25.5  1.94e-143 8.47e-140     7</span></a>
<a class="sourceLine" id="cb5-72" data-line-number="72"><span class="co">#&gt;  9 CTAG2        153.            8.31 0.675 12.3  7.76e- 35 2.46e- 32     5</span></a>
<a class="sourceLine" id="cb5-73" data-line-number="73"><span class="co">#&gt; 10 TMLHE        188.            8.25 1.05   7.88 3.34e- 15 1.62e- 13     5</span></a>
<a class="sourceLine" id="cb5-74" data-line-number="74"><span class="co">#&gt; # ... with 1,898 more rows</span></a>
<a class="sourceLine" id="cb5-75" data-line-number="75"></a>
<a class="sourceLine" id="cb5-76" data-line-number="76"><span class="co"># "-AS" denotes antisense genes, we can find all those</span></a>
<a class="sourceLine" id="cb5-77" data-line-number="77">gene_tibble_as &lt;-<span class="st"> </span>gene_tibble <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="kw">str_detect</span>(gene, <span class="st">"-AS"</span>))</a>
<a class="sourceLine" id="cb5-78" data-line-number="78">gene_tibble_as</a>
<a class="sourceLine" id="cb5-79" data-line-number="79"><span class="co">#&gt; # A tibble: 23 x 7</span></a>
<a class="sourceLine" id="cb5-80" data-line-number="80"><span class="co">#&gt;    gene       baseMean log2FoldChange lfcSE  stat   pvalue     padj</span></a>
<a class="sourceLine" id="cb5-81" data-line-number="81"><span class="co">#&gt;    &lt;chr&gt;         &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-82" data-line-number="82"><span class="co">#&gt;  1 ELOVL2-AS1     5.47           6.71 1.41   4.75 2.05e- 6 2.20e- 5</span></a>
<a class="sourceLine" id="cb5-83" data-line-number="83"><span class="co">#&gt;  2 GAS6-AS2      19.5            2.48 0.478  5.18 2.19e- 7 2.85e- 6</span></a>
<a class="sourceLine" id="cb5-84" data-line-number="84"><span class="co">#&gt;  3 ZEB1-AS1      26.3            2.22 0.423  5.26 1.45e- 7 1.96e- 6</span></a>
<a class="sourceLine" id="cb5-85" data-line-number="85"><span class="co">#&gt;  4 FGF13-AS1     12.1            1.99 0.576  3.46 5.44e- 4 3.06e- 3</span></a>
<a class="sourceLine" id="cb5-86" data-line-number="86"><span class="co">#&gt;  5 C1RL-AS1      30.9            1.49 0.416  3.58 3.46e- 4 2.07e- 3</span></a>
<a class="sourceLine" id="cb5-87" data-line-number="87"><span class="co">#&gt;  6 RAP2C-AS1     38.5            1.48 0.382  3.87 1.07e- 4 7.50e- 4</span></a>
<a class="sourceLine" id="cb5-88" data-line-number="88"><span class="co">#&gt;  7 KANSL1-AS1    30.8            1.45 0.455  3.19 1.42e- 3 6.91e- 3</span></a>
<a class="sourceLine" id="cb5-89" data-line-number="89"><span class="co">#&gt;  8 DPP10-AS1    115.             1.20 0.254  4.71 2.42e- 6 2.58e- 5</span></a>
<a class="sourceLine" id="cb5-90" data-line-number="90"><span class="co">#&gt;  9 CCNT2-AS1     43.7            1.18 0.319  3.70 2.14e- 4 1.37e- 3</span></a>
<a class="sourceLine" id="cb5-91" data-line-number="91"><span class="co">#&gt; 10 DHRS4-AS1    694.             1.13 0.170  6.65 2.98e-11 7.79e-10</span></a>
<a class="sourceLine" id="cb5-92" data-line-number="92"><span class="co">#&gt; # ... with 13 more rows</span></a>
<a class="sourceLine" id="cb5-93" data-line-number="93"></a>
<a class="sourceLine" id="cb5-94" data-line-number="94"><span class="co"># also add annotations of our own</span></a>
<a class="sourceLine" id="cb5-95" data-line-number="95">gene_tibble_as2 &lt;-<span class="st"> </span>gene_tibble_as <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">str_c</span>(<span class="st">"HUMAN-"</span>, gene))</a>
<a class="sourceLine" id="cb5-96" data-line-number="96">gene_tibble_as2</a>
<a class="sourceLine" id="cb5-97" data-line-number="97"><span class="co">#&gt; # A tibble: 23 x 7</span></a>
<a class="sourceLine" id="cb5-98" data-line-number="98"><span class="co">#&gt;    gene             baseMean log2FoldChange lfcSE  stat   pvalue     padj</span></a>
<a class="sourceLine" id="cb5-99" data-line-number="99"><span class="co">#&gt;    &lt;chr&gt;               &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb5-100" data-line-number="100"><span class="co">#&gt;  1 HUMAN-ELOVL2-AS1     5.47           6.71 1.41   4.75 2.05e- 6 2.20e- 5</span></a>
<a class="sourceLine" id="cb5-101" data-line-number="101"><span class="co">#&gt;  2 HUMAN-GAS6-AS2      19.5            2.48 0.478  5.18 2.19e- 7 2.85e- 6</span></a>
<a class="sourceLine" id="cb5-102" data-line-number="102"><span class="co">#&gt;  3 HUMAN-ZEB1-AS1      26.3            2.22 0.423  5.26 1.45e- 7 1.96e- 6</span></a>
<a class="sourceLine" id="cb5-103" data-line-number="103"><span class="co">#&gt;  4 HUMAN-FGF13-AS1     12.1            1.99 0.576  3.46 5.44e- 4 3.06e- 3</span></a>
<a class="sourceLine" id="cb5-104" data-line-number="104"><span class="co">#&gt;  5 HUMAN-C1RL-AS1      30.9            1.49 0.416  3.58 3.46e- 4 2.07e- 3</span></a>
<a class="sourceLine" id="cb5-105" data-line-number="105"><span class="co">#&gt;  6 HUMAN-RAP2C-AS1     38.5            1.48 0.382  3.87 1.07e- 4 7.50e- 4</span></a>
<a class="sourceLine" id="cb5-106" data-line-number="106"><span class="co">#&gt;  7 HUMAN-KANSL1-AS1    30.8            1.45 0.455  3.19 1.42e- 3 6.91e- 3</span></a>
<a class="sourceLine" id="cb5-107" data-line-number="107"><span class="co">#&gt;  8 HUMAN-DPP10-AS1    115.             1.20 0.254  4.71 2.42e- 6 2.58e- 5</span></a>
<a class="sourceLine" id="cb5-108" data-line-number="108"><span class="co">#&gt;  9 HUMAN-CCNT2-AS1     43.7            1.18 0.319  3.70 2.14e- 4 1.37e- 3</span></a>
<a class="sourceLine" id="cb5-109" data-line-number="109"><span class="co">#&gt; 10 HUMAN-DHRS4-AS1    694.             1.13 0.170  6.65 2.98e-11 7.79e-10</span></a>
<a class="sourceLine" id="cb5-110" data-line-number="110"><span class="co">#&gt; # ... with 13 more rows</span></a></code></pre></div>
</div>
<div id="various-conversions-in-r---biomart" class="section level2">
<h2 class="hasAnchor">
<a href="#various-conversions-in-r---biomart" class="anchor"></a>various conversions in R - BioMart</h2>
<p>BioMart can be used outside of R too: <a href="http://www.ensembl.org/biomart/martview" class="uri">http://www.ensembl.org/biomart/martview</a></p>
<p>We will be using <code>biomaRt</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># BiocManager::install("biomaRt", version = "3.8")</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(biomaRt)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co"># biomaRt links to many useful datasets</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/listDatasets">listDatasets</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>, <span class="dt">host =</span> <span class="st">"www.ensembl.org"</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() <span class="co"># hundreds of organisms</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; # A tibble: 138 x 3</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt;    dataset               description                     version          </span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt;    &lt;I(chr)&gt;              &lt;I(chr)&gt;                        &lt;I(chr)&gt;         </span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt;  1 acalliptera_gene_ens… Eastern happy genes (fAstCal1.… fAstCal1.2       </span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt;  2 acarolinensis_gene_e… Anole lizard genes (AnoCar2.0)  AnoCar2.0        </span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt;  3 acitrinellus_gene_en… Midas cichlid genes (Midas_v5)  Midas_v5         </span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt;  4 amelanoleuca_gene_en… Panda genes (ailMel1)           ailMel1          </span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt;  5 amexicanus_gene_ense… Cave fish genes (Astyanax_mexi… Astyanax_mexican…</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt;  6 anancymaae_gene_ense… Ma's night monkey genes (Anan_… Anan_2.0         </span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="co">#&gt;  7 aocellaris_gene_ense… Clown anemonefish genes (AmpOc… AmpOce1.0        </span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="co">#&gt;  8 apercula_gene_ensembl Orange clownfish genes (Nemo_v… Nemo_v1          </span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="co">#&gt;  9 aplatyrhynchos_gene_… Duck genes (BGI_duck_1.0)       BGI_duck_1.0     </span></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"><span class="co">#&gt; 10 apolyacanthus_gene_e… Spiny chromis genes (ASM210954… ASM210954v1      </span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19"><span class="co">#&gt; # ... with 128 more rows</span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20"></a>
<a class="sourceLine" id="cb6-21" data-line-number="21"><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/listAttributes">listAttributes</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() <span class="co"># thousands of rows/attributes</span></a>
<a class="sourceLine" id="cb6-22" data-line-number="22"><span class="co">#&gt; # A tibble: 2,579 x 3</span></a>
<a class="sourceLine" id="cb6-23" data-line-number="23"><span class="co">#&gt;    name                          description                  page        </span></a>
<a class="sourceLine" id="cb6-24" data-line-number="24"><span class="co">#&gt;    &lt;chr&gt;                         &lt;chr&gt;                        &lt;chr&gt;       </span></a>
<a class="sourceLine" id="cb6-25" data-line-number="25"><span class="co">#&gt;  1 ensembl_gene_id               Gene stable ID               feature_page</span></a>
<a class="sourceLine" id="cb6-26" data-line-number="26"><span class="co">#&gt;  2 ensembl_gene_id_version       Gene stable ID version       feature_page</span></a>
<a class="sourceLine" id="cb6-27" data-line-number="27"><span class="co">#&gt;  3 ensembl_transcript_id         Transcript stable ID         feature_page</span></a>
<a class="sourceLine" id="cb6-28" data-line-number="28"><span class="co">#&gt;  4 ensembl_transcript_id_version Transcript stable ID version feature_page</span></a>
<a class="sourceLine" id="cb6-29" data-line-number="29"><span class="co">#&gt;  5 ensembl_peptide_id            Protein stable ID            feature_page</span></a>
<a class="sourceLine" id="cb6-30" data-line-number="30"><span class="co">#&gt;  6 ensembl_peptide_id_version    Protein stable ID version    feature_page</span></a>
<a class="sourceLine" id="cb6-31" data-line-number="31"><span class="co">#&gt;  7 ensembl_exon_id               Exon stable ID               feature_page</span></a>
<a class="sourceLine" id="cb6-32" data-line-number="32"><span class="co">#&gt;  8 description                   Gene description             feature_page</span></a>
<a class="sourceLine" id="cb6-33" data-line-number="33"><span class="co">#&gt;  9 chromosome_name               Chromosome/scaffold name     feature_page</span></a>
<a class="sourceLine" id="cb6-34" data-line-number="34"><span class="co">#&gt; 10 start_position                Gene start (bp)              feature_page</span></a>
<a class="sourceLine" id="cb6-35" data-line-number="35"><span class="co">#&gt; # ... with 2,569 more rows</span></a>
<a class="sourceLine" id="cb6-36" data-line-number="36"></a>
<a class="sourceLine" id="cb6-37" data-line-number="37"><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/listFilters">listFilters</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() <span class="co"># ways to filter the data</span></a>
<a class="sourceLine" id="cb6-38" data-line-number="38"><span class="co">#&gt; # A tibble: 373 x 2</span></a>
<a class="sourceLine" id="cb6-39" data-line-number="39"><span class="co">#&gt;    name               description                           </span></a>
<a class="sourceLine" id="cb6-40" data-line-number="40"><span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;                                 </span></a>
<a class="sourceLine" id="cb6-41" data-line-number="41"><span class="co">#&gt;  1 chromosome_name    Chromosome/scaffold name              </span></a>
<a class="sourceLine" id="cb6-42" data-line-number="42"><span class="co">#&gt;  2 start              Start                                 </span></a>
<a class="sourceLine" id="cb6-43" data-line-number="43"><span class="co">#&gt;  3 end                End                                   </span></a>
<a class="sourceLine" id="cb6-44" data-line-number="44"><span class="co">#&gt;  4 band_start         Band Start                            </span></a>
<a class="sourceLine" id="cb6-45" data-line-number="45"><span class="co">#&gt;  5 band_end           Band End                              </span></a>
<a class="sourceLine" id="cb6-46" data-line-number="46"><span class="co">#&gt;  6 marker_start       Marker Start                          </span></a>
<a class="sourceLine" id="cb6-47" data-line-number="47"><span class="co">#&gt;  7 marker_end         Marker End                            </span></a>
<a class="sourceLine" id="cb6-48" data-line-number="48"><span class="co">#&gt;  8 encode_region      Encode region                         </span></a>
<a class="sourceLine" id="cb6-49" data-line-number="49"><span class="co">#&gt;  9 strand             Strand                                </span></a>
<a class="sourceLine" id="cb6-50" data-line-number="50"><span class="co">#&gt; 10 chromosomal_region e.g. 1:100:10000:-1, 1:100000:200000:1</span></a>
<a class="sourceLine" id="cb6-51" data-line-number="51"><span class="co">#&gt; # ... with 363 more rows</span></a></code></pre></div>
<p>common usage 1: id to gene symbol conversion</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># gene id to gene symbol conversion</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">human_mart &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">ensg_sym &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"hgnc_symbol"</span>, <span class="st">"ensembl_gene_id"</span>), </a>
<a class="sourceLine" id="cb7-4" data-line-number="4">                 <span class="dt">filters =</span> <span class="st">"ensembl_gene_id"</span>, </a>
<a class="sourceLine" id="cb7-5" data-line-number="5">                 <span class="dt">mart =</span> human_mart, </a>
<a class="sourceLine" id="cb7-6" data-line-number="6">                 <span class="dt">values =</span> ensg_tibble<span class="op">$</span>gene) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() <span class="co"># with . version #, can't find matches</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7">ensg_sym</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt; # A tibble: 0 x 2</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt; # ... with 2 variables: hgnc_symbol &lt;lgl&gt;, ensembl_gene_id &lt;lgl&gt;</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"></a>
<a class="sourceLine" id="cb7-11" data-line-number="11">ensg_sym &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"hgnc_symbol"</span>, <span class="st">"ensembl_gene_id"</span>), </a>
<a class="sourceLine" id="cb7-12" data-line-number="12">                 <span class="dt">filters =</span> <span class="st">"ensembl_gene_id"</span>, </a>
<a class="sourceLine" id="cb7-13" data-line-number="13">                 <span class="dt">mart =</span> human_mart, </a>
<a class="sourceLine" id="cb7-14" data-line-number="14">                 <span class="dt">values =</span> ensg_tibble2<span class="op">$</span>gene) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb7-15" data-line-number="15">ensg_sym</a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">#&gt; # A tibble: 5 x 2</span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co">#&gt;   hgnc_symbol  ensembl_gene_id</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18"><span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;          </span></a>
<a class="sourceLine" id="cb7-19" data-line-number="19"><span class="co">#&gt; 1 TNFRSF9      ENSG00000049249</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20"><span class="co">#&gt; 2 SUGT1P1      ENSG00000226823</span></a>
<a class="sourceLine" id="cb7-21" data-line-number="21"><span class="co">#&gt; 3 PIP5K1P2     ENSG00000229435</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22"><span class="co">#&gt; 4 TNFRSF14-AS1 ENSG00000238164</span></a>
<a class="sourceLine" id="cb7-23" data-line-number="23"><span class="co">#&gt; 5 ""           ENSG00000266470</span></a></code></pre></div>
<p>common usage 2: functional GO term annotation</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/listAttributes">listAttributes</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="kw">str_detect</span>(description, <span class="st">"GO"</span>)) <span class="co"># look at attributes related to GO </span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; # A tibble: 7 x 3</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#&gt;   name                   description             page        </span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="co">#&gt;   &lt;chr&gt;                  &lt;chr&gt;                   &lt;chr&gt;       </span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co">#&gt; 1 go_id                  GO term accession       feature_page</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#&gt; 2 name_1006              GO term name            feature_page</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="co">#&gt; 3 definition_1006        GO term definition      feature_page</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#&gt; 4 go_linkage_type        GO term evidence code   feature_page</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#&gt; 5 namespace_1003         GO domain               feature_page</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">#&gt; 6 goslim_goa_accession   GOSlim GOA Accession(s) feature_page</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">#&gt; 7 goslim_goa_description GOSlim GOA Description  feature_page</span></a></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">drerio_mart &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ensembl"</span>, <span class="dt">dataset =</span> <span class="st">"drerio_gene_ensembl"</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">loc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"genes_drerio.csv"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">genes_drerio &lt;-<span class="st"> </span><span class="kw">read_csv</span>(loc)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">genes_drerio</a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt; # A tibble: 20 x 2</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt;    gene       p_val_adj</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt;    &lt;chr&gt;          &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">#&gt;  1 tbx18      1.36e-275</span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">#&gt;  2 tmsb4x     5.40e-222</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="co">#&gt;  3 tjp1a      4.89e-211</span></a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt;  4 zgc:77517  1.36e-206</span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">#&gt;  5 tgfbi      1.54e-195</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co">#&gt;  6 tubb5      7.41e-180</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co">#&gt;  7 tnc        1.58e-176</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"><span class="co">#&gt;  8 tpm4a      1.13e-172</span></a>
<a class="sourceLine" id="cb9-17" data-line-number="17"><span class="co">#&gt;  9 zgc:162612 5.63e-132</span></a>
<a class="sourceLine" id="cb9-18" data-line-number="18"><span class="co">#&gt; 10 zgc:153867 1.27e-127</span></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="co">#&gt; 11 ywhaz      2.34e-125</span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="co">#&gt; 12 tnn        5.90e-120</span></a>
<a class="sourceLine" id="cb9-21" data-line-number="21"><span class="co">#&gt; 13 tuba8l3    2.63e-117</span></a>
<a class="sourceLine" id="cb9-22" data-line-number="22"><span class="co">#&gt; 14 tmem176    4.77e-104</span></a>
<a class="sourceLine" id="cb9-23" data-line-number="23"><span class="co">#&gt; 15 tpm3       1.20e- 88</span></a>
<a class="sourceLine" id="cb9-24" data-line-number="24"><span class="co">#&gt; 16 xbp1       1.71e- 88</span></a>
<a class="sourceLine" id="cb9-25" data-line-number="25"><span class="co">#&gt; 17 tmem258    2.58e- 76</span></a>
<a class="sourceLine" id="cb9-26" data-line-number="26"><span class="co">#&gt; 18 tuba1a     4.09e- 71</span></a>
<a class="sourceLine" id="cb9-27" data-line-number="27"><span class="co">#&gt; 19 tbx2a      1.87e- 60</span></a>
<a class="sourceLine" id="cb9-28" data-line-number="28"><span class="co">#&gt; 20 wu:fc21g02 4.14e- 54</span></a>
<a class="sourceLine" id="cb9-29" data-line-number="29"></a>
<a class="sourceLine" id="cb9-30" data-line-number="30">genesGO &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"zfin_id_symbol"</span>, <span class="st">"go_id"</span>, <span class="st">"name_1006"</span>, <span class="st">"namespace_1003"</span>), </a>
<a class="sourceLine" id="cb9-31" data-line-number="31">                 <span class="dt">filters =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"zfin_id_symbol"</span>, <span class="st">"with_go"</span>), </a>
<a class="sourceLine" id="cb9-32" data-line-number="32">                 <span class="dt">mart =</span> drerio_mart, </a>
<a class="sourceLine" id="cb9-33" data-line-number="33">                 <span class="dt">values =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(genes_drerio<span class="op">$</span>gene, <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb9-34" data-line-number="34">genesGO</a>
<a class="sourceLine" id="cb9-35" data-line-number="35"><span class="co">#&gt; # A tibble: 129 x 4</span></a>
<a class="sourceLine" id="cb9-36" data-line-number="36"><span class="co">#&gt;    zfin_id_symbol go_id     name_1006                     namespace_1003  </span></a>
<a class="sourceLine" id="cb9-37" data-line-number="37"><span class="co">#&gt;    &lt;chr&gt;          &lt;chr&gt;     &lt;chr&gt;                         &lt;chr&gt;           </span></a>
<a class="sourceLine" id="cb9-38" data-line-number="38"><span class="co">#&gt;  1 tbx18          GO:00036… DNA binding                   molecular_funct…</span></a>
<a class="sourceLine" id="cb9-39" data-line-number="39"><span class="co">#&gt;  2 tbx18          GO:00037… DNA-binding transcription fa… molecular_funct…</span></a>
<a class="sourceLine" id="cb9-40" data-line-number="40"><span class="co">#&gt;  3 tbx18          GO:00063… regulation of transcription,… biological_proc…</span></a>
<a class="sourceLine" id="cb9-41" data-line-number="41"><span class="co">#&gt;  4 tbx18          GO:00056… nucleus                       cellular_compon…</span></a>
<a class="sourceLine" id="cb9-42" data-line-number="42"><span class="co">#&gt;  5 tbx2a          GO:00036… DNA binding                   molecular_funct…</span></a>
<a class="sourceLine" id="cb9-43" data-line-number="43"><span class="co">#&gt;  6 tbx2a          GO:00037… DNA-binding transcription fa… molecular_funct…</span></a>
<a class="sourceLine" id="cb9-44" data-line-number="44"><span class="co">#&gt;  7 tbx2a          GO:00063… regulation of transcription,… biological_proc…</span></a>
<a class="sourceLine" id="cb9-45" data-line-number="45"><span class="co">#&gt;  8 tbx2a          GO:00056… nucleus                       cellular_compon…</span></a>
<a class="sourceLine" id="cb9-46" data-line-number="46"><span class="co">#&gt;  9 tbx2a          GO:00055… protein binding               molecular_funct…</span></a>
<a class="sourceLine" id="cb9-47" data-line-number="47"><span class="co">#&gt; 10 tbx2a          GO:00487… pronephros development        biological_proc…</span></a>
<a class="sourceLine" id="cb9-48" data-line-number="48"><span class="co">#&gt; # ... with 119 more rows</span></a>
<a class="sourceLine" id="cb9-49" data-line-number="49"></a>
<a class="sourceLine" id="cb9-50" data-line-number="50"><span class="co"># join GO dataframe with original RNA-seq/scRNA-seq data if needed</span></a>
<a class="sourceLine" id="cb9-51" data-line-number="51">genes_drerio_joinedGO &lt;-<span class="st"> </span><span class="kw">left_join</span>(genes_drerio, genesGO, </a>
<a class="sourceLine" id="cb9-52" data-line-number="52">                                   <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"gene"</span> =<span class="st"> "zfin_id_symbol"</span>))</a>
<a class="sourceLine" id="cb9-53" data-line-number="53">genes_drerio_joinedGO</a>
<a class="sourceLine" id="cb9-54" data-line-number="54"><span class="co">#&gt; # A tibble: 130 x 5</span></a>
<a class="sourceLine" id="cb9-55" data-line-number="55"><span class="co">#&gt;    gene   p_val_adj go_id    name_1006                    namespace_1003  </span></a>
<a class="sourceLine" id="cb9-56" data-line-number="56"><span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;                        &lt;chr&gt;           </span></a>
<a class="sourceLine" id="cb9-57" data-line-number="57"><span class="co">#&gt;  1 tbx18  1.36e-275 GO:0003… DNA binding                  molecular_funct…</span></a>
<a class="sourceLine" id="cb9-58" data-line-number="58"><span class="co">#&gt;  2 tbx18  1.36e-275 GO:0003… DNA-binding transcription f… molecular_funct…</span></a>
<a class="sourceLine" id="cb9-59" data-line-number="59"><span class="co">#&gt;  3 tbx18  1.36e-275 GO:0006… regulation of transcription… biological_proc…</span></a>
<a class="sourceLine" id="cb9-60" data-line-number="60"><span class="co">#&gt;  4 tbx18  1.36e-275 GO:0005… nucleus                      cellular_compon…</span></a>
<a class="sourceLine" id="cb9-61" data-line-number="61"><span class="co">#&gt;  5 tmsb4x 5.40e-222 GO:0003… actin monomer binding        molecular_funct…</span></a>
<a class="sourceLine" id="cb9-62" data-line-number="62"><span class="co">#&gt;  6 tmsb4x 5.40e-222 GO:0007… actin filament organization  biological_proc…</span></a>
<a class="sourceLine" id="cb9-63" data-line-number="63"><span class="co">#&gt;  7 tmsb4x 5.40e-222 GO:0030… regulation of cell migration biological_proc…</span></a>
<a class="sourceLine" id="cb9-64" data-line-number="64"><span class="co">#&gt;  8 tmsb4x 5.40e-222 GO:0042… sequestering of actin monom… biological_proc…</span></a>
<a class="sourceLine" id="cb9-65" data-line-number="65"><span class="co">#&gt;  9 tmsb4x 5.40e-222 GO:0005… cytoplasm                    cellular_compon…</span></a>
<a class="sourceLine" id="cb9-66" data-line-number="66"><span class="co">#&gt; 10 tmsb4x 5.40e-222 GO:0001… heart looping                biological_proc…</span></a>
<a class="sourceLine" id="cb9-67" data-line-number="67"><span class="co">#&gt; # ... with 120 more rows</span></a>
<a class="sourceLine" id="cb9-68" data-line-number="68"></a>
<a class="sourceLine" id="cb9-69" data-line-number="69"><span class="co"># or, get all genes with GO term of interest</span></a>
<a class="sourceLine" id="cb9-70" data-line-number="70">genesGO_filtered &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"zfin_id_symbol"</span>), </a>
<a class="sourceLine" id="cb9-71" data-line-number="71">                  <span class="dt">filters =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"zfin_id_symbol"</span>, <span class="st">"go"</span>), </a>
<a class="sourceLine" id="cb9-72" data-line-number="72">                  <span class="dt">mart =</span> drerio_mart, </a>
<a class="sourceLine" id="cb9-73" data-line-number="73">                  <span class="dt">values =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(genes_drerio<span class="op">$</span>gene, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"GO:0003677"</span>)))</a>
<a class="sourceLine" id="cb9-74" data-line-number="74">genesGO_filtered</a>
<a class="sourceLine" id="cb9-75" data-line-number="75"><span class="co">#&gt;   zfin_id_symbol</span></a>
<a class="sourceLine" id="cb9-76" data-line-number="76"><span class="co">#&gt; 1          tbx18</span></a>
<a class="sourceLine" id="cb9-77" data-line-number="77"><span class="co">#&gt; 2          tbx2a</span></a>
<a class="sourceLine" id="cb9-78" data-line-number="78"><span class="co">#&gt; 3     zgc:162612</span></a></code></pre></div>
<p>common usage 3: ortholog conversion between species</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># example of zebrafish to human gene symbol conversion</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">genes_human &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getLDS">getLDS</a></span>(<span class="dt">attributes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"hgnc_symbol"</span>), </a>
<a class="sourceLine" id="cb10-3" data-line-number="3">                      <span class="dt">filters =</span> <span class="st">"hgnc_symbol"</span>, </a>
<a class="sourceLine" id="cb10-4" data-line-number="4">                      <span class="dt">mart =</span> human_mart, </a>
<a class="sourceLine" id="cb10-5" data-line-number="5">                      <span class="dt">attributesL =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"zfin_id_symbol"</span>), </a>
<a class="sourceLine" id="cb10-6" data-line-number="6">                      <span class="dt">martL =</span> drerio_mart, </a>
<a class="sourceLine" id="cb10-7" data-line-number="7">                      <span class="dt">uniqueRows=</span>T, </a>
<a class="sourceLine" id="cb10-8" data-line-number="8">                      <span class="dt">values =</span> genes_drerio<span class="op">$</span>gene) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb10-9" data-line-number="9">genes_human</a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="co">#&gt; # A tibble: 8 x 2</span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="co">#&gt;   HGNC.symbol ZFIN.symbol</span></a>
<a class="sourceLine" id="cb10-12" data-line-number="12"><span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;      </span></a>
<a class="sourceLine" id="cb10-13" data-line-number="13"><span class="co">#&gt; 1 XBP1        xbp1       </span></a>
<a class="sourceLine" id="cb10-14" data-line-number="14"><span class="co">#&gt; 2 TPM3        tpm3       </span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15"><span class="co">#&gt; 3 TGFBI       tgfbi      </span></a>
<a class="sourceLine" id="cb10-16" data-line-number="16"><span class="co">#&gt; 4 TBX18       tbx18      </span></a>
<a class="sourceLine" id="cb10-17" data-line-number="17"><span class="co">#&gt; 5 YWHAZ       ywhabb     </span></a>
<a class="sourceLine" id="cb10-18" data-line-number="18"><span class="co">#&gt; 6 YWHAZ       ywhaba     </span></a>
<a class="sourceLine" id="cb10-19" data-line-number="19"><span class="co">#&gt; 7 TMEM258     tmem258    </span></a>
<a class="sourceLine" id="cb10-20" data-line-number="20"><span class="co">#&gt; 8 TUBA1A      tuba1a</span></a></code></pre></div>
</div>
<div id="exercise-2" class="section level2">
<h2 class="hasAnchor">
<a href="#exercise-2" class="anchor"></a>Exercise 2</h2>
<ol style="list-style-type: decimal">
<li><p>Delete one argument from code above used to generate <code>genesGO_filtered</code> to obtain a list of all genes with GO id of “<a href="GO:0003677" class="uri">GO:0003677</a>”.</p></li>
<li><p>Take a look at <code>genesGO</code>. How would you add GO-CC back into the brauer_gene_exp?</p></li>
</ol>
</div>
<div id="additional-resourcesreading" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-resourcesreading" class="anchor"></a>Additional resources/reading</h2>
<p>More sophisticated analysis than simply gene lists: Find enriched sequence elements in promoters, UTRs, etc. See <a href="https://bedtools.readthedocs.io/en/latest/" class="uri">https://bedtools.readthedocs.io/en/latest/</a>, <a href="http://meme-suite.org/" class="uri">http://meme-suite.org/</a>, etc</p>
<p>WGCNA: weighted correlation network analysis</p>
<p>SCENIC: infer gene regulatory networks and cell types from single-cell RNA-seq data</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#did-experiment-ran-through-analysis-pipeline-what-now">Did experiment, ran through analysis pipeline, what now?</a></li>
      <li><a href="#retrieve-gene-lists-from-rna-seq-or-scrna-seq">Retrieve gene lists from RNA-seq or scRNA-seq</a></li>
      <li><a href="#feed-gene-lists-to-other-software-outside-of-r-get-enrichment-score-back">Feed gene lists to other software outside of R, get enrichment score back</a></li>
      <li><a href="#sometimes-transformation-of-names-is-needed---stringr-package-part-of-tidyverse">Sometimes transformation of names is needed - <code>stringr</code> package, part of tidyverse</a></li>
      <li><a href="#various-conversions-in-r---biomart">various conversions in R - BioMart</a></li>
      <li><a href="#exercise-2">Exercise 2</a></li>
      <li><a href="#additional-resourcesreading">Additional resources/reading</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy, Rui Fu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
