<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Problem Set 4 • pbda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Problem Set 4">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pbda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/practical-data-analysis">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Problem Set 4</h1>
                        <h4 class="author">Your Name Here</h4>
            
            <h4 class="date">12/11/2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/master/vignettes/problem-set-4.Rmd"><code>vignettes/problem-set-4.Rmd</code></a></small>
      <div class="hidden name"><code>problem-set-4.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>Here is a <a href="https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/vignettes/problem-set-4.Rmd">link</a> to the text of these exercises.</p>
</div>
<div id="question-1" class="section level1">
<h1 class="hasAnchor">
<a href="#question-1" class="anchor"></a>Question 1</h1>
<p>A second 10X Genomics single cell RNA-seq dataset is included in the pbda package. This dataset is derived from a human solid tumor. Get its path using the code below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">tumor_10x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"tumor_gene_bc_matrices"</span>, <span class="st">"hg19"</span>, <span class="dt">package =</span> <span class="st">"pbda"</span>)</a></code></pre></div>
<p>Using the new data, create a new Seurat object named <code>tumor</code>. Next, extract the list of mitochondrial genes in the matrix, calculate the percentage of reads that are derived from the mitochondrial genome in each cell, and add this to a metadata slot called “percent.mito” in the <code>tumor</code> object.</p>
<p>Now, using the <code>GenePlot()</code> function, plot nUMI vs. nGene and nUMI vs. percent.mito. Based on these plots, filter the data to remove dead cells and likely doublets (using <code>FilterCells()</code>).</p>
<div id="strategy" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-2" class="section level1">
<h1 class="hasAnchor">
<a href="#question-2" class="anchor"></a>Question 2</h1>
<p>Now that you’ve filtered your cells, normalize the data (<code>NormalizeData()</code>) then identify variable genes using the code below (the cutoffs have been adjusted for this dataset - don’t worry about interpreting the VMR plot):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">tumor &lt;-<span class="st"> </span><span class="kw">FindVariableGenes</span>(tumor,</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">                          <span class="dt">mean.function =</span> ExpMean,</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">                          <span class="dt">dispersion.function =</span> LogVMR, </a>
<a class="sourceLine" id="cb2-4" data-line-number="4">                          <span class="dt">x.low.cutoff =</span> <span class="fl">0.025</span>,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">                          <span class="dt">x.high.cutoff =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">                          <span class="dt">y.cutoff =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>Next, scale the data and run PCA on your <code>tumor</code> object using the variable genes you’ve just identified.</p>
<p>Finally, use an elbow plot and/or PC heatmaps to select the prinicpal components to use for tSNE and clustering. Justify your choice of PCs (do they capture the full variation present in the data?).</p>
<div id="strategy-1" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy-1" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-1" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-1" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-3" class="section level1">
<h1 class="hasAnchor">
<a href="#question-3" class="anchor"></a>Question 3</h1>
<p>Using the PCs selected in Question 2, run the tSNE algorithm on <code>tumor</code> and identify clusters using <code>FindClusters()</code>. Plot the resulting tSNE. How many clusters are identified? Does this appear reasonable based on the structure present in the tSNE projection?</p>
<div id="strategy-2" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy-2" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-2" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-2" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="question-4" class="section level1">
<h1 class="hasAnchor">
<a href="#question-4" class="anchor"></a>Question 4</h1>
<p>Included in the package is a matrix called <code>tx_rates</code> containing transcription rates for ~7000 genes. Transcription rates were measured every 15 minutes for 3 hours after exposure of dendritic cells to LPS, which activates innate immunity responses. The processed data was obtained from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56977">GEO</a></p>
<p>For this exercise perform the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Exclude genes that have rates of 0 across all samples</p></li>
<li><p>Log transform the data (add a pseudocount of 1 to avoid Infinite values)</p></li>
<li><p>Normalize the data to the 0 hour time point</p></li>
<li><p>run kmeans clustering on the log transformed and normalized matrix. Use k = between 5 and 10 clusters.</p></li>
<li><p>Make a heatmap with rows split based on the clusters identified in step 4. Use the code below to generate a color palette. Discuss any interesting patterns in your clusters in the interpretation.</p></li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">cols_to_use &lt;-<span class="st"> </span>circlize<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">0</span>, <span class="dv">4</span>), </a>
<a class="sourceLine" id="cb3-2" data-line-number="2">                                    <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co"># pseudocode</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw">Heatmap</span>(mat, <span class="dt">col =</span> cols_to_use, ...</a></code></pre></div>
<div id="strategy-3" class="section level2">
<h2 class="hasAnchor">
<a href="#strategy-3" class="anchor"></a>Strategy</h2>
</div>
<div id="interpretation-3" class="section level2">
<h2 class="hasAnchor">
<a href="#interpretation-3" class="anchor"></a>Interpretation</h2>
</div>
</div>
<div id="bonus-super-fun-but-optional-exercise" class="section level1">
<h1 class="hasAnchor">
<a href="#bonus-super-fun-but-optional-exercise" class="anchor"></a>Bonus super fun but optional exercise:</h1>
<p>Using your tidyverse skills plot the average transcription rate for each cluster identified in Q4 using ggplot. Facet by cluster. (Hint gather and tidy the matrix, then join with the cluster assignments)</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li>
<a href="#question-1">Question 1</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy">Strategy</a></li>
      <li><a href="#interpretation">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-2">Question 2</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy-1">Strategy</a></li>
      <li><a href="#interpretation-1">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-3">Question 3</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy-2">Strategy</a></li>
      <li><a href="#interpretation-2">Interpretation</a></li>
      </ul>
</li>
      <li>
<a href="#question-4">Question 4</a><ul class="nav nav-pills nav-stacked">
<li><a href="#strategy-3">Strategy</a></li>
      <li><a href="#interpretation-3">Interpretation</a></li>
      </ul>
</li>
      <li><a href="#bonus-super-fun-but-optional-exercise">Bonus super fun but optional exercise:</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy, Rui Fu, Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
