<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Problem Set 2 • pbda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Problem Set 2">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pbda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview, Data Import and Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Analyzing qPCR data</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Basic Data Types and Structures</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/practical-data-analysis">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Problem Set 2</h1>
                        <h4 class="author">YOUR NAME HERE</h4>
            
            <h4 class="date">11/30/2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/practical-data-analysis/blob/master/vignettes/problem-set-2.Rmd"><code>vignettes/problem-set-2.Rmd</code></a></small>
      <div class="hidden name"><code>problem-set-2.Rmd</code></div>

    </div>

    
    
<div id="analyze-a-qpcr-experiment" class="section level2">
<h2 class="hasAnchor">
<a href="#analyze-a-qpcr-experiment" class="anchor"></a>Analyze a qPCR experiment</h2>
<p>Here is a <a href="https://raw.githubusercontent.com/IDPT7810/practical-data-analysis/master/vignettes/problem-set-2.Rmd">link</a> to the Rmarkdown for this assignment. Download that and update your name at the top.</p>
<p>You are investigating how the transcription factor MAGIC regulates the expression of two genes WAND and POOF. You use wild-type and mutant lines with different forms of the MAGIC gene and examine the expression of MAGIC, WAND, and POOF. The experiment is a time-course after you add an activator of MAGIC. You collect 4 time points (0, 12, 24, 48 hour) with three biological replicates.</p>
<p>The cells have three mutant forms of MAGIC:</p>
<ul>
<li>a hypomorphic allele (<code>MAGIC-hypo</code>) that is known to reduce its activity</li>
<li>an siRNA-mediated knockdown of MAGIC (<code>MAGIC-siRNA</code>)</li>
<li>a deletion of the MAGIC gene (<code>MAGIC-null</code>)</li>
</ul>
<p>You collect expression data in a qPCR experiment in a 384-well plate (a common format for these experiments). The data in CSV format are available in the class package. One CSV contains expression levels and the other contains sample names coded as <code>cell_time_gene_rt_rep</code>.</p>
<p>This code will give you the path and file names of the CSVs.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># devtools::install_github('rnabioco/eda')</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">qpcr384_data_csv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"qpcr-data-384.csv"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">qpcr384_names_csv &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"qpcr-names-384.csv"</span>, <span class="dt">package =</span> <span class="st">'pbda'</span>)</a></code></pre></div>
<div id="exercise-1" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-1" class="anchor"></a>Exercise 1</h3>
<p>Use a function from the <a href="http://readr.tidyverse.org/"><code>readr</code></a> library–part of the tidyverse–to load the <strong>CSV</strong> data. You should have two new tibbles.</p>
</div>
<div id="exercise-2" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-2" class="anchor"></a>Exercise 2</h3>
<p>Inspect these tibbles and make note of their resemblance to a 384 plate. Tidy these tibbles into this format:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># A tibble: 128 x 6</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">cell  time  gene    rt   exp_mean      exp_var</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="dv">1</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span> ACTIN     <span class="op">-</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">0.000000e+00</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="dv">2</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span> ACTIN     <span class="op">+</span><span class="st">   </span><span class="fl">2.333333</span> <span class="fl">4.133333e-01</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="dv">3</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span> MAGIC     <span class="op">-</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">0.000000e+00</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="dv">4</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span> MAGIC     <span class="op">+</span><span class="st">  </span><span class="fl">13.000000</span> <span class="fl">9.750000e+00</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="dv">5</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span>  POOF     <span class="op">-</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">0.000000e+00</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="dv">6</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span>  POOF     <span class="op">+</span><span class="st"> </span><span class="fl">981.333333</span> <span class="fl">5.717973e+05</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="dv">7</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span>  WAND     <span class="op">-</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">0.000000e+00</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="dv">8</span> MAGIC<span class="op">-</span>hypo     <span class="dv">0</span>  WAND     <span class="op">+</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">1.900000e-01</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="dv">9</span> MAGIC<span class="op">-</span>hypo    <span class="dv">12</span> ACTIN     <span class="op">-</span><span class="st">   </span><span class="fl">1.000000</span> <span class="fl">0.000000e+00</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="dv">10</span> MAGIC<span class="op">-</span>hypo    <span class="dv">12</span> ACTIN     <span class="op">+</span><span class="st">  </span><span class="fl">10.000000</span> <span class="fl">1.600000e-01</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co"># ... with 118 more rows</span></a></code></pre></div>
<p>Note that this table does not have <code>row</code> and <code>col</code> (they have already been dropped) and that the replicates have been grouped and summarized by their mean (<code>exp_mean</code>) and variance (<code>exp_var</code>).</p>
</div>
<div id="exercise-3" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-3" class="anchor"></a>Exercise 3</h3>
<p>Thare are two sets of qPCR reactions: one where reverse transcriptase was added to the RNA sample, and one where it was not. The <code>rt</code> variable reflects this by noting samples with <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> and <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>.</p>
<p>Make two plots of the distribution of expression values for all sample values. In one plot, use <code>geom_histogram()</code> and facet by <code>rt</code>. In the other plot, change the x-axis scale using <code>scale_x_log10()</code>.</p>
<p>What do these plots tell you about the values from the <code>rt == "-"</code> samples?</p>
</div>
<div id="exercise-4" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-4" class="anchor"></a>Exercise 4</h3>
<p>Create a plot of expression versus time for each of the MAGIC cell types. At this point you can remove the <code>rt == "-"</code> controls. You will need to plot expression value on a log-scale to see differences.</p>
<ul>
<li><p>In which cell lines is the expression of <code>WAND</code> affected? In what way? Can you make a statement about this (e.g., XXX is required for YYY expression)?</p></li>
<li><p>Compare the data per-<code>cell</code> and per-<code>gene</code> by creating two separate plots by grouping (e.g., using facets) one or the other of those variables. Which of these is more useful to see differences in gene expression?</p></li>
</ul>
</div>
<div id="exercise-5" class="section level3">
<h3 class="hasAnchor">
<a href="#exercise-5" class="anchor"></a>Exercise 5</h3>
<p>Normalize the expression data dividing each of the MAGIC, WAND, and POOF value by the ACTIN values. You will need to use <code>spread()</code> to rearrange the data for calculation, and then <code>gather()</code> to reformat for plotting. Re-create the plots from question 4 with this normalized data. Did your interpretation change?</p>
</div>
</div>
<div id="exploratory-analysis-of-biological-data" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory-analysis-of-biological-data" class="anchor"></a>Exploratory analysis of biological data</h2>
<p>Explore the <code>brauer_gene_exp</code> and <code>yeast_protein_prop</code> data sets to address <strong>your own hypothesis</strong>. Here is an example.</p>
<div id="hypothesis" class="section level3">
<h3 class="hasAnchor">
<a href="#hypothesis" class="anchor"></a>Hypothesis</h3>
<p>Sulfate limitation of yeast causes downregulation of genes containing high levels of cysteine and methionine residues.</p>
</div>
<div id="approach" class="section level3">
<h3 class="hasAnchor">
<a href="#approach" class="anchor"></a>Approach</h3>
<ul>
<li>Calculate cysteine / methionine content for each gene using the <code>yeast_protein_prop</code> table.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a></code></pre></div>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✔ ggplot2 3.2.1     ✔ purrr   0.3.3
## ✔ tibble  2.1.3     ✔ dplyr   0.8.3
## ✔ tidyr   1.0.0     ✔ stringr 1.4.0
## ✔ readr   1.3.1     ✔ forcats 0.4.0</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pbda)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">sulfurs &lt;-<span class="st"> </span>yeast_prot_prop <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="st">  </span><span class="kw">select</span>(ORF, Cys, Met) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="st">  </span><span class="kw">group_by</span>(ORF) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">n_sulfur =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(Cys, Met)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="st">  </span><span class="kw">rename</span>(<span class="dt">systematic_name =</span> ORF) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="st">  </span><span class="kw">select</span>(systematic_name, n_sulfur)</a></code></pre></div>
<ul>
<li>Test whether gene expression and <code>n_sulfur</code> are correlated.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(broom)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">brauer_gene_exp <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="st">  </span><span class="kw">select</span>(systematic_name, nutrient, expression) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="st">  </span><span class="kw">left_join</span>(sulfurs) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>systematic_name) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="st">  </span><span class="kw">group_by</span>(nutrient) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="st">  </span><span class="kw">nest</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="st">  </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb8-10" data-line-number="10">    <span class="dt">cor =</span> <span class="kw">map</span>(data, <span class="op">~</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/cor.test.html">cor.test</a></span>(.<span class="op">$</span>n_sulfur, .<span class="op">$</span>expression)))</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="st">  </span><span class="kw">select</span>(nutrient, cor) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="st">  </span><span class="kw">unnest</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="st">  </span><span class="kw">arrange</span>(<span class="op">-</span>estimate)</a></code></pre></div>
<pre><code>## Joining, by = "systematic_name"</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(cor)`</code></pre>
<pre><code>## # A tibble: 6 x 9
## # Groups:   nutrient [6]
##   nutrient estimate statistic   p.value parameter conf.low conf.high method
##   &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt; 
## 1 Ammonia   0.127       23.2  5.00e-118     32989  0.116      0.137  Pears…
## 2 Leucine   0.0449       8.16 3.48e- 16     33027  0.0341     0.0556 Pears…
## 3 Phospha…  0.0420       7.63 2.47e- 14     32919  0.0312     0.0528 Pears…
## 4 Glucose   0.0367       6.68 2.49e- 11     32987  0.0260     0.0475 Pears…
## 5 Sulfate   0.00680      1.23 2.18e-  1     32751 -0.00403    0.0176 Pears…
## 6 Uracil   -0.0313      -5.67 1.44e-  8     32858 -0.0421    -0.0205 Pears…
## # … with 1 more variable: alternative &lt;chr&gt;</code></pre>
<ul>
<li>Identify genes that are repressed under sulfate starvation conditions. Use the linear modeling approach to identify genes with negative slopes.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(broom)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">models &lt;-<span class="st"> </span>brauer_gene_exp <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">  </span><span class="kw">select</span>(systematic_name<span class="op">:</span>expression) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="st">  </span><span class="kw">group_by</span>(systematic_name, nutrient) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="st">  </span><span class="kw">nest</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="st">  </span><span class="co"># head() %&gt;% </span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="st">  </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb12-8" data-line-number="8">    <span class="dt">model =</span> <span class="kw">map</span>(data, <span class="op">~</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(expression <span class="op">~</span><span class="st"> </span>rate, <span class="dt">data =</span> .x)))</a>
<a class="sourceLine" id="cb12-9" data-line-number="9">  ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-10" data-line-number="10"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>data)</a></code></pre></div>
<pre><code>## Warning in summary.lm(x): essentially perfect fit: summary may be unreliable</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># Now unnest and grab relevant data, i.e. genes with negative slopes</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">models &lt;-<span class="st"> </span>models <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="st">  </span><span class="kw">unnest</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(term <span class="op">==</span><span class="st"> "rate"</span> <span class="op">&amp;</span><span class="st"> </span>estimate <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(nutrient)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="st">  </span><span class="kw">select</span>(systematic_name, nutrient, estimate)</a></code></pre></div>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(model)`</code></pre>
<ul>
<li>Cross-reference the two tables above to count the overlaps. What is a good control here? Need to ask whether the downregulation of these genes is <em>specific</em> to sulfate limitation. Are these same genes affected by limitation of other nutrients?</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">combined &lt;-<span class="st"> </span><span class="kw">left_join</span>(sulfurs, models) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>systematic_name) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(nutrient))</a></code></pre></div>
<pre><code>## Joining, by = "systematic_name"</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(cowplot)</a></code></pre></div>
<pre><code>## 
## ********************************************************
## Note: As of version 1.0.0, cowplot does not change the
##   default ggplot2 theme anymore. To recover the previous
##   behavior, execute:
##   theme_set(theme_cowplot())
## ********************************************************</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">ggplot</span>(combined, <span class="kw">aes</span>(<span class="dt">x =</span> n_sulfur, <span class="dt">y =</span> estimate)) <span class="op">+</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="st">  </span><span class="kw">scale_x_log10</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="st">  </span><span class="kw">facet_grid</span>(<span class="op">~</span>nutrient)</a></code></pre></div>
<p><img src="problem-set-2_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div id="interpretation" class="section level3">
<h3 class="hasAnchor">
<a href="#interpretation" class="anchor"></a>Interpretation</h3>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#analyze-a-qpcr-experiment">Analyze a qPCR experiment</a></li>
      <li><a href="#exploratory-analysis-of-biological-data">Exploratory analysis of biological data</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy, Rui Fu, Ryan Sheridan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
